{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cf69005d-ff5d-4b9a-80ef-e2956ad7f370",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "#import pybedtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cd094c9b-46c7-45d6-b3bb-d58d2098afa2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method_descriptor:\n",
      "\n",
      "split(self, /, sep=None, maxsplit=-1)\n",
      "    Return a list of the words in the string, using sep as the delimiter string.\n",
      "    \n",
      "    sep\n",
      "      The delimiter according which to split the string.\n",
      "      None (the default value) means split according to any whitespace,\n",
      "      and discard empty strings from the result.\n",
      "    maxsplit\n",
      "      Maximum number of splits to do.\n",
      "      -1 (the default value) means no limit.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(str.split)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "024b4ec9-4dbc-46c9-a40e-5c8a5c4d25ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>genome</th>\n",
       "      <th>identifier</th>\n",
       "      <th>file</th>\n",
       "      <th>clip</th>\n",
       "      <th>scale</th>\n",
       "      <th>sum_stat</th>\n",
       "      <th>description</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF833POA</td>\n",
       "      <td>/home/drk/tillage/datasets/human/dnase/encode/...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>cerebellum male adult (27 years) and male adul...</td>\n",
       "      <td>DNASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF110QGM</td>\n",
       "      <td>/home/drk/tillage/datasets/human/dnase/encode/...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>frontal cortex male adult (27 years) and male ...</td>\n",
       "      <td>DNASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF880MKD</td>\n",
       "      <td>/home/drk/tillage/datasets/human/dnase/encode/...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>chorion</td>\n",
       "      <td>DNASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF463ZLQ</td>\n",
       "      <td>/home/drk/tillage/datasets/human/dnase/encode/...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>Ishikawa treated with 0.02% dimethyl sulfoxide...</td>\n",
       "      <td>DNASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF890OGQ</td>\n",
       "      <td>/home/drk/tillage/datasets/human/dnase/encode/...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>GM03348</td>\n",
       "      <td>DNASE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index  genome   identifier   \n",
       "0      0       0  ENCFF833POA  \\\n",
       "1      1       0  ENCFF110QGM   \n",
       "2      2       0  ENCFF880MKD   \n",
       "3      3       0  ENCFF463ZLQ   \n",
       "4      4       0  ENCFF890OGQ   \n",
       "\n",
       "                                                file  clip  scale sum_stat   \n",
       "0  /home/drk/tillage/datasets/human/dnase/encode/...    32      2     mean  \\\n",
       "1  /home/drk/tillage/datasets/human/dnase/encode/...    32      2     mean   \n",
       "2  /home/drk/tillage/datasets/human/dnase/encode/...    32      2     mean   \n",
       "3  /home/drk/tillage/datasets/human/dnase/encode/...    32      2     mean   \n",
       "4  /home/drk/tillage/datasets/human/dnase/encode/...    32      2     mean   \n",
       "\n",
       "                                         description    exp  \n",
       "0  cerebellum male adult (27 years) and male adul...  DNASE  \n",
       "1  frontal cortex male adult (27 years) and male ...  DNASE  \n",
       "2                                            chorion  DNASE  \n",
       "3  Ishikawa treated with 0.02% dimethyl sulfoxide...  DNASE  \n",
       "4                                            GM03348  DNASE  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in targets\n",
    "data_dir = \"/omics/groups/OE0540/internal/users/mikulik/master_thesis/data/\"\n",
    "targets = pd.read_csv(f\"{data_dir}gcs_basenj/targets_human.txt\", sep=\"\\t\")\n",
    "targets[[\"exp\", \"description\"]] = targets.description.str.split(\":\",n=1, expand=True)#[0]\n",
    "targets.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b8ce8b18-a91b-48a0-81cd-9589772dec07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>genome</th>\n",
       "      <th>identifier</th>\n",
       "      <th>file</th>\n",
       "      <th>clip</th>\n",
       "      <th>scale</th>\n",
       "      <th>sum_stat</th>\n",
       "      <th>description</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4949</th>\n",
       "      <td>4949</td>\n",
       "      <td>0</td>\n",
       "      <td>CNhs11760</td>\n",
       "      <td>/home/drk/tillage/datasets/human/cage/fantom/C...</td>\n",
       "      <td>384</td>\n",
       "      <td>1</td>\n",
       "      <td>sum</td>\n",
       "      <td>aorta, adult, pool1</td>\n",
       "      <td>CAGE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      index  genome identifier   \n",
       "4949   4949       0  CNhs11760  \\\n",
       "\n",
       "                                                   file  clip  scale sum_stat   \n",
       "4949  /home/drk/tillage/datasets/human/cage/fantom/C...   384      1      sum  \\\n",
       "\n",
       "              description   exp  \n",
       "4949  aorta, adult, pool1  CAGE  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "targets[targets.identifier == \"CNhs11760\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9e271422-13b3-4152-b7b4-463b2913ad5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>genome</th>\n",
       "      <th>identifier</th>\n",
       "      <th>file</th>\n",
       "      <th>clip</th>\n",
       "      <th>scale</th>\n",
       "      <th>sum_stat</th>\n",
       "      <th>description</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>596</th>\n",
       "      <td>596</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF233ORX</td>\n",
       "      <td>/home/drk/tillage/datasets/human/dnase/encode/...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>ascending aorta female adult (51 year)</td>\n",
       "      <td>DNASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1494</th>\n",
       "      <td>1494</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF477GTH</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K27ac:thoracic aorta male adult (37 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1643</th>\n",
       "      <td>1643</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF383WLC</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K9me3:aorta male adult (34 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1659</th>\n",
       "      <td>1659</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF651IGX</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K27ac:ascending aorta female adult (53 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1702</th>\n",
       "      <td>1702</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF129RUF</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K36me3:thoracic aorta male adult (37 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1774</th>\n",
       "      <td>1774</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF573NUO</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K27me3:ascending aorta female adult (51 year)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1837</th>\n",
       "      <td>1837</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF151YZO</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K4me3:ascending aorta female adult (53 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1860</th>\n",
       "      <td>1860</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF182VPC</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K27me3:aorta female adult (30 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2063</th>\n",
       "      <td>2063</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF397BWR</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K27me3:aorta male adult (34 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2085</th>\n",
       "      <td>2085</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF082ADJ</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K4me1:ascending aorta female adult (51 year)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2227</th>\n",
       "      <td>2227</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF953LUQ</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K36me3:ascending aorta female adult (53 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2420</th>\n",
       "      <td>2420</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF721SMI</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K27ac:thoracic aorta male adult (54 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2433</th>\n",
       "      <td>2433</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF763TYO</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K27ac:aorta female adult (30 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2443</th>\n",
       "      <td>2443</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF261LVJ</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K4me1:aorta male adult (34 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2714</th>\n",
       "      <td>2714</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF652OON</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K4me1:thoracic aorta male adult (54 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2810</th>\n",
       "      <td>2810</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF771GOX</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K9me3:ascending aorta female adult (51 year)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3060</th>\n",
       "      <td>3060</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF039JTK</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K27ac:aorta male adult (34 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3269</th>\n",
       "      <td>3269</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF178UXM</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K27me3:ascending aorta female adult (53 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3440</th>\n",
       "      <td>3440</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF757WJJ</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K4me3:ascending aorta female adult (51 year)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3504</th>\n",
       "      <td>3504</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF075CNP</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>CTCF:thoracic aorta male adult (37 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3523</th>\n",
       "      <td>3523</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF968LVP</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K36me3:aorta male adult (34 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3632</th>\n",
       "      <td>3632</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF918NVX</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K4me1:ascending aorta female adult (53 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3683</th>\n",
       "      <td>3683</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF050VHG</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K9me3:ascending aorta female adult (53 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3727</th>\n",
       "      <td>3727</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF191EYK</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K36me3:ascending aorta female adult (51 year)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3942</th>\n",
       "      <td>3942</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF749KKG</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K4me1:thoracic aorta male adult (37 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4065</th>\n",
       "      <td>4065</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF296KKE</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K4me1:aorta female adult (30 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4262</th>\n",
       "      <td>4262</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF249NHD</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K9me3:thoracic aorta male adult (54 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4319</th>\n",
       "      <td>4319</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF713DRZ</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K4me3:thoracic aorta male adult (54 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4337</th>\n",
       "      <td>4337</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF008EYF</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K27me3:thoracic aorta male adult (37 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4384</th>\n",
       "      <td>4384</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF432XVD</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K4me3:aorta male adult (34 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4401</th>\n",
       "      <td>4401</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF328OPT</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K4me3:aorta female adult (30 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4403</th>\n",
       "      <td>4403</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF142LSJ</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>CTCF:ascending aorta female adult (51 year)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4438</th>\n",
       "      <td>4438</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF644PXV</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K9me3:thoracic aorta male adult (37 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4463</th>\n",
       "      <td>4463</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF109PKU</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K27ac:ascending aorta female adult (51 year)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4466</th>\n",
       "      <td>4466</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF715CEI</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K4me3:thoracic aorta male adult (37 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4478</th>\n",
       "      <td>4478</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF703NUB</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K36me3:thoracic aorta male adult (54 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4483</th>\n",
       "      <td>4483</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF281FXG</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K36me3:aorta female adult (30 years)</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4949</th>\n",
       "      <td>4949</td>\n",
       "      <td>0</td>\n",
       "      <td>CNhs11760</td>\n",
       "      <td>/home/drk/tillage/datasets/human/cage/fantom/C...</td>\n",
       "      <td>384</td>\n",
       "      <td>1</td>\n",
       "      <td>sum</td>\n",
       "      <td>aorta, adult, pool1</td>\n",
       "      <td>CAGE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      index  genome   identifier   \n",
       "596     596       0  ENCFF233ORX  \\\n",
       "1494   1494       0  ENCFF477GTH   \n",
       "1643   1643       0  ENCFF383WLC   \n",
       "1659   1659       0  ENCFF651IGX   \n",
       "1702   1702       0  ENCFF129RUF   \n",
       "1774   1774       0  ENCFF573NUO   \n",
       "1837   1837       0  ENCFF151YZO   \n",
       "1860   1860       0  ENCFF182VPC   \n",
       "2063   2063       0  ENCFF397BWR   \n",
       "2085   2085       0  ENCFF082ADJ   \n",
       "2227   2227       0  ENCFF953LUQ   \n",
       "2420   2420       0  ENCFF721SMI   \n",
       "2433   2433       0  ENCFF763TYO   \n",
       "2443   2443       0  ENCFF261LVJ   \n",
       "2714   2714       0  ENCFF652OON   \n",
       "2810   2810       0  ENCFF771GOX   \n",
       "3060   3060       0  ENCFF039JTK   \n",
       "3269   3269       0  ENCFF178UXM   \n",
       "3440   3440       0  ENCFF757WJJ   \n",
       "3504   3504       0  ENCFF075CNP   \n",
       "3523   3523       0  ENCFF968LVP   \n",
       "3632   3632       0  ENCFF918NVX   \n",
       "3683   3683       0  ENCFF050VHG   \n",
       "3727   3727       0  ENCFF191EYK   \n",
       "3942   3942       0  ENCFF749KKG   \n",
       "4065   4065       0  ENCFF296KKE   \n",
       "4262   4262       0  ENCFF249NHD   \n",
       "4319   4319       0  ENCFF713DRZ   \n",
       "4337   4337       0  ENCFF008EYF   \n",
       "4384   4384       0  ENCFF432XVD   \n",
       "4401   4401       0  ENCFF328OPT   \n",
       "4403   4403       0  ENCFF142LSJ   \n",
       "4438   4438       0  ENCFF644PXV   \n",
       "4463   4463       0  ENCFF109PKU   \n",
       "4466   4466       0  ENCFF715CEI   \n",
       "4478   4478       0  ENCFF703NUB   \n",
       "4483   4483       0  ENCFF281FXG   \n",
       "4949   4949       0    CNhs11760   \n",
       "\n",
       "                                                   file  clip  scale sum_stat   \n",
       "596   /home/drk/tillage/datasets/human/dnase/encode/...    32      2     mean  \\\n",
       "1494  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "1643  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "1659  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "1702  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "1774  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "1837  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "1860  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "2063  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "2085  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "2227  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "2420  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "2433  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "2443  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "2714  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "2810  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "3060  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "3269  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "3440  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "3504  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "3523  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "3632  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "3683  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "3727  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "3942  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "4065  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "4262  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "4319  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "4337  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "4384  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "4401  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "4403  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "4438  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "4463  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "4466  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "4478  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "4483  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "4949  /home/drk/tillage/datasets/human/cage/fantom/C...   384      1      sum   \n",
       "\n",
       "                                           description    exp  \n",
       "596             ascending aorta female adult (51 year)  DNASE  \n",
       "1494      H3K27ac:thoracic aorta male adult (37 years)   CHIP  \n",
       "1643               H3K9me3:aorta male adult (34 years)   CHIP  \n",
       "1659   H3K27ac:ascending aorta female adult (53 years)   CHIP  \n",
       "1702     H3K36me3:thoracic aorta male adult (37 years)   CHIP  \n",
       "1774   H3K27me3:ascending aorta female adult (51 year)   CHIP  \n",
       "1837   H3K4me3:ascending aorta female adult (53 years)   CHIP  \n",
       "1860            H3K27me3:aorta female adult (30 years)   CHIP  \n",
       "2063              H3K27me3:aorta male adult (34 years)   CHIP  \n",
       "2085    H3K4me1:ascending aorta female adult (51 year)   CHIP  \n",
       "2227  H3K36me3:ascending aorta female adult (53 years)   CHIP  \n",
       "2420      H3K27ac:thoracic aorta male adult (54 years)   CHIP  \n",
       "2433             H3K27ac:aorta female adult (30 years)   CHIP  \n",
       "2443               H3K4me1:aorta male adult (34 years)   CHIP  \n",
       "2714      H3K4me1:thoracic aorta male adult (54 years)   CHIP  \n",
       "2810    H3K9me3:ascending aorta female adult (51 year)   CHIP  \n",
       "3060               H3K27ac:aorta male adult (34 years)   CHIP  \n",
       "3269  H3K27me3:ascending aorta female adult (53 years)   CHIP  \n",
       "3440    H3K4me3:ascending aorta female adult (51 year)   CHIP  \n",
       "3504         CTCF:thoracic aorta male adult (37 years)   CHIP  \n",
       "3523              H3K36me3:aorta male adult (34 years)   CHIP  \n",
       "3632   H3K4me1:ascending aorta female adult (53 years)   CHIP  \n",
       "3683   H3K9me3:ascending aorta female adult (53 years)   CHIP  \n",
       "3727   H3K36me3:ascending aorta female adult (51 year)   CHIP  \n",
       "3942      H3K4me1:thoracic aorta male adult (37 years)   CHIP  \n",
       "4065             H3K4me1:aorta female adult (30 years)   CHIP  \n",
       "4262      H3K9me3:thoracic aorta male adult (54 years)   CHIP  \n",
       "4319      H3K4me3:thoracic aorta male adult (54 years)   CHIP  \n",
       "4337     H3K27me3:thoracic aorta male adult (37 years)   CHIP  \n",
       "4384               H3K4me3:aorta male adult (34 years)   CHIP  \n",
       "4401             H3K4me3:aorta female adult (30 years)   CHIP  \n",
       "4403       CTCF:ascending aorta female adult (51 year)   CHIP  \n",
       "4438      H3K9me3:thoracic aorta male adult (37 years)   CHIP  \n",
       "4463    H3K27ac:ascending aorta female adult (51 year)   CHIP  \n",
       "4466      H3K4me3:thoracic aorta male adult (37 years)   CHIP  \n",
       "4478     H3K36me3:thoracic aorta male adult (54 years)   CHIP  \n",
       "4483            H3K36me3:aorta female adult (30 years)   CHIP  \n",
       "4949                               aorta, adult, pool1   CAGE  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "targets[targets.description.str.contains(\"aorta\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b70b503b-c13f-4e07-aa34-52aa47379753",
   "metadata": {},
   "source": [
    "The `clip` parameter seems to have one of three values: 32, 64 and 384, depending on the experiment. For ChIP-seq experiments a clip value of 64 was used for all experiments starting the identfier with \"ENC\", so data derived from ENCODE. Conversely, the clip value was 32 for all experiments starting the identifier with \"GSM\". \n",
    "\n",
    "QUESTION: Why are they using different clip values for the two?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d8b29d71-b278-45a3-9569-21563a0782f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNASE: [32]\n",
      "ATAC: [64]\n",
      "CHIP: [32 64]\n",
      "CAGE: [384]\n"
     ]
    }
   ],
   "source": [
    "for exp in targets.exp.unique():\n",
    "    print(f\"{exp}: {targets[targets['exp'] == exp]['clip'].unique()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ec7f3f74-a3ff-44a4-b82f-4d85948bfe50",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([3832.,    0.,    0.,    0.,    0.,    0.,    0.,    0.,    0.,\n",
       "         159.]),\n",
       " array([32. , 35.2, 38.4, 41.6, 44.8, 48. , 51.2, 54.4, 57.6, 60.8, 64. ]),\n",
       " <BarContainer object of 10 artists>)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjEAAAGgCAYAAABbvTaPAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAstElEQVR4nO3df1BV953/8deNwK0aOCvgvZe7ocZu0NWg2R3swKXZ+Bt1QsivWW3p3tGp1WRjNIw6SbR/xO6kYrITbTrOWuNmYlSy5I/GNhkNkYyRrIso0jJRa1y70Q1WrlgXLmDpxZLP949+cyZXUIOi+IHnY+bMcM55n3M/5zOf5L783HPu9RhjjAAAACxzR383AAAA4HoQYgAAgJUIMQAAwEqEGAAAYCVCDAAAsBIhBgAAWIkQAwAArESIAQAAViLEAAAAKxFiAACAlW4oxJSWlsrj8aikpMTdZozRmjVrFAwGNXToUE2ZMkXHjh2LOy4Wi2np0qVKT0/X8OHDVVRUpDNnzsTVNDc3KxwOy3EcOY6jcDislpaWG2kuAAAYQBKu98Da2lq99tprmjhxYtz2l19+WevXr9fWrVs1ZswYvfjii5o5c6ZOnDih5ORkSVJJSYnee+89lZeXKy0tTStWrFBhYaHq6uo0ZMgQSVJxcbHOnDmjiooKSdLixYsVDof13nvvfa32ffHFFzp79qySk5Pl8Xiu9zIBAMAtZIxRW1ubgsGg7rjjGnMt5jq0tbWZrKwsU1lZaSZPnmyeeeYZY4wxX3zxhQkEAmbdunVu7Z/+9CfjOI75+c9/bowxpqWlxSQmJpry8nK35ve//7254447TEVFhTHGmN/+9rdGkqmpqXFrDhw4YCSZTz/99Gu1saGhwUhiYWFhYWFhsXBpaGi45nv9dc3ELFmyRA8++KBmzJihF1980d1+6tQpRSIRFRQUuNu8Xq8mT56s6upqPfHEE6qrq9OlS5fiaoLBoLKzs1VdXa1Zs2bpwIEDchxHubm5bk1eXp4cx1F1dbXGjh3brU2xWEyxWMxdN///x7kbGhqUkpJyPZcJAABusdbWVmVmZrqf3lxNr0NMeXm5fv3rX6u2trbbvkgkIkny+/1x2/1+v/73f//XrUlKStKIESO61Xx5fCQSkc/n63Z+n8/n1lyutLRUP/7xj7ttT0lJIcQAAGCZr3MrSK9u7G1oaNAzzzyjHTt26Bvf+MbXfmFjzDUbc3lNT/VXO8+qVasUjUbdpaGh4aqvBwAA7NarEFNXV6empibl5OQoISFBCQkJqqqq0s9+9jMlJCS4MzCXz5Y0NTW5+wKBgDo7O9Xc3HzVmnPnznV7/fPnz3eb5fmS1+t1Z12YfQEAYODrVYiZPn26jhw5ovr6eneZNGmSvv/976u+vl7f+ta3FAgEVFlZ6R7T2dmpqqoq5efnS5JycnKUmJgYV9PY2KijR4+6NaFQSNFoVIcOHXJrDh48qGg06tYAAIDBrVf3xCQnJys7Oztu2/Dhw5WWluZuLykp0dq1a5WVlaWsrCytXbtWw4YNU3FxsSTJcRwtXLhQK1asUFpamlJTU7Vy5UpNmDBBM2bMkCSNGzdOs2fP1qJFi7R582ZJf3nEurCwsMebegEAwOBz3d8TcyXPPvusOjo69NRTT6m5uVm5ubnas2dP3F3GGzZsUEJCgubOnauOjg5Nnz5dW7dudb8jRpLKysq0bNky9ymmoqIibdy4sa+bCwAALOUxXz6LPMC0trbKcRxFo1HujwEAwBK9ef/mt5MAAICVCDEAAMBKhBgAAGAlQgwAALASIQYAAFiJEAMAAKxEiAEAAFYixAAAACv1+Tf2DhZ3P7+rv5vQa6fXPdjfTQAAoM8wEwMAAKxEiAEAAFYixAAAACsRYgAAgJUIMQAAwEqEGAAAYCVCDAAAsBIhBgAAWIkQAwAArESIAQAAViLEAAAAKxFiAACAlQgxAADASoQYAABgJUIMAACwEiEGAABYiRADAACsRIgBAABWIsQAAAArEWIAAICVCDEAAMBKhBgAAGAlQgwAALASIQYAAFiJEAMAAKxEiAEAAFbqVYjZtGmTJk6cqJSUFKWkpCgUCun999939y9YsEAejyduycvLiztHLBbT0qVLlZ6eruHDh6uoqEhnzpyJq2lublY4HJbjOHIcR+FwWC0tLdd/lQAAYMDpVYi56667tG7dOh0+fFiHDx/WtGnT9PDDD+vYsWNuzezZs9XY2Oguu3fvjjtHSUmJdu7cqfLycu3fv1/t7e0qLCxUV1eXW1NcXKz6+npVVFSooqJC9fX1CofDN3ipAABgIEnoTfFDDz0Ut/6Tn/xEmzZtUk1Nje69915JktfrVSAQ6PH4aDSq119/Xdu3b9eMGTMkSTt27FBmZqY+/PBDzZo1S8ePH1dFRYVqamqUm5srSdqyZYtCoZBOnDihsWPH9voiAQDAwHPd98R0dXWpvLxcFy9eVCgUcrfv27dPPp9PY8aM0aJFi9TU1OTuq6ur06VLl1RQUOBuCwaDys7OVnV1tSTpwIEDchzHDTCSlJeXJ8dx3JqexGIxtba2xi0AAGDg6nWIOXLkiO688055vV49+eST2rlzp8aPHy9JmjNnjsrKyrR371698sorqq2t1bRp0xSLxSRJkUhESUlJGjFiRNw5/X6/IpGIW+Pz+bq9rs/nc2t6Ulpa6t5D4ziOMjMze3tpAADAIr36OEmSxo4dq/r6erW0tOgXv/iF5s+fr6qqKo0fP17z5s1z67KzszVp0iSNGjVKu3bt0mOPPXbFcxpj5PF43PWv/n2lmsutWrVKy5cvd9dbW1sJMgAADGC9DjFJSUm65557JEmTJk1SbW2tXn31VW3evLlbbUZGhkaNGqWTJ09KkgKBgDo7O9Xc3Bw3G9PU1KT8/Hy35ty5c93Odf78efn9/iu2y+v1yuv19vZyAACApW74e2KMMe7HRZe7cOGCGhoalJGRIUnKyclRYmKiKisr3ZrGxkYdPXrUDTGhUEjRaFSHDh1yaw4ePKhoNOrWAAAA9GomZvXq1ZozZ44yMzPV1tam8vJy7du3TxUVFWpvb9eaNWv0+OOPKyMjQ6dPn9bq1auVnp6uRx99VJLkOI4WLlyoFStWKC0tTampqVq5cqUmTJjgPq00btw4zZ49W4sWLXJndxYvXqzCwkKeTAIAAK5ehZhz584pHA6rsbFRjuNo4sSJqqio0MyZM9XR0aEjR45o27ZtamlpUUZGhqZOnaq3335bycnJ7jk2bNighIQEzZ07Vx0dHZo+fbq2bt2qIUOGuDVlZWVatmyZ+xRTUVGRNm7c2EeXDAAABgKPMcb0dyNuhtbWVjmOo2g0qpSUlD4//93P7+rzc95sp9c92N9NAADgqnrz/s1vJwEAACsRYgAAgJUIMQAAwEqEGAAAYCVCDAAAsBIhBgAAWIkQAwAArESIAQAAViLEAAAAKxFiAACAlQgxAADASoQYAABgJUIMAACwEiEGAABYiRADAACsRIgBAABWIsQAAAArEWIAAICVCDEAAMBKhBgAAGAlQgwAALASIQYAAFiJEAMAAKxEiAEAAFYixAAAACsRYgAAgJUIMQAAwEqEGAAAYCVCDAAAsBIhBgAAWIkQAwAArESIAQAAViLEAAAAKxFiAACAlQgxAADASr0KMZs2bdLEiROVkpKilJQUhUIhvf/+++5+Y4zWrFmjYDCooUOHasqUKTp27FjcOWKxmJYuXar09HQNHz5cRUVFOnPmTFxNc3OzwuGwHMeR4zgKh8NqaWm5/qsEAAADTq9CzF133aV169bp8OHDOnz4sKZNm6aHH37YDSovv/yy1q9fr40bN6q2tlaBQEAzZ85UW1ube46SkhLt3LlT5eXl2r9/v9rb21VYWKiuri63pri4WPX19aqoqFBFRYXq6+sVDof76JIBAMBA4DHGmBs5QWpqqv71X/9VP/jBDxQMBlVSUqLnnntO0l9mXfx+v1566SU98cQTikajGjlypLZv36558+ZJks6ePavMzEzt3r1bs2bN0vHjxzV+/HjV1NQoNzdXklRTU6NQKKRPP/1UY8eO/Vrtam1tleM4ikajSklJuZFL7NHdz+/q83PebKfXPdjfTQAA4Kp68/593ffEdHV1qby8XBcvXlQoFNKpU6cUiURUUFDg1ni9Xk2ePFnV1dWSpLq6Ol26dCmuJhgMKjs72605cOCAHMdxA4wk5eXlyXEct6YnsVhMra2tcQsAABi4eh1ijhw5ojvvvFNer1dPPvmkdu7cqfHjxysSiUiS/H5/XL3f73f3RSIRJSUlacSIEVet8fl83V7X5/O5NT0pLS1176FxHEeZmZm9vTQAAGCRXoeYsWPHqr6+XjU1Nfrnf/5nzZ8/X7/97W/d/R6PJ67eGNNt2+Uur+mp/lrnWbVqlaLRqLs0NDR83UsCAAAW6nWISUpK0j333KNJkyaptLRU9913n1599VUFAgFJ6jZb0tTU5M7OBAIBdXZ2qrm5+ao1586d6/a658+f7zbL81Ver9d9aurLBQAADFw3/D0xxhjFYjGNHj1agUBAlZWV7r7Ozk5VVVUpPz9fkpSTk6PExMS4msbGRh09etStCYVCikajOnTokFtz8OBBRaNRtwYAACChN8WrV6/WnDlzlJmZqba2NpWXl2vfvn2qqKiQx+NRSUmJ1q5dq6ysLGVlZWnt2rUaNmyYiouLJUmO42jhwoVasWKF0tLSlJqaqpUrV2rChAmaMWOGJGncuHGaPXu2Fi1apM2bN0uSFi9erMLCwq/9ZBIAABj4ehVizp07p3A4rMbGRjmOo4kTJ6qiokIzZ86UJD377LPq6OjQU089pebmZuXm5mrPnj1KTk52z7FhwwYlJCRo7ty56ujo0PTp07V161YNGTLErSkrK9OyZcvcp5iKioq0cePGvrheAAAwQNzw98TcrviemO74nhgAwO3ulnxPDAAAQH8ixAAAACsRYgAAgJUIMQAAwEqEGAAAYCVCDAAAsBIhBgAAWIkQAwAArESIAQAAViLEAAAAKxFiAACAlQgxAADASoQYAABgJUIMAACwEiEGAABYiRADAACsRIgBAABWIsQAAAArEWIAAICVCDEAAMBKhBgAAGAlQgwAALASIQYAAFiJEAMAAKxEiAEAAFYixAAAACsRYgAAgJUIMQAAwEqEGAAAYCVCDAAAsBIhBgAAWIkQAwAArESIAQAAViLEAAAAKxFiAACAlXoVYkpLS/Xtb39bycnJ8vl8euSRR3TixIm4mgULFsjj8cQteXl5cTWxWExLly5Venq6hg8frqKiIp05cyauprm5WeFwWI7jyHEchcNhtbS0XN9VAgCAAadXIaaqqkpLlixRTU2NKisr9ec//1kFBQW6ePFiXN3s2bPV2NjoLrt3747bX1JSop07d6q8vFz79+9Xe3u7CgsL1dXV5dYUFxervr5eFRUVqqioUH19vcLh8A1cKgAAGEgSelNcUVERt/7GG2/I5/Oprq5ODzzwgLvd6/UqEAj0eI5oNKrXX39d27dv14wZMyRJO3bsUGZmpj788EPNmjVLx48fV0VFhWpqapSbmytJ2rJli0KhkE6cOKGxY8f26iIBAMDAc0P3xESjUUlSampq3PZ9+/bJ5/NpzJgxWrRokZqamtx9dXV1unTpkgoKCtxtwWBQ2dnZqq6uliQdOHBAjuO4AUaS8vLy5DiOW3O5WCym1tbWuAUAAAxc1x1ijDFavny57r//fmVnZ7vb58yZo7KyMu3du1evvPKKamtrNW3aNMViMUlSJBJRUlKSRowYEXc+v9+vSCTi1vh8vm6v6fP53JrLlZaWuvfPOI6jzMzM6700AABggV59nPRVTz/9tD755BPt378/bvu8efPcv7OzszVp0iSNGjVKu3bt0mOPPXbF8xlj5PF43PWv/n2lmq9atWqVli9f7q63trYSZAAAGMCuayZm6dKlevfdd/XRRx/prrvuumptRkaGRo0apZMnT0qSAoGAOjs71dzcHFfX1NQkv9/v1pw7d67buc6fP+/WXM7r9SolJSVuAQAAA1evQowxRk8//bTeeecd7d27V6NHj77mMRcuXFBDQ4MyMjIkSTk5OUpMTFRlZaVb09jYqKNHjyo/P1+SFAqFFI1GdejQIbfm4MGDikajbg0AABjcevVx0pIlS/TWW2/pV7/6lZKTk937UxzH0dChQ9Xe3q41a9bo8ccfV0ZGhk6fPq3Vq1crPT1djz76qFu7cOFCrVixQmlpaUpNTdXKlSs1YcIE92mlcePGafbs2Vq0aJE2b94sSVq8eLEKCwt5MgkAAEjqZYjZtGmTJGnKlClx29944w0tWLBAQ4YM0ZEjR7Rt2za1tLQoIyNDU6dO1dtvv63k5GS3fsOGDUpISNDcuXPV0dGh6dOna+vWrRoyZIhbU1ZWpmXLlrlPMRUVFWnjxo3Xe50AAGCA8RhjTH834mZobW2V4ziKRqM35f6Yu5/f1efnvNlOr3uwv5sAAMBV9eb9m99OAgAAViLEAAAAKxFiAACAlQgxAADASoQYAABgJUIMAACwEiEGAABYiRADAACsRIgBAABWIsQAAAArEWIAAICVCDEAAMBKhBgAAGAlQgwAALASIQYAAFiJEAMAAKxEiAEAAFYixAAAACsRYgAAgJUIMQAAwEqEGAAAYCVCDAAAsBIhBgAAWIkQAwAArESIAQAAViLEAAAAKxFiAACAlQgxAADASoQYAABgJUIMAACwEiEGAABYiRADAACsRIgBAABWIsQAAAArEWIAAICVehViSktL9e1vf1vJycny+Xx65JFHdOLEibgaY4zWrFmjYDCooUOHasqUKTp27FhcTSwW09KlS5Wenq7hw4erqKhIZ86ciatpbm5WOByW4zhyHEfhcFgtLS3Xd5UAAGDA6VWIqaqq0pIlS1RTU6PKykr9+c9/VkFBgS5evOjWvPzyy1q/fr02btyo2tpaBQIBzZw5U21tbW5NSUmJdu7cqfLycu3fv1/t7e0qLCxUV1eXW1NcXKz6+npVVFSooqJC9fX1CofDfXDJAABgIPAYY8z1Hnz+/Hn5fD5VVVXpgQcekDFGwWBQJSUleu655yT9ZdbF7/frpZde0hNPPKFoNKqRI0dq+/btmjdvniTp7NmzyszM1O7duzVr1iwdP35c48ePV01NjXJzcyVJNTU1CoVC+vTTTzV27Nhrtq21tVWO4ygajSolJeV6L/GK7n5+V5+f82Y7ve7B/m4CAABX1Zv37xu6JyYajUqSUlNTJUmnTp1SJBJRQUGBW+P1ejV58mRVV1dLkurq6nTp0qW4mmAwqOzsbLfmwIEDchzHDTCSlJeXJ8dx3JrLxWIxtba2xi0AAGDguu4QY4zR8uXLdf/99ys7O1uSFIlEJEl+vz+u1u/3u/sikYiSkpI0YsSIq9b4fL5ur+nz+dyay5WWlrr3zziOo8zMzOu9NAAAYIHrDjFPP/20PvnkE/3Hf/xHt30ejydu3RjTbdvlLq/pqf5q51m1apWi0ai7NDQ0fJ3LAAAAlrquELN06VK9++67+uijj3TXXXe52wOBgCR1my1pampyZ2cCgYA6OzvV3Nx81Zpz5851e93z5893m+X5ktfrVUpKStwCAAAGrl6FGGOMnn76ab3zzjvau3evRo8eHbd/9OjRCgQCqqysdLd1dnaqqqpK+fn5kqScnBwlJibG1TQ2Nuro0aNuTSgUUjQa1aFDh9yagwcPKhqNujUAAGBwS+hN8ZIlS/TWW2/pV7/6lZKTk90ZF8dxNHToUHk8HpWUlGjt2rXKyspSVlaW1q5dq2HDhqm4uNitXbhwoVasWKG0tDSlpqZq5cqVmjBhgmbMmCFJGjdunGbPnq1FixZp8+bNkqTFixersLDwaz2ZBAAABr5ehZhNmzZJkqZMmRK3/Y033tCCBQskSc8++6w6Ojr01FNPqbm5Wbm5udqzZ4+Sk5Pd+g0bNighIUFz585VR0eHpk+frq1bt2rIkCFuTVlZmZYtW+Y+xVRUVKSNGzdezzUCAIAB6Ia+J+Z2xvfEdMf3xAAAbne37HtiAAAA+gshBgAAWIkQAwAArESIAQAAViLEAAAAKxFiAACAlQgxAADASoQYAABgJUIMAACwEiEGAABYiRADAACsRIgBAABWIsQAAAArEWIAAICVCDEAAMBKhBgAAGAlQgwAALASIQYAAFiJEAMAAKxEiAEAAFYixAAAACsRYgAAgJUIMQAAwEqEGAAAYCVCDAAAsBIhBgAAWIkQAwAArESIAQAAViLEAAAAKxFiAACAlQgxAADASoQYAABgJUIMAACwEiEGAABYiRADAACs1OsQ8/HHH+uhhx5SMBiUx+PRL3/5y7j9CxYskMfjiVvy8vLiamKxmJYuXar09HQNHz5cRUVFOnPmTFxNc3OzwuGwHMeR4zgKh8NqaWnp9QUCAICBqdch5uLFi7rvvvu0cePGK9bMnj1bjY2N7rJ79+64/SUlJdq5c6fKy8u1f/9+tbe3q7CwUF1dXW5NcXGx6uvrVVFRoYqKCtXX1yscDve2uQAAYIBK6O0Bc+bM0Zw5c65a4/V6FQgEetwXjUb1+uuva/v27ZoxY4YkaceOHcrMzNSHH36oWbNm6fjx46qoqFBNTY1yc3MlSVu2bFEoFNKJEyc0duzYbueNxWKKxWLuemtra28vDQAAWOSm3BOzb98++Xw+jRkzRosWLVJTU5O7r66uTpcuXVJBQYG7LRgMKjs7W9XV1ZKkAwcOyHEcN8BIUl5enhzHcWsuV1pa6n705DiOMjMzb8alAQCA20Sfh5g5c+aorKxMe/fu1SuvvKLa2lpNmzbNnSWJRCJKSkrSiBEj4o7z+/2KRCJujc/n63Zun8/n1lxu1apVikaj7tLQ0NDHVwYAAG4nvf446VrmzZvn/p2dna1JkyZp1KhR2rVrlx577LErHmeMkcfjcde/+veVar7K6/XK6/XeQMsBAIBNbvoj1hkZGRo1apROnjwpSQoEAurs7FRzc3NcXVNTk/x+v1tz7ty5buc6f/68WwMAAAa3mx5iLly4oIaGBmVkZEiScnJylJiYqMrKSremsbFRR48eVX5+viQpFAopGo3q0KFDbs3BgwcVjUbdGgAAMLj1+uOk9vZ2/e53v3PXT506pfr6eqWmpio1NVVr1qzR448/royMDJ0+fVqrV69Wenq6Hn30UUmS4zhauHChVqxYobS0NKWmpmrlypWaMGGC+7TSuHHjNHv2bC1atEibN2+WJC1evFiFhYU9PpkEAAAGn16HmMOHD2vq1Knu+vLlyyVJ8+fP16ZNm3TkyBFt27ZNLS0tysjI0NSpU/X2228rOTnZPWbDhg1KSEjQ3Llz1dHRoenTp2vr1q0aMmSIW1NWVqZly5a5TzEVFRVd9btpAADA4OIxxpj+bsTN0NraKsdxFI1GlZKS0ufnv/v5XX1+zpvt9LoH+7sJAABcVW/ev/ntJAAAYCVCDAAAsBIhBgAAWIkQAwAArESIAQAAViLEAAAAKxFiAACAlQgxAADASoQYAABgJUIMAACwEiEGAABYiRADAACsRIgBAABWIsQAAAArEWIAAICVCDEAAMBKhBgAAGAlQgwAALASIQYAAFiJEAMAAKxEiAEAAFYixAAAACsRYgAAgJUIMQAAwEqEGAAAYCVCDAAAsBIhBgAAWIkQAwAArESIAQAAViLEAAAAKxFiAACAlQgxAADASoQYAABgJUIMAACwEiEGAABYqdch5uOPP9ZDDz2kYDAoj8ejX/7yl3H7jTFas2aNgsGghg4dqilTpujYsWNxNbFYTEuXLlV6erqGDx+uoqIinTlzJq6mublZ4XBYjuPIcRyFw2G1tLT0+gIBAMDA1OsQc/HiRd13333auHFjj/tffvllrV+/Xhs3blRtba0CgYBmzpyptrY2t6akpEQ7d+5UeXm59u/fr/b2dhUWFqqrq8utKS4uVn19vSoqKlRRUaH6+nqFw+HruEQAADAQeYwx5roP9ni0c+dOPfLII5L+MgsTDAZVUlKi5557TtJfZl38fr9eeuklPfHEE4pGoxo5cqS2b9+uefPmSZLOnj2rzMxM7d69W7NmzdLx48c1fvx41dTUKDc3V5JUU1OjUCikTz/9VGPHju3Wllgsplgs5q63trYqMzNT0WhUKSkp13uJV3T387v6/Jw32+l1D/Z3EwAAuKrW1lY5jvO13r/79J6YU6dOKRKJqKCgwN3m9Xo1efJkVVdXS5Lq6up06dKluJpgMKjs7Gy35sCBA3Icxw0wkpSXlyfHcdyay5WWlrofPTmOo8zMzL68NAAAcJvp0xATiUQkSX6/P2673+9390UiESUlJWnEiBFXrfH5fN3O7/P53JrLrVq1StFo1F0aGhpu+HoAAMDtK+FmnNTj8cStG2O6bbvc5TU91V/tPF6vV16v9zpaCwAAbNSnMzGBQECSus2WNDU1ubMzgUBAnZ2dam5uvmrNuXPnup3//Pnz3WZ5AADA4NSnIWb06NEKBAKqrKx0t3V2dqqqqkr5+fmSpJycHCUmJsbVNDY26ujRo25NKBRSNBrVoUOH3JqDBw8qGo26NQAAYHDr9cdJ7e3t+t3vfueunzp1SvX19UpNTdU3v/lNlZSUaO3atcrKylJWVpbWrl2rYcOGqbi4WJLkOI4WLlyoFStWKC0tTampqVq5cqUmTJigGTNmSJLGjRun2bNna9GiRdq8ebMkafHixSosLOzxySQAADD49DrEHD58WFOnTnXXly9fLkmaP3++tm7dqmeffVYdHR166qmn1NzcrNzcXO3Zs0fJycnuMRs2bFBCQoLmzp2rjo4OTZ8+XVu3btWQIUPcmrKyMi1btsx9iqmoqOiK300DAAAGnxv6npjbWW+eM78efE8MAAB9r9++JwYAAOBWIcQAAAArEWIAAICVCDEAAMBKhBgAAGAlQgwAALASIQYAAFiJEAMAAKxEiAEAAFYixAAAACsRYgAAgJUIMQAAwEqEGAAAYCVCDAAAsBIhBgAAWIkQAwAArESIAQAAViLEAAAAKxFiAACAlQgxAADASoQYAABgJUIMAACwEiEGAABYiRADAACsRIgBAABWIsQAAAArEWIAAICVCDEAAMBKhBgAAGAlQgwAALASIQYAAFiJEAMAAKxEiAEAAFYixAAAACsRYgAAgJX6PMSsWbNGHo8nbgkEAu5+Y4zWrFmjYDCooUOHasqUKTp27FjcOWKxmJYuXar09HQNHz5cRUVFOnPmTF83FQAAWOymzMTce++9amxsdJcjR464+15++WWtX79eGzduVG1trQKBgGbOnKm2tja3pqSkRDt37lR5ebn279+v9vZ2FRYWqqur62Y0FwAAWCjhppw0ISFu9uVLxhj99Kc/1Y9+9CM99thjkqQ333xTfr9fb731lp544glFo1G9/vrr2r59u2bMmCFJ2rFjhzIzM/Xhhx9q1qxZPb5mLBZTLBZz11tbW2/ClQEAgNvFTZmJOXnypILBoEaPHq3vfve7+uyzzyRJp06dUiQSUUFBgVvr9Xo1efJkVVdXS5Lq6up06dKluJpgMKjs7Gy3pielpaVyHMddMjMzb8alAQCA20Sfh5jc3Fxt27ZNH3zwgbZs2aJIJKL8/HxduHBBkUhEkuT3++OO8fv97r5IJKKkpCSNGDHiijU9WbVqlaLRqLs0NDT08ZUBAIDbSZ9/nDRnzhz37wkTJigUCulv/uZv9OabbyovL0+S5PF44o4xxnTbdrlr1Xi9Xnm93htoOQAAsMlNf8R6+PDhmjBhgk6ePOneJ3P5jEpTU5M7OxMIBNTZ2anm5uYr1gAAANz0EBOLxXT8+HFlZGRo9OjRCgQCqqysdPd3dnaqqqpK+fn5kqScnBwlJibG1TQ2Nuro0aNuDQAAQJ9/nLRy5Uo99NBD+uY3v6mmpia9+OKLam1t1fz58+XxeFRSUqK1a9cqKytLWVlZWrt2rYYNG6bi4mJJkuM4WrhwoVasWKG0tDSlpqZq5cqVmjBhgvu0EgAAQJ+HmDNnzuh73/ue/vCHP2jkyJHKy8tTTU2NRo0aJUl69tln1dHRoaeeekrNzc3Kzc3Vnj17lJyc7J5jw4YNSkhI0Ny5c9XR0aHp06dr69atGjJkSF83FwAAWMpjjDH93YibobW1VY7jKBqNKiUlpc/Pf/fzu/r8nDfb6XUP9ncTAAC4qt68f/PbSQAAwEqEGAAAYCVCDAAAsBIhBgAAWIkQAwAArESIAQAAViLEAAAAKxFiAACAlQgxAADASoQYAABgJUIMAACwEiEGAABYiRADAACsRIgBAABWIsQAAAArEWIAAICVCDEAAMBKhBgAAGAlQgwAALASIQYAAFiJEAMAAKxEiAEAAFYixAAAACsRYgAAgJUS+rsBAABAuvv5Xf3dhF47ve7Bfn19ZmIAAICVCDEAAMBKhBgAAGAlQgwAALASIQYAAFiJEAMAAKxEiAEAAFYixAAAACsRYgAAgJVu+xDzb//2bxo9erS+8Y1vKCcnR//5n//Z300CAAC3gds6xLz99tsqKSnRj370I/3mN7/RP/zDP2jOnDn6/PPP+7tpAACgn93WIWb9+vVauHChfvjDH2rcuHH66U9/qszMTG3atKm/mwYAAPrZbfsDkJ2dnaqrq9Pzzz8ft72goEDV1dXd6mOxmGKxmLsejUYlSa2trTelfV/E/nhTznsz3ay+AADcON5X4s9pjLlm7W0bYv7whz+oq6tLfr8/brvf71ckEulWX1paqh//+MfdtmdmZt60NtrG+Wl/twAAMJDczPeVtrY2OY5z1ZrbNsR8yePxxK0bY7ptk6RVq1Zp+fLl7voXX3yh//u//1NaWlqP9YNBa2urMjMz1dDQoJSUlP5uzm2Dfrky+qZn9MuV0Tc9o1+u7Fp9Y4xRW1ubgsHgNc9124aY9PR0DRkypNusS1NTU7fZGUnyer3yer1x2/7qr/7qZjbRGikpKfxH1AP65crom57RL1dG3/SMfrmyq/XNtWZgvnTb3tiblJSknJwcVVZWxm2vrKxUfn5+P7UKAADcLm7bmRhJWr58ucLhsCZNmqRQKKTXXntNn3/+uZ588sn+bhoAAOhnt3WImTdvni5cuKB/+Zd/UWNjo7Kzs7V7926NGjWqv5tmBa/XqxdeeKHbx2yDHf1yZfRNz+iXK6Nveka/XFlf9o3HfJ1nmAAAAG4zt+09MQAAAFdDiAEAAFYixAAAACsRYgAAgJUIMQAAwEqEGMtt2rRJEydOdL/5MBQK6f3333f3L1iwQB6PJ27Jy8vrxxb3j9LSUnk8HpWUlLjbjDFas2aNgsGghg4dqilTpujYsWP918h+0lPfDNZxs2bNmm7XHQgE3P2Ddcxcq18G63iRpN///vf6p3/6J6WlpWnYsGH6u7/7O9XV1bn7B+uYka7dN30xbm7r74nBtd11111at26d7rnnHknSm2++qYcffli/+c1vdO+990qSZs+erTfeeMM9JikpqV/a2l9qa2v12muvaeLEiXHbX375Za1fv15bt27VmDFj9OKLL2rmzJk6ceKEkpOT+6m1t9aV+kYavOPm3nvv1YcffuiuDxkyxP17MI+Zq/WLNDjHS3Nzs77zne9o6tSpev/99+Xz+fQ///M/cT95M1jHzNfpG6kPxo3BgDNixAjz7//+78YYY+bPn28efvjh/m1QP2prazNZWVmmsrLSTJ482TzzzDPGGGO++OILEwgEzLp169zaP/3pT8ZxHPPzn/+8n1p7a12pb4wZvOPmhRdeMPfdd1+P+wbzmLlavxgzeMfLc889Z+6///4r7h/MY+ZafWNM34wbPk4aQLq6ulReXq6LFy8qFAq52/ft2yefz6cxY8Zo0aJFampq6sdW3lpLlizRgw8+qBkzZsRtP3XqlCKRiAoKCtxtXq9XkydPVnV19a1uZr+4Ut98abCOm5MnTyoYDGr06NH67ne/q88++0wSY+ZK/fKlwThe3n33XU2aNEn/+I//KJ/Pp7//+7/Xli1b3P2Decxcq2++dKPjhhAzABw5ckR33nmnvF6vnnzySe3cuVPjx4+XJM2ZM0dlZWXau3evXnnlFdXW1mratGmKxWL93Oqbr7y8XL/+9a9VWlrabd+Xv45++S+i+/3+br+cPhBdrW+kwTtucnNztW3bNn3wwQfasmWLIpGI8vPzdeHChUE9Zq7WL9LgHS+fffaZNm3apKysLH3wwQd68skntWzZMm3btk3S4P7/zLX6RuqjcXND8zi4LcRiMXPy5ElTW1trnn/+eZOenm6OHTvWY+3Zs2dNYmKi+cUvfnGLW3lrff7558bn85n6+np321c/Mvmv//ovI8mcPXs27rgf/vCHZtasWbeyqbfctfqmJ4Nl3Fyuvb3d+P1+88orrwzqMXO5r/ZLTwbLeElMTDShUChu29KlS01eXp4xZnD/f+ZafdOT6xk3zMQMAElJSbrnnns0adIklZaW6r777tOrr77aY21GRoZGjRqlkydP3uJW3lp1dXVqampSTk6OEhISlJCQoKqqKv3sZz9TQkKC+y+jy/811NTU1O1fTQPNtfqmq6ur2zGDZdxcbvjw4ZowYYJOnjzpPo0zGMfM5b7aLz0ZLOMlIyPDnfX+0rhx4/T5559L0qAeM9fqmysd09txQ4gZgIwxV5yOu3DhghoaGpSRkXGLW3VrTZ8+XUeOHFF9fb27TJo0Sd///vdVX1+vb33rWwoEAqqsrHSP6ezsVFVVlfLz8/ux5Tfftfrm8qdOpMEzbi4Xi8V0/PhxZWRkaPTo0YN2zFzuq/3Sk8EyXr7zne/oxIkTcdv++7//W6NGjZKkQT1mrtU3PbmucXO9U0W4Paxatcp8/PHH5tSpU+aTTz4xq1evNnfccYfZs2ePaWtrMytWrDDV1dXm1KlT5qOPPjKhUMj89V//tWltbe3vpt9yl39ksm7dOuM4jnnnnXfMkSNHzPe+9z2TkZEx6PtmMI+bFStWmH379pnPPvvM1NTUmMLCQpOcnGxOnz5tjBm8Y+Zq/TKYx8uhQ4dMQkKC+clPfmJOnjxpysrKzLBhw8yOHTvcmsE6Zq7VN301bggxlvvBD35gRo0aZZKSkszIkSPN9OnTzZ49e4wxxvzxj380BQUFZuTIkSYxMdF885vfNPPnzzeff/55P7e6f1weYr744gvzwgsvmEAgYLxer3nggQfMkSNH+q+B/eirfTOYx828efNMRkaGSUxMNMFg0Dz22GNx95cN1jFztX4ZzOPFGGPee+89k52dbbxer/nbv/1b89prr8XtH6xjxpir901fjRuPMcZc73QRAABAf+GeGAAAYCVCDAAAsBIhBgAAWIkQAwAArESIAQAAViLEAAAAKxFiAACAlQgxAADASoQYAABgJUIMAACwEiEGAABY6f8B3b+nItULRLEAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(targets[targets[\"exp\"] == \"CHIP\"][\"clip\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e9e95729-51a9-4d56-bfe3-558802b52320",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All CHIP-seq experiments with clip = 64 have an identifier that starts with GSM: True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>genome</th>\n",
       "      <th>identifier</th>\n",
       "      <th>file</th>\n",
       "      <th>clip</th>\n",
       "      <th>scale</th>\n",
       "      <th>sum_stat</th>\n",
       "      <th>description</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4516</th>\n",
       "      <td>4516</td>\n",
       "      <td>0</td>\n",
       "      <td>GSM1208590</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/geo/GSM1...</td>\n",
       "      <td>64</td>\n",
       "      <td>1</td>\n",
       "      <td>sum</td>\n",
       "      <td>.:batch1_chrom1_LoVo_ARNT_Mouse_PassedQC / LoV...</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4517</th>\n",
       "      <td>4517</td>\n",
       "      <td>0</td>\n",
       "      <td>GSM1208591</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/geo/GSM1...</td>\n",
       "      <td>64</td>\n",
       "      <td>1</td>\n",
       "      <td>sum</td>\n",
       "      <td>.:batch1_chrom1_LoVo_ASCL2_Mouse_PassedQC / Lo...</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4518</th>\n",
       "      <td>4518</td>\n",
       "      <td>0</td>\n",
       "      <td>GSM1208592</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/geo/GSM1...</td>\n",
       "      <td>64</td>\n",
       "      <td>1</td>\n",
       "      <td>sum</td>\n",
       "      <td>.:batch1_chrom1_LoVo_ATOH1_Rabbit_PassedQC / L...</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4519</th>\n",
       "      <td>4519</td>\n",
       "      <td>0</td>\n",
       "      <td>GSM1208593</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/geo/GSM1...</td>\n",
       "      <td>64</td>\n",
       "      <td>1</td>\n",
       "      <td>sum</td>\n",
       "      <td>.:batch1_chrom1_LoVo_BARX2_Goat_PassedQC / LoV...</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4520</th>\n",
       "      <td>4520</td>\n",
       "      <td>0</td>\n",
       "      <td>GSM1208594</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/geo/GSM1...</td>\n",
       "      <td>64</td>\n",
       "      <td>1</td>\n",
       "      <td>sum</td>\n",
       "      <td>.:batch1_chrom1_LoVo_CAMTA2_Goat_PassedQC / Lo...</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      index  genome  identifier  \\\n",
       "4516   4516       0  GSM1208590   \n",
       "4517   4517       0  GSM1208591   \n",
       "4518   4518       0  GSM1208592   \n",
       "4519   4519       0  GSM1208593   \n",
       "4520   4520       0  GSM1208594   \n",
       "\n",
       "                                                   file  clip  scale sum_stat  \\\n",
       "4516  /home/drk/tillage/datasets/human/chip/geo/GSM1...    64      1      sum   \n",
       "4517  /home/drk/tillage/datasets/human/chip/geo/GSM1...    64      1      sum   \n",
       "4518  /home/drk/tillage/datasets/human/chip/geo/GSM1...    64      1      sum   \n",
       "4519  /home/drk/tillage/datasets/human/chip/geo/GSM1...    64      1      sum   \n",
       "4520  /home/drk/tillage/datasets/human/chip/geo/GSM1...    64      1      sum   \n",
       "\n",
       "                                            description   exp  \n",
       "4516  .:batch1_chrom1_LoVo_ARNT_Mouse_PassedQC / LoV...  CHIP  \n",
       "4517  .:batch1_chrom1_LoVo_ASCL2_Mouse_PassedQC / Lo...  CHIP  \n",
       "4518  .:batch1_chrom1_LoVo_ATOH1_Rabbit_PassedQC / L...  CHIP  \n",
       "4519  .:batch1_chrom1_LoVo_BARX2_Goat_PassedQC / LoV...  CHIP  \n",
       "4520  .:batch1_chrom1_LoVo_CAMTA2_Goat_PassedQC / Lo...  CHIP  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chip_64 = targets[(targets[\"exp\"] == \"CHIP\") & (targets[\"clip\"] == 64)]\n",
    "print(f\"All CHIP-seq experiments with clip = 64 have an identifier that starts with GSM: {chip_64['identifier'].str.startswith('GSM').sum() == len(chip_64)}\")\n",
    "chip_64.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e8e0e8e7-0b65-4ada-b73d-0a7745c08806",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All CHIP-seq experiments with clip = 64 have an identifier that starts with ENC: True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>genome</th>\n",
       "      <th>identifier</th>\n",
       "      <th>file</th>\n",
       "      <th>clip</th>\n",
       "      <th>scale</th>\n",
       "      <th>sum_stat</th>\n",
       "      <th>description</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>684</th>\n",
       "      <td>684</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF991NDB</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>CTCF:MCF-7</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>685</th>\n",
       "      <td>685</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF831FHP</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>TAF1:MCF-7</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>686</th>\n",
       "      <td>686</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF776DPQ</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K4me3:GM12878</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>687</th>\n",
       "      <td>687</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF279CYY</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>CTCF:GM12878</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688</th>\n",
       "      <td>688</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF340JIF</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K27ac:GM12878</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index  genome   identifier  \\\n",
       "684    684       0  ENCFF991NDB   \n",
       "685    685       0  ENCFF831FHP   \n",
       "686    686       0  ENCFF776DPQ   \n",
       "687    687       0  ENCFF279CYY   \n",
       "688    688       0  ENCFF340JIF   \n",
       "\n",
       "                                                  file  clip  scale sum_stat  \\\n",
       "684  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "685  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "686  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "687  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "688  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "\n",
       "         description   exp  \n",
       "684       CTCF:MCF-7  CHIP  \n",
       "685       TAF1:MCF-7  CHIP  \n",
       "686  H3K4me3:GM12878  CHIP  \n",
       "687     CTCF:GM12878  CHIP  \n",
       "688  H3K27ac:GM12878  CHIP  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chip_32 = targets[(targets[\"exp\"] == \"CHIP\") & (targets[\"clip\"] == 32)]\n",
    "print(f\"All CHIP-seq experiments with clip = 64 have an identifier that starts with ENC: {chip_32['identifier'].str.startswith('ENC').sum() == len(chip_32)}\")\n",
    "\n",
    "chip_32.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56fe5744-fa45-4f8a-a20f-684e954f7b5d",
   "metadata": {},
   "source": [
    "The `sum_stat` parameter is either \n",
    "* mean\n",
    "    - DNASE\n",
    "    - CHIP from ENCODE\n",
    "* sum \n",
    "    - ATAC\n",
    "    - CHIP from GSM\n",
    "    - CAGE "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a27c0f60-ca5c-44e7-b73a-d4e6324d3b61",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNASE: ['mean']\n",
      "ATAC: ['sum']\n",
      "CHIP: ['mean' 'sum']\n",
      "CAGE: ['sum']\n"
     ]
    }
   ],
   "source": [
    "for exp in targets.exp.unique():\n",
    "    print(f\"{exp}: {targets[targets['exp'] == exp]['sum_stat'].unique()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1de085b2-b4c4-4529-9f8f-7e765fc33eae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All CHIP-seq experiments with sum_stat = sum have an identifier that starts with GSM: True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>genome</th>\n",
       "      <th>identifier</th>\n",
       "      <th>file</th>\n",
       "      <th>clip</th>\n",
       "      <th>scale</th>\n",
       "      <th>sum_stat</th>\n",
       "      <th>description</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4516</th>\n",
       "      <td>4516</td>\n",
       "      <td>0</td>\n",
       "      <td>GSM1208590</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/geo/GSM1...</td>\n",
       "      <td>64</td>\n",
       "      <td>1</td>\n",
       "      <td>sum</td>\n",
       "      <td>.:batch1_chrom1_LoVo_ARNT_Mouse_PassedQC / LoV...</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4517</th>\n",
       "      <td>4517</td>\n",
       "      <td>0</td>\n",
       "      <td>GSM1208591</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/geo/GSM1...</td>\n",
       "      <td>64</td>\n",
       "      <td>1</td>\n",
       "      <td>sum</td>\n",
       "      <td>.:batch1_chrom1_LoVo_ASCL2_Mouse_PassedQC / Lo...</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4518</th>\n",
       "      <td>4518</td>\n",
       "      <td>0</td>\n",
       "      <td>GSM1208592</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/geo/GSM1...</td>\n",
       "      <td>64</td>\n",
       "      <td>1</td>\n",
       "      <td>sum</td>\n",
       "      <td>.:batch1_chrom1_LoVo_ATOH1_Rabbit_PassedQC / L...</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4519</th>\n",
       "      <td>4519</td>\n",
       "      <td>0</td>\n",
       "      <td>GSM1208593</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/geo/GSM1...</td>\n",
       "      <td>64</td>\n",
       "      <td>1</td>\n",
       "      <td>sum</td>\n",
       "      <td>.:batch1_chrom1_LoVo_BARX2_Goat_PassedQC / LoV...</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4520</th>\n",
       "      <td>4520</td>\n",
       "      <td>0</td>\n",
       "      <td>GSM1208594</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/geo/GSM1...</td>\n",
       "      <td>64</td>\n",
       "      <td>1</td>\n",
       "      <td>sum</td>\n",
       "      <td>.:batch1_chrom1_LoVo_CAMTA2_Goat_PassedQC / Lo...</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      index  genome  identifier  \\\n",
       "4516   4516       0  GSM1208590   \n",
       "4517   4517       0  GSM1208591   \n",
       "4518   4518       0  GSM1208592   \n",
       "4519   4519       0  GSM1208593   \n",
       "4520   4520       0  GSM1208594   \n",
       "\n",
       "                                                   file  clip  scale sum_stat  \\\n",
       "4516  /home/drk/tillage/datasets/human/chip/geo/GSM1...    64      1      sum   \n",
       "4517  /home/drk/tillage/datasets/human/chip/geo/GSM1...    64      1      sum   \n",
       "4518  /home/drk/tillage/datasets/human/chip/geo/GSM1...    64      1      sum   \n",
       "4519  /home/drk/tillage/datasets/human/chip/geo/GSM1...    64      1      sum   \n",
       "4520  /home/drk/tillage/datasets/human/chip/geo/GSM1...    64      1      sum   \n",
       "\n",
       "                                            description   exp  \n",
       "4516  .:batch1_chrom1_LoVo_ARNT_Mouse_PassedQC / LoV...  CHIP  \n",
       "4517  .:batch1_chrom1_LoVo_ASCL2_Mouse_PassedQC / Lo...  CHIP  \n",
       "4518  .:batch1_chrom1_LoVo_ATOH1_Rabbit_PassedQC / L...  CHIP  \n",
       "4519  .:batch1_chrom1_LoVo_BARX2_Goat_PassedQC / LoV...  CHIP  \n",
       "4520  .:batch1_chrom1_LoVo_CAMTA2_Goat_PassedQC / Lo...  CHIP  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chip_sum = targets[(targets[\"exp\"] == \"CHIP\") & (targets[\"sum_stat\"] == \"sum\")]\n",
    "print(f\"All CHIP-seq experiments with sum_stat = sum have an identifier that starts with GSM: {chip_sum['identifier'].str.startswith('GSM').sum() == len(chip_sum)}\")\n",
    "chip_sum.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "11768aa3-15a8-4bbf-9eb9-3be9ee7de473",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All CHIP-seq experiments with sum_stat = mean have an identifier that starts with ENC: True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>genome</th>\n",
       "      <th>identifier</th>\n",
       "      <th>file</th>\n",
       "      <th>clip</th>\n",
       "      <th>scale</th>\n",
       "      <th>sum_stat</th>\n",
       "      <th>description</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>684</th>\n",
       "      <td>684</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF991NDB</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>CTCF:MCF-7</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>685</th>\n",
       "      <td>685</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF831FHP</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>TAF1:MCF-7</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>686</th>\n",
       "      <td>686</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF776DPQ</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K4me3:GM12878</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>687</th>\n",
       "      <td>687</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF279CYY</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>CTCF:GM12878</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688</th>\n",
       "      <td>688</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF340JIF</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K27ac:GM12878</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index  genome   identifier  \\\n",
       "684    684       0  ENCFF991NDB   \n",
       "685    685       0  ENCFF831FHP   \n",
       "686    686       0  ENCFF776DPQ   \n",
       "687    687       0  ENCFF279CYY   \n",
       "688    688       0  ENCFF340JIF   \n",
       "\n",
       "                                                  file  clip  scale sum_stat  \\\n",
       "684  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "685  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "686  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "687  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "688  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "\n",
       "         description   exp  \n",
       "684       CTCF:MCF-7  CHIP  \n",
       "685       TAF1:MCF-7  CHIP  \n",
       "686  H3K4me3:GM12878  CHIP  \n",
       "687     CTCF:GM12878  CHIP  \n",
       "688  H3K27ac:GM12878  CHIP  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chip_mean = targets[(targets[\"exp\"] == \"CHIP\") & (targets[\"sum_stat\"] == \"mean\")]\n",
    "print(f\"All CHIP-seq experiments with sum_stat = mean have an identifier that starts with ENC: {chip_mean['identifier'].str.startswith('ENC').sum() == len(chip_mean)}\")\n",
    "chip_mean.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97e63590-600c-476a-a19a-3ac28fc00413",
   "metadata": {},
   "source": [
    "The `scale` parameter is set to\n",
    "\n",
    "* 2\n",
    "    - DNASE\n",
    "    - CHIP from ENCODE\n",
    "* 1\n",
    "    - ATAC\n",
    "    - CHIP from GSM\n",
    "    - CAGE\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bbb4b6c7-d48e-4990-a6fe-64dc129ffd38",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNASE: [2]\n",
      "ATAC: [1]\n",
      "CHIP: [2 1]\n",
      "CAGE: [1]\n"
     ]
    }
   ],
   "source": [
    "for exp in targets.exp.unique():\n",
    "    print(f\"{exp}: {targets[targets['exp'] == exp]['scale'].unique()}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b3836f04-cb54-4606-9861-6b4f98a8536d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All CHIP-seq experiments with scale = 2 have an identifier that starts with ENC: True\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>genome</th>\n",
       "      <th>identifier</th>\n",
       "      <th>file</th>\n",
       "      <th>clip</th>\n",
       "      <th>scale</th>\n",
       "      <th>sum_stat</th>\n",
       "      <th>description</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>684</th>\n",
       "      <td>684</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF991NDB</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>CTCF:MCF-7</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>685</th>\n",
       "      <td>685</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF831FHP</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>TAF1:MCF-7</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>686</th>\n",
       "      <td>686</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF776DPQ</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K4me3:GM12878</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>687</th>\n",
       "      <td>687</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF279CYY</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>CTCF:GM12878</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>688</th>\n",
       "      <td>688</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF340JIF</td>\n",
       "      <td>/home/drk/tillage/datasets/human/chip/encode/E...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>H3K27ac:GM12878</td>\n",
       "      <td>CHIP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     index  genome   identifier  \\\n",
       "684    684       0  ENCFF991NDB   \n",
       "685    685       0  ENCFF831FHP   \n",
       "686    686       0  ENCFF776DPQ   \n",
       "687    687       0  ENCFF279CYY   \n",
       "688    688       0  ENCFF340JIF   \n",
       "\n",
       "                                                  file  clip  scale sum_stat  \\\n",
       "684  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "685  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "686  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "687  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "688  /home/drk/tillage/datasets/human/chip/encode/E...    32      2     mean   \n",
       "\n",
       "         description   exp  \n",
       "684       CTCF:MCF-7  CHIP  \n",
       "685       TAF1:MCF-7  CHIP  \n",
       "686  H3K4me3:GM12878  CHIP  \n",
       "687     CTCF:GM12878  CHIP  \n",
       "688  H3K27ac:GM12878  CHIP  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chip_mean = targets[(targets[\"exp\"] == \"CHIP\") & (targets[\"scale\"] == 2)]\n",
    "print(f\"All CHIP-seq experiments with scale = 2 have an identifier that starts with ENC: {chip_mean['identifier'].str.startswith('ENC').sum() == len(chip_mean)}\")\n",
    "chip_mean.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a01a415a-b76c-4f69-97af-9cc3c34ba3d3",
   "metadata": {},
   "source": [
    "## Supplementary table\n",
    "\n",
    "From the supplementary table from Enformer we can see that only files from Fantom and GEO have been reprocessed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a47aebfd-fb40-42ba-9de7-aca1947d9b66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome</th>\n",
       "      <th>experiment</th>\n",
       "      <th>identifier</th>\n",
       "      <th>source</th>\n",
       "      <th>clip</th>\n",
       "      <th>reprocessed</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>human</td>\n",
       "      <td>dnase</td>\n",
       "      <td>ENCFF833POA</td>\n",
       "      <td>encode</td>\n",
       "      <td>32</td>\n",
       "      <td>False</td>\n",
       "      <td>cerebellum male adult (27 years) and male adul...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>human</td>\n",
       "      <td>dnase</td>\n",
       "      <td>ENCFF110QGM</td>\n",
       "      <td>encode</td>\n",
       "      <td>32</td>\n",
       "      <td>False</td>\n",
       "      <td>frontal cortex male adult (27 years) and male ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human</td>\n",
       "      <td>dnase</td>\n",
       "      <td>ENCFF880MKD</td>\n",
       "      <td>encode</td>\n",
       "      <td>32</td>\n",
       "      <td>False</td>\n",
       "      <td>chorion\\n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>human</td>\n",
       "      <td>dnase</td>\n",
       "      <td>ENCFF463ZLQ</td>\n",
       "      <td>encode</td>\n",
       "      <td>32</td>\n",
       "      <td>False</td>\n",
       "      <td>Ishikawa treated with 0.02% dimethyl sulfoxide...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human</td>\n",
       "      <td>dnase</td>\n",
       "      <td>ENCFF890OGQ</td>\n",
       "      <td>encode</td>\n",
       "      <td>32</td>\n",
       "      <td>False</td>\n",
       "      <td>GM03348\\n</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  genome experiment   identifier  source  clip  reprocessed  \\\n",
       "0  human      dnase  ENCFF833POA  encode    32        False   \n",
       "1  human      dnase  ENCFF110QGM  encode    32        False   \n",
       "2  human      dnase  ENCFF880MKD  encode    32        False   \n",
       "3  human      dnase  ENCFF463ZLQ  encode    32        False   \n",
       "4  human      dnase  ENCFF890OGQ  encode    32        False   \n",
       "\n",
       "                                         description  \n",
       "0  cerebellum male adult (27 years) and male adul...  \n",
       "1  frontal cortex male adult (27 years) and male ...  \n",
       "2                                          chorion\\n  \n",
       "3  Ishikawa treated with 0.02% dimethyl sulfoxide...  \n",
       "4                                          GM03348\\n  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "supp = pd.read_excel(\"/omics/groups/OE0540/internal/users/mikulik/master_thesis/data/pcbi.1008050.s019.xlsx\")#, sheet_name= \"pcbi.1008050.s019\")#sep=\" \")\n",
    "supp = supp.rename(columns={\"identifiers\":\"identifier\"})\n",
    "supp.head()\n",
    "\n",
    "#test = test.append({\"genome\":\"human\", \"experiment\":\"dnase\", \"identifiers\":\"A\", \"source\":\"encode\", \"clip\":32, \"reprocessed\": True, \"description\":0}, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7377bf48-81f4-4b69-af2a-b9ea32d1f7fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "source    reprocessed\n",
       "encode    False          5525\n",
       "fantom    True            995\n",
       "geo       True            351\n",
       "uw-atlas  False            85\n",
       "dtype: int64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Only Fantom and GEO files have been reprocessed!\n",
    "supp.groupby([\"source\", \"reprocessed\"]).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "46f4f2c4-24ce-4ec7-8529-ba1c36006933",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mouse: (1643, 7)\n",
      "Human: (5313, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome</th>\n",
       "      <th>experiment</th>\n",
       "      <th>identifier</th>\n",
       "      <th>source</th>\n",
       "      <th>clip</th>\n",
       "      <th>reprocessed</th>\n",
       "      <th>description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>human</td>\n",
       "      <td>dnase</td>\n",
       "      <td>ENCFF833POA</td>\n",
       "      <td>encode</td>\n",
       "      <td>32</td>\n",
       "      <td>False</td>\n",
       "      <td>cerebellum male adult (27 years) and male adul...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>human</td>\n",
       "      <td>dnase</td>\n",
       "      <td>ENCFF110QGM</td>\n",
       "      <td>encode</td>\n",
       "      <td>32</td>\n",
       "      <td>False</td>\n",
       "      <td>frontal cortex male adult (27 years) and male ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>human</td>\n",
       "      <td>dnase</td>\n",
       "      <td>ENCFF880MKD</td>\n",
       "      <td>encode</td>\n",
       "      <td>32</td>\n",
       "      <td>False</td>\n",
       "      <td>chorion\\n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>human</td>\n",
       "      <td>dnase</td>\n",
       "      <td>ENCFF463ZLQ</td>\n",
       "      <td>encode</td>\n",
       "      <td>32</td>\n",
       "      <td>False</td>\n",
       "      <td>Ishikawa treated with 0.02% dimethyl sulfoxide...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>human</td>\n",
       "      <td>dnase</td>\n",
       "      <td>ENCFF890OGQ</td>\n",
       "      <td>encode</td>\n",
       "      <td>32</td>\n",
       "      <td>False</td>\n",
       "      <td>GM03348\\n</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  genome experiment   identifier  source  clip  reprocessed  \\\n",
       "0  human      dnase  ENCFF833POA  encode    32        False   \n",
       "1  human      dnase  ENCFF110QGM  encode    32        False   \n",
       "2  human      dnase  ENCFF880MKD  encode    32        False   \n",
       "3  human      dnase  ENCFF463ZLQ  encode    32        False   \n",
       "4  human      dnase  ENCFF890OGQ  encode    32        False   \n",
       "\n",
       "                                         description  \n",
       "0  cerebellum male adult (27 years) and male adul...  \n",
       "1  frontal cortex male adult (27 years) and male ...  \n",
       "2                                          chorion\\n  \n",
       "3  Ishikawa treated with 0.02% dimethyl sulfoxide...  \n",
       "4                                          GM03348\\n  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(f'Mouse: {supp[supp[\"genome\"] == \"mouse\"].shape}')\n",
    "human_supp = supp[supp[\"genome\"] == \"human\"]\n",
    "print(f'Human: {human_supp.shape}')\n",
    "human_supp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c941c125-073b-4dd4-9537-edc105752603",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>genome</th>\n",
       "      <th>identifier</th>\n",
       "      <th>file</th>\n",
       "      <th>clip</th>\n",
       "      <th>scale</th>\n",
       "      <th>sum_stat</th>\n",
       "      <th>description</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF833POA</td>\n",
       "      <td>/home/drk/tillage/datasets/human/dnase/encode/...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>cerebellum male adult (27 years) and male adul...</td>\n",
       "      <td>DNASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF110QGM</td>\n",
       "      <td>/home/drk/tillage/datasets/human/dnase/encode/...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>frontal cortex male adult (27 years) and male ...</td>\n",
       "      <td>DNASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF880MKD</td>\n",
       "      <td>/home/drk/tillage/datasets/human/dnase/encode/...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>chorion</td>\n",
       "      <td>DNASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF463ZLQ</td>\n",
       "      <td>/home/drk/tillage/datasets/human/dnase/encode/...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>Ishikawa treated with 0.02% dimethyl sulfoxide...</td>\n",
       "      <td>DNASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF890OGQ</td>\n",
       "      <td>/home/drk/tillage/datasets/human/dnase/encode/...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>GM03348</td>\n",
       "      <td>DNASE</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index  genome   identifier  \\\n",
       "0      0       0  ENCFF833POA   \n",
       "1      1       0  ENCFF110QGM   \n",
       "2      2       0  ENCFF880MKD   \n",
       "3      3       0  ENCFF463ZLQ   \n",
       "4      4       0  ENCFF890OGQ   \n",
       "\n",
       "                                                file  clip  scale sum_stat  \\\n",
       "0  /home/drk/tillage/datasets/human/dnase/encode/...    32      2     mean   \n",
       "1  /home/drk/tillage/datasets/human/dnase/encode/...    32      2     mean   \n",
       "2  /home/drk/tillage/datasets/human/dnase/encode/...    32      2     mean   \n",
       "3  /home/drk/tillage/datasets/human/dnase/encode/...    32      2     mean   \n",
       "4  /home/drk/tillage/datasets/human/dnase/encode/...    32      2     mean   \n",
       "\n",
       "                                         description    exp  \n",
       "0  cerebellum male adult (27 years) and male adul...  DNASE  \n",
       "1  frontal cortex male adult (27 years) and male ...  DNASE  \n",
       "2                                            chorion  DNASE  \n",
       "3  Ishikawa treated with 0.02% dimethyl sulfoxide...  DNASE  \n",
       "4                                            GM03348  DNASE  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "targets.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b0d683f9-7978-4baa-827f-fd1017548d7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>genome_x</th>\n",
       "      <th>identifier</th>\n",
       "      <th>file</th>\n",
       "      <th>clip</th>\n",
       "      <th>scale</th>\n",
       "      <th>sum_stat</th>\n",
       "      <th>description_x</th>\n",
       "      <th>exp</th>\n",
       "      <th>genome_y</th>\n",
       "      <th>experiment</th>\n",
       "      <th>source</th>\n",
       "      <th>reprocessed</th>\n",
       "      <th>description_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF833POA</td>\n",
       "      <td>/home/drk/tillage/datasets/human/dnase/encode/...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>cerebellum male adult (27 years) and male adul...</td>\n",
       "      <td>DNASE</td>\n",
       "      <td>human</td>\n",
       "      <td>dnase</td>\n",
       "      <td>encode</td>\n",
       "      <td>False</td>\n",
       "      <td>cerebellum male adult (27 years) and male adul...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF110QGM</td>\n",
       "      <td>/home/drk/tillage/datasets/human/dnase/encode/...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>frontal cortex male adult (27 years) and male ...</td>\n",
       "      <td>DNASE</td>\n",
       "      <td>human</td>\n",
       "      <td>dnase</td>\n",
       "      <td>encode</td>\n",
       "      <td>False</td>\n",
       "      <td>frontal cortex male adult (27 years) and male ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF880MKD</td>\n",
       "      <td>/home/drk/tillage/datasets/human/dnase/encode/...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>chorion</td>\n",
       "      <td>DNASE</td>\n",
       "      <td>human</td>\n",
       "      <td>dnase</td>\n",
       "      <td>encode</td>\n",
       "      <td>False</td>\n",
       "      <td>chorion\\n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF463ZLQ</td>\n",
       "      <td>/home/drk/tillage/datasets/human/dnase/encode/...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>Ishikawa treated with 0.02% dimethyl sulfoxide...</td>\n",
       "      <td>DNASE</td>\n",
       "      <td>human</td>\n",
       "      <td>dnase</td>\n",
       "      <td>encode</td>\n",
       "      <td>False</td>\n",
       "      <td>Ishikawa treated with 0.02% dimethyl sulfoxide...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>ENCFF890OGQ</td>\n",
       "      <td>/home/drk/tillage/datasets/human/dnase/encode/...</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>mean</td>\n",
       "      <td>GM03348</td>\n",
       "      <td>DNASE</td>\n",
       "      <td>human</td>\n",
       "      <td>dnase</td>\n",
       "      <td>encode</td>\n",
       "      <td>False</td>\n",
       "      <td>GM03348\\n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5308</th>\n",
       "      <td>5308</td>\n",
       "      <td>0</td>\n",
       "      <td>CNhs14239</td>\n",
       "      <td>/home/drk/tillage/datasets/human/cage/fantom/C...</td>\n",
       "      <td>384</td>\n",
       "      <td>1</td>\n",
       "      <td>sum</td>\n",
       "      <td>epithelioid sarcoma cell line:HS-ES-2R</td>\n",
       "      <td>CAGE</td>\n",
       "      <td>human</td>\n",
       "      <td>cage</td>\n",
       "      <td>fantom</td>\n",
       "      <td>True</td>\n",
       "      <td>epithelioid sarcoma cell line:HS-ES-2R\\n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5309</th>\n",
       "      <td>5309</td>\n",
       "      <td>0</td>\n",
       "      <td>CNhs14240</td>\n",
       "      <td>/home/drk/tillage/datasets/human/cage/fantom/C...</td>\n",
       "      <td>384</td>\n",
       "      <td>1</td>\n",
       "      <td>sum</td>\n",
       "      <td>squamous cell lung carcinoma cell line:RERF-LC-AI</td>\n",
       "      <td>CAGE</td>\n",
       "      <td>human</td>\n",
       "      <td>cage</td>\n",
       "      <td>fantom</td>\n",
       "      <td>True</td>\n",
       "      <td>squamous cell lung carcinoma cell line:RERF-LC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5310</th>\n",
       "      <td>5310</td>\n",
       "      <td>0</td>\n",
       "      <td>CNhs14241</td>\n",
       "      <td>/home/drk/tillage/datasets/human/cage/fantom/C...</td>\n",
       "      <td>384</td>\n",
       "      <td>1</td>\n",
       "      <td>sum</td>\n",
       "      <td>gastric cancer cell line:GSS</td>\n",
       "      <td>CAGE</td>\n",
       "      <td>human</td>\n",
       "      <td>cage</td>\n",
       "      <td>fantom</td>\n",
       "      <td>True</td>\n",
       "      <td>gastric cancer cell line:GSS\\n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5311</th>\n",
       "      <td>5311</td>\n",
       "      <td>0</td>\n",
       "      <td>CNhs14244</td>\n",
       "      <td>/home/drk/tillage/datasets/human/cage/fantom/C...</td>\n",
       "      <td>384</td>\n",
       "      <td>1</td>\n",
       "      <td>sum</td>\n",
       "      <td>carcinoid cell line:NCI-H727</td>\n",
       "      <td>CAGE</td>\n",
       "      <td>human</td>\n",
       "      <td>cage</td>\n",
       "      <td>fantom</td>\n",
       "      <td>True</td>\n",
       "      <td>carcinoid cell line:NCI-H727\\n</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5312</th>\n",
       "      <td>5312</td>\n",
       "      <td>0</td>\n",
       "      <td>CNhs14245</td>\n",
       "      <td>/home/drk/tillage/datasets/human/cage/fantom/C...</td>\n",
       "      <td>384</td>\n",
       "      <td>1</td>\n",
       "      <td>sum</td>\n",
       "      <td>lung adenocarcinoma, papillary cell line:NCI-H441</td>\n",
       "      <td>CAGE</td>\n",
       "      <td>human</td>\n",
       "      <td>cage</td>\n",
       "      <td>fantom</td>\n",
       "      <td>True</td>\n",
       "      <td>lung adenocarcinoma, papillary cell line:NCI-H...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5313 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index  genome_x   identifier  \\\n",
       "0         0         0  ENCFF833POA   \n",
       "1         1         0  ENCFF110QGM   \n",
       "2         2         0  ENCFF880MKD   \n",
       "3         3         0  ENCFF463ZLQ   \n",
       "4         4         0  ENCFF890OGQ   \n",
       "...     ...       ...          ...   \n",
       "5308   5308         0    CNhs14239   \n",
       "5309   5309         0    CNhs14240   \n",
       "5310   5310         0    CNhs14241   \n",
       "5311   5311         0    CNhs14244   \n",
       "5312   5312         0    CNhs14245   \n",
       "\n",
       "                                                   file  clip  scale sum_stat  \\\n",
       "0     /home/drk/tillage/datasets/human/dnase/encode/...    32      2     mean   \n",
       "1     /home/drk/tillage/datasets/human/dnase/encode/...    32      2     mean   \n",
       "2     /home/drk/tillage/datasets/human/dnase/encode/...    32      2     mean   \n",
       "3     /home/drk/tillage/datasets/human/dnase/encode/...    32      2     mean   \n",
       "4     /home/drk/tillage/datasets/human/dnase/encode/...    32      2     mean   \n",
       "...                                                 ...   ...    ...      ...   \n",
       "5308  /home/drk/tillage/datasets/human/cage/fantom/C...   384      1      sum   \n",
       "5309  /home/drk/tillage/datasets/human/cage/fantom/C...   384      1      sum   \n",
       "5310  /home/drk/tillage/datasets/human/cage/fantom/C...   384      1      sum   \n",
       "5311  /home/drk/tillage/datasets/human/cage/fantom/C...   384      1      sum   \n",
       "5312  /home/drk/tillage/datasets/human/cage/fantom/C...   384      1      sum   \n",
       "\n",
       "                                          description_x    exp genome_y  \\\n",
       "0     cerebellum male adult (27 years) and male adul...  DNASE    human   \n",
       "1     frontal cortex male adult (27 years) and male ...  DNASE    human   \n",
       "2                                               chorion  DNASE    human   \n",
       "3     Ishikawa treated with 0.02% dimethyl sulfoxide...  DNASE    human   \n",
       "4                                               GM03348  DNASE    human   \n",
       "...                                                 ...    ...      ...   \n",
       "5308             epithelioid sarcoma cell line:HS-ES-2R   CAGE    human   \n",
       "5309  squamous cell lung carcinoma cell line:RERF-LC-AI   CAGE    human   \n",
       "5310                       gastric cancer cell line:GSS   CAGE    human   \n",
       "5311                       carcinoid cell line:NCI-H727   CAGE    human   \n",
       "5312  lung adenocarcinoma, papillary cell line:NCI-H441   CAGE    human   \n",
       "\n",
       "     experiment  source  reprocessed  \\\n",
       "0         dnase  encode        False   \n",
       "1         dnase  encode        False   \n",
       "2         dnase  encode        False   \n",
       "3         dnase  encode        False   \n",
       "4         dnase  encode        False   \n",
       "...         ...     ...          ...   \n",
       "5308       cage  fantom         True   \n",
       "5309       cage  fantom         True   \n",
       "5310       cage  fantom         True   \n",
       "5311       cage  fantom         True   \n",
       "5312       cage  fantom         True   \n",
       "\n",
       "                                          description_y  \n",
       "0     cerebellum male adult (27 years) and male adul...  \n",
       "1     frontal cortex male adult (27 years) and male ...  \n",
       "2                                             chorion\\n  \n",
       "3     Ishikawa treated with 0.02% dimethyl sulfoxide...  \n",
       "4                                             GM03348\\n  \n",
       "...                                                 ...  \n",
       "5308           epithelioid sarcoma cell line:HS-ES-2R\\n  \n",
       "5309  squamous cell lung carcinoma cell line:RERF-LC...  \n",
       "5310                     gastric cancer cell line:GSS\\n  \n",
       "5311                     carcinoid cell line:NCI-H727\\n  \n",
       "5312  lung adenocarcinoma, papillary cell line:NCI-H...  \n",
       "\n",
       "[5313 rows x 14 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.merge(targets, human_supp, how = \"inner\", on = [\"identifier\", \"clip\"])#, \"description\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b36b689-ca16-41ac-8911-f20a260fdee9",
   "metadata": {},
   "source": [
    "# Basenji1 Supplementary Table Fantom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "868f9440-1e6a-49cc-8c9b-39324a562585",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Identifier</th>\n",
       "      <th>Data</th>\n",
       "      <th>Project</th>\n",
       "      <th>BioRep</th>\n",
       "      <th>TechRep</th>\n",
       "      <th>Sequencing</th>\n",
       "      <th>Label</th>\n",
       "      <th>AccessDate</th>\n",
       "      <th>URL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CNhs11051</td>\n",
       "      <td>CAGE</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>single</td>\n",
       "      <td>Adipocyte - breast, donor1</td>\n",
       "      <td>1/16/2017</td>\n",
       "      <td>ftp://ftp.ddbj.nig.ac.jp/ddbj_database/dra/fas...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CNhs11969</td>\n",
       "      <td>CAGE</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>single</td>\n",
       "      <td>Adipocyte - breast, donor2</td>\n",
       "      <td>1/16/2017</td>\n",
       "      <td>ftp://ftp.ddbj.nig.ac.jp/ddbj_database/dra/fas...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CNhs11054</td>\n",
       "      <td>CAGE</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>single</td>\n",
       "      <td>Adipocyte - omental, donor1</td>\n",
       "      <td>1/16/2017</td>\n",
       "      <td>ftp://ftp.ddbj.nig.ac.jp/ddbj_database/dra/fas...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CNhs12067</td>\n",
       "      <td>CAGE</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>single</td>\n",
       "      <td>Adipocyte - omental, donor2</td>\n",
       "      <td>1/16/2017</td>\n",
       "      <td>ftp://ftp.ddbj.nig.ac.jp/ddbj_database/dra/fas...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CNhs12068</td>\n",
       "      <td>CAGE</td>\n",
       "      <td>FANTOM</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>single</td>\n",
       "      <td>Adipocyte - omental, donor3</td>\n",
       "      <td>1/16/2017</td>\n",
       "      <td>ftp://ftp.ddbj.nig.ac.jp/ddbj_database/dra/fas...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4224</th>\n",
       "      <td>GSM817172</td>\n",
       "      <td>DNASE</td>\n",
       "      <td>Roadmap</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>single</td>\n",
       "      <td>thymus, fetal day127 M</td>\n",
       "      <td>2/3/2017</td>\n",
       "      <td>ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-insta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4225</th>\n",
       "      <td>GSM665823</td>\n",
       "      <td>DNASE</td>\n",
       "      <td>Roadmap</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>single</td>\n",
       "      <td>thymus, fetal day147 F</td>\n",
       "      <td>2/3/2017</td>\n",
       "      <td>ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-insta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4226</th>\n",
       "      <td>GSM878627</td>\n",
       "      <td>DNASE</td>\n",
       "      <td>Roadmap</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>single</td>\n",
       "      <td>thymus, fetal day97 M</td>\n",
       "      <td>2/3/2017</td>\n",
       "      <td>ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-insta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4227</th>\n",
       "      <td>GSM774230</td>\n",
       "      <td>DNASE</td>\n",
       "      <td>Roadmap</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>single</td>\n",
       "      <td>thymus, fetal day98 F</td>\n",
       "      <td>2/3/2017</td>\n",
       "      <td>ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-insta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4228</th>\n",
       "      <td>GSM701537</td>\n",
       "      <td>DNASE</td>\n",
       "      <td>Roadmap</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>single</td>\n",
       "      <td>thymus, fetal day98 F</td>\n",
       "      <td>2/3/2017</td>\n",
       "      <td>ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-insta...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4229 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Identifier   Data  Project  BioRep  TechRep Sequencing  \\\n",
       "0     CNhs11051   CAGE   FANTOM       1        1     single   \n",
       "1     CNhs11969   CAGE   FANTOM       2        1     single   \n",
       "2     CNhs11054   CAGE   FANTOM       1        1     single   \n",
       "3     CNhs12067   CAGE   FANTOM       2        1     single   \n",
       "4     CNhs12068   CAGE   FANTOM       3        1     single   \n",
       "...         ...    ...      ...     ...      ...        ...   \n",
       "4224  GSM817172  DNASE  Roadmap       1        1     single   \n",
       "4225  GSM665823  DNASE  Roadmap       1        1     single   \n",
       "4226  GSM878627  DNASE  Roadmap       1        1     single   \n",
       "4227  GSM774230  DNASE  Roadmap       1        1     single   \n",
       "4228  GSM701537  DNASE  Roadmap       2        1     single   \n",
       "\n",
       "                            Label AccessDate  \\\n",
       "0      Adipocyte - breast, donor1  1/16/2017   \n",
       "1      Adipocyte - breast, donor2  1/16/2017   \n",
       "2     Adipocyte - omental, donor1  1/16/2017   \n",
       "3     Adipocyte - omental, donor2  1/16/2017   \n",
       "4     Adipocyte - omental, donor3  1/16/2017   \n",
       "...                           ...        ...   \n",
       "4224       thymus, fetal day127 M   2/3/2017   \n",
       "4225       thymus, fetal day147 F   2/3/2017   \n",
       "4226        thymus, fetal day97 M   2/3/2017   \n",
       "4227        thymus, fetal day98 F   2/3/2017   \n",
       "4228        thymus, fetal day98 F   2/3/2017   \n",
       "\n",
       "                                                    URL  \n",
       "0     ftp://ftp.ddbj.nig.ac.jp/ddbj_database/dra/fas...  \n",
       "1     ftp://ftp.ddbj.nig.ac.jp/ddbj_database/dra/fas...  \n",
       "2     ftp://ftp.ddbj.nig.ac.jp/ddbj_database/dra/fas...  \n",
       "3     ftp://ftp.ddbj.nig.ac.jp/ddbj_database/dra/fas...  \n",
       "4     ftp://ftp.ddbj.nig.ac.jp/ddbj_database/dra/fas...  \n",
       "...                                                 ...  \n",
       "4224  ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-insta...  \n",
       "4225  ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-insta...  \n",
       "4226  ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-insta...  \n",
       "4227  ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-insta...  \n",
       "4228  ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-insta...  \n",
       "\n",
       "[4229 rows x 9 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_excel(\"/omics/groups/OE0540/internal/users/mikulik/master_thesis/data/Supplemental_Table_S1.xls\", sheet_name=\"Sheet1\", header=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "fada2ed0-e8e4-4de0-a94a-119cf4bc935e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_31585/2567327423.py:1: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.\n",
      "  a = pd.read_csv(\"/omics/groups/OE0540/internal/users/mikulik/master_thesis/data/Supplemental_Table_S1_Basenji.csv\",sep='delimiter', header=None)# header=0, sep=\",\", on_bad_lines='skip')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Identifier;Data;Project;BioRep;TechRep;Sequenc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CNhs11051;CAGE;FANTOM;1;1;single;Adipocyte - b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CNhs11969;CAGE;FANTOM;2;1;single;Adipocyte - b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CNhs11054;CAGE;FANTOM;1;1;single;Adipocyte - o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CNhs12067;CAGE;FANTOM;2;1;single;Adipocyte - o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4225</th>\n",
       "      <td>GSM817172;DNASE;Roadmap;1;1;single;thymus, fet...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4226</th>\n",
       "      <td>GSM665823;DNASE;Roadmap;1;1;single;thymus, fet...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4227</th>\n",
       "      <td>GSM878627;DNASE;Roadmap;1;1;single;thymus, fet...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4228</th>\n",
       "      <td>GSM774230;DNASE;Roadmap;1;1;single;thymus, fet...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4229</th>\n",
       "      <td>GSM701537;DNASE;Roadmap;2;1;single;thymus, fet...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4230 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      0\n",
       "0     Identifier;Data;Project;BioRep;TechRep;Sequenc...\n",
       "1     CNhs11051;CAGE;FANTOM;1;1;single;Adipocyte - b...\n",
       "2     CNhs11969;CAGE;FANTOM;2;1;single;Adipocyte - b...\n",
       "3     CNhs11054;CAGE;FANTOM;1;1;single;Adipocyte - o...\n",
       "4     CNhs12067;CAGE;FANTOM;2;1;single;Adipocyte - o...\n",
       "...                                                 ...\n",
       "4225  GSM817172;DNASE;Roadmap;1;1;single;thymus, fet...\n",
       "4226  GSM665823;DNASE;Roadmap;1;1;single;thymus, fet...\n",
       "4227  GSM878627;DNASE;Roadmap;1;1;single;thymus, fet...\n",
       "4228  GSM774230;DNASE;Roadmap;1;1;single;thymus, fet...\n",
       "4229  GSM701537;DNASE;Roadmap;2;1;single;thymus, fet...\n",
       "\n",
       "[4230 rows x 1 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = pd.read_csv(\"/omics/group^s/OE0540/internal/users/mikulik/master_thesis/data/Supplemental_Table_S1_Basenji.csv\",sep='delimiter', header=None)# header=0, sep=\",\", on_bad_lines='skip')\n",
    "a"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a9716fc-d521-4af5-8e69-d39ea0ccacbb",
   "metadata": {},
   "source": [
    "# Region Dataframe\n",
    "\n",
    "From this it seems that the overlap between the individual model sequences is 65473 in the Basenji2/Enformer data at least. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "788e2e4b-3719-4b6d-869b-6177d02d0ee1",
   "metadata": {},
   "source": [
    "### Basenji1 data\n",
    "\n",
    "The data for basenji1 I got from `https://github.com/calico/basenji/blob/master/manuscripts/genome_research2018/get_data.sh`\n",
    "\n",
    "AS described in the paper these sequences are non-overlapping! If there is a gap this is because of the gap files or blacklisted regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "38ef5c7a-2c54-4d6f-8e17-f6fcc6d67090",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>subset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>163560601</td>\n",
       "      <td>163691673</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr14</td>\n",
       "      <td>50546679</td>\n",
       "      <td>50677751</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7</td>\n",
       "      <td>93980486</td>\n",
       "      <td>94111558</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>92888465</td>\n",
       "      <td>93019537</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr20</td>\n",
       "      <td>51689140</td>\n",
       "      <td>51820212</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom      start        end subset\n",
       "0   chr2  163560601  163691673  train\n",
       "1  chr14   50546679   50677751  train\n",
       "2   chr7   93980486   94111558  train\n",
       "3   chrX   92888465   93019537  train\n",
       "4  chr20   51689140   51820212  train"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "basenji1 = pd.read_csv(f\"{data_dir}l131k_w128.bed\", sep=\"\\t\", header=None)\n",
    "basenji1.columns = ['chrom', 'start', 'end', 'subset']\n",
    "basenji1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "47aae525-3d4c-49ff-a213-5a366259920f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>subset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12700</th>\n",
       "      <td>chr1</td>\n",
       "      <td>825115</td>\n",
       "      <td>956187</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4072</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1138175</td>\n",
       "      <td>1269247</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4097</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1286804</td>\n",
       "      <td>1417876</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6010</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1417876</td>\n",
       "      <td>1548948</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3642</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1548948</td>\n",
       "      <td>1680020</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6758</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247403790</td>\n",
       "      <td>247534862</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1921</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247534862</td>\n",
       "      <td>247665934</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8909</th>\n",
       "      <td>chr1</td>\n",
       "      <td>247665934</td>\n",
       "      <td>247797006</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5150</th>\n",
       "      <td>chr1</td>\n",
       "      <td>248697213</td>\n",
       "      <td>248828285</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10623</th>\n",
       "      <td>chr1</td>\n",
       "      <td>249061968</td>\n",
       "      <td>249193040</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1166 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom      start        end subset\n",
       "12700  chr1     825115     956187  train\n",
       "4072   chr1    1138175    1269247  train\n",
       "4097   chr1    1286804    1417876  train\n",
       "6010   chr1    1417876    1548948  train\n",
       "3642   chr1    1548948    1680020  train\n",
       "...     ...        ...        ...    ...\n",
       "6758   chr1  247403790  247534862  train\n",
       "1921   chr1  247534862  247665934  train\n",
       "8909   chr1  247665934  247797006  train\n",
       "5150   chr1  248697213  248828285  train\n",
       "10623  chr1  249061968  249193040  train\n",
       "\n",
       "[1166 rows x 4 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "basenji1[basenji1[\"chrom\"] == \"chr1\"].sort_values(by=\"start\", ascending=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "92ce923d-8409-41f2-ae4f-6408a4d182d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17557"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1286804 - 1269247"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81869a71-9d46-4d3f-9b90-36aa49d3cf8a",
   "metadata": {},
   "source": [
    "### Basenji2/Enformer data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "5309e8d3-a6ee-4e53-b5ee-4709e920cfb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 38171 human sequences and 33521 mouse sequences.\n"
     ]
    }
   ],
   "source": [
    "human_regions = pd.read_csv(f\"{data_dir}gcs_basenj/human/sequences.bed\", sep=\"\\t\", header=None)\n",
    "human_regions.columns = ['chrom', 'start', 'end', 'subset']\n",
    "mouse_regions = pd.read_csv(f\"{data_dir}gcs_basenj/mouse/sequences.bed\", sep=\"\\t\", header=None)\n",
    "mouse_regions.columns = ['chrom', 'start', 'end', 'subset']\n",
    "print(f\"There are {len(human_regions)} human sequences and {len(mouse_regions)} mouse sequences.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "60bf70ca-81f8-4fc8-9750-41f4f341abc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>subset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr18</td>\n",
       "      <td>928386</td>\n",
       "      <td>1059458</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>113630947</td>\n",
       "      <td>113762019</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>18427720</td>\n",
       "      <td>18558792</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16</td>\n",
       "      <td>85805681</td>\n",
       "      <td>85936753</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3</td>\n",
       "      <td>158386188</td>\n",
       "      <td>158517260</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom      start        end subset\n",
       "0  chr18     928386    1059458  train\n",
       "1   chr4  113630947  113762019  train\n",
       "2  chr11   18427720   18558792  train\n",
       "3  chr16   85805681   85936753  train\n",
       "4   chr3  158386188  158517260  train"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_regions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "2a287cbd-e066-4aca-8113-7b2592545e88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method NDFrame.head of       chrom      start        end subset  seq_len\n",
       "11833  chr1     848384     979456  train   131072\n",
       "20752  chr1     913983    1045055  train   131072\n",
       "32487  chr1     979582    1110654  train   131072\n",
       "25406  chr1    1045181    1176253  train   131072\n",
       "18812  chr1    1110780    1241852  train   131072\n",
       "...     ...        ...        ...    ...      ...\n",
       "35294  chrX  155141683  155272755  valid   131072\n",
       "35373  chrX  155485750  155616822  valid   131072\n",
       "34326  chrX  155600439  155731511  valid   131072\n",
       "35856  chrX  155715128  155846200  valid   131072\n",
       "35704  chrX  155829817  155960889  valid   131072\n",
       "\n",
       "[38171 rows x 5 columns]>"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_df[\"seq_len\"] = region_df.end - region_df.start\n",
    "region_df = region_df.sort_values([\"chrom\", \"start\"])\n",
    "region_df.head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "8570dbeb-8b62-4c38-bcf4-69f489f2b702",
   "metadata": {},
   "outputs": [],
   "source": [
    "overlap = [0]\n",
    "for i in range(len(region_df)):\n",
    "    if i == 0:\n",
    "        continue\n",
    "    if i == len(region_df)-1:\n",
    "        overlap.append(0)\n",
    "    else:\n",
    "        overlap.append(int(region_df.iloc[i].end -  region_df.iloc[i+1].start) )\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "206aa582-d5d4-4c17-b492-fc7d67901e43",
   "metadata": {},
   "outputs": [],
   "source": [
    "region_df[\"overlap\"] = overlap\n",
    "region_df[\"seq_len_2\"] = region_df.seq_len/2\n",
    "region_df[\"difference\"] = region_df.seq_len_2 - region_df.overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "070a6626-df2b-4ba6-b89a-cb3559575944",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>subset</th>\n",
       "      <th>seq_len</th>\n",
       "      <th>overlap</th>\n",
       "      <th>seq_len/2</th>\n",
       "      <th>seq_len_2</th>\n",
       "      <th>difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11833</th>\n",
       "      <td>chr1</td>\n",
       "      <td>848384</td>\n",
       "      <td>979456</td>\n",
       "      <td>train</td>\n",
       "      <td>131072</td>\n",
       "      <td>0</td>\n",
       "      <td>65536.0</td>\n",
       "      <td>65536.0</td>\n",
       "      <td>65536.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20752</th>\n",
       "      <td>chr1</td>\n",
       "      <td>913983</td>\n",
       "      <td>1045055</td>\n",
       "      <td>train</td>\n",
       "      <td>131072</td>\n",
       "      <td>65473</td>\n",
       "      <td>65536.0</td>\n",
       "      <td>65536.0</td>\n",
       "      <td>63.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32487</th>\n",
       "      <td>chr1</td>\n",
       "      <td>979582</td>\n",
       "      <td>1110654</td>\n",
       "      <td>train</td>\n",
       "      <td>131072</td>\n",
       "      <td>65473</td>\n",
       "      <td>65536.0</td>\n",
       "      <td>65536.0</td>\n",
       "      <td>63.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25406</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1045181</td>\n",
       "      <td>1176253</td>\n",
       "      <td>train</td>\n",
       "      <td>131072</td>\n",
       "      <td>65473</td>\n",
       "      <td>65536.0</td>\n",
       "      <td>65536.0</td>\n",
       "      <td>63.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18812</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1110780</td>\n",
       "      <td>1241852</td>\n",
       "      <td>train</td>\n",
       "      <td>131072</td>\n",
       "      <td>65473</td>\n",
       "      <td>65536.0</td>\n",
       "      <td>65536.0</td>\n",
       "      <td>63.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom    start      end subset  seq_len  overlap  seq_len/2  seq_len_2  \\\n",
       "11833  chr1   848384   979456  train   131072        0    65536.0    65536.0   \n",
       "20752  chr1   913983  1045055  train   131072    65473    65536.0    65536.0   \n",
       "32487  chr1   979582  1110654  train   131072    65473    65536.0    65536.0   \n",
       "25406  chr1  1045181  1176253  train   131072    65473    65536.0    65536.0   \n",
       "18812  chr1  1110780  1241852  train   131072    65473    65536.0    65536.0   \n",
       "\n",
       "       difference  \n",
       "11833     65536.0  \n",
       "20752        63.0  \n",
       "32487        63.0  \n",
       "25406        63.0  \n",
       "18812        63.0  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ad9f598-198c-4567-a26d-59a64abee32c",
   "metadata": {},
   "source": [
    "Are all sequences the same length?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "d1dd472f-d8df-4ee4-817e-c0fe710062e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>subset</th>\n",
       "      <th>seq_len</th>\n",
       "      <th>overlap</th>\n",
       "      <th>seq_len/2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chrom, start, end, subset, seq_len, overlap, seq_len/2]\n",
       "Index: []"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_df[region_df.seq_len != 131072]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e0a8615-73f6-4ca4-a375-ed848ae05f72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['train', 'test', 'valid'], dtype=object)"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region_df.subset.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4f8e022-0ff7-4ba3-bb38-662c21359601",
   "metadata": {},
   "source": [
    "# Centering around TSS\n",
    "\n",
    "In the original publication you get the impression that all sequences are centered around the TSS. At least for the examples they visualize this is the case. Also, they say that they compute the correlation for CAGE signal using the 3 bins around the TSS only, but not the entire sequence. For this I need to know whether all input sequences we train on are actually centered around the TSS. From my exploration of the sequences from above and the Basenji2 data preprocessing scripts it seem like this is not the case, since they split the contigs of a chromosome into contiguous 131kb regions with an overlap of 65,472. \n",
    "\n",
    "**Canonical Ensembl transcript**\n",
    "For protein-coding gnees this is the transcript that on balance has the highest coverage of conserved exons, highest expression, longest coding sequenc. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7a66c7ab-f989-4632-882f-a348cc11f4fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  start    end strand            gene_id  \\\n",
       "0    chr1  11869  14409      +  ENSG00000223972.5   \n",
       "1    chr1  11869  14409      +  ENSG00000223972.5   \n",
       "2    chr1  11869  12227      +  ENSG00000223972.5   \n",
       "3    chr1  12613  12721      +  ENSG00000223972.5   \n",
       "4    chr1  13221  14409      +  ENSG00000223972.5   \n",
       "\n",
       "                            gene_type gene_name  \n",
       "0  transcribed_unprocessed_pseudogene   DDX11L1  \n",
       "1  transcribed_unprocessed_pseudogene   DDX11L1  \n",
       "2  transcribed_unprocessed_pseudogene   DDX11L1  \n",
       "3  transcribed_unprocessed_pseudogene   DDX11L1  \n",
       "4  transcribed_unprocessed_pseudogene   DDX11L1  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "data_dir = \"/omics/groups/OE0540/internal/users/mikulik/master_thesis/data/\"\n",
    "gene_list = pd.read_csv(os.path.join(data_dir, \"gencode_v32_hg38.tsv\"), sep=\"\\t\", index_col=0)\n",
    "gene_list.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d669470c-7cfe-4e8f-a0c5-77356de09392",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_10870/3930467463.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_list_sub[\"length\"] = gene_list_sub.end - gene_list_sub.start\n",
      "/tmp/ipykernel_10870/3930467463.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_list_sub[\"gene_length\"] = np.where(gene_list_sub.strand == \"+\", gene_list_sub.end - gene_list_sub.start, gene_list_sub.start - gene_list_sub.end)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname  start    end strand            gene_id  \\\n",
       "0    chr1  11869  14409      +  ENSG00000223972.5   \n",
       "1    chr1  11869  14409      +  ENSG00000223972.5   \n",
       "2    chr1  11869  12227      +  ENSG00000223972.5   \n",
       "3    chr1  12613  12721      +  ENSG00000223972.5   \n",
       "4    chr1  13221  14409      +  ENSG00000223972.5   \n",
       "\n",
       "                            gene_type gene_name  \n",
       "0  transcribed_unprocessed_pseudogene   DDX11L1  \n",
       "1  transcribed_unprocessed_pseudogene   DDX11L1  \n",
       "2  transcribed_unprocessed_pseudogene   DDX11L1  \n",
       "3  transcribed_unprocessed_pseudogene   DDX11L1  \n",
       "4  transcribed_unprocessed_pseudogene   DDX11L1  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_list_sub = gene_list.drop_duplicates(keep=\"first\", inplace=False)\n",
    "gene_list_sub[\"length\"] = gene_list_sub.end - gene_list_sub.start\n",
    "gene_list_sub[\"gene_length\"] = np.where(gene_list_sub.strand == \"+\", gene_list_sub.end - gene_list_sub.start, gene_list_sub.start - gene_list_sub.end)\n",
    "gene_list.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "67320c7c-fb67-4d83-b5dd-7f67ee5dd246",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of genes in total is: 2899774, but only 2554032 of these are protein-coding genes.\n"
     ]
    }
   ],
   "source": [
    "print(f\"The number of genes in total is: {len(gene_list)}, but only {len(gene_list[gene_list['gene_type'] == 'protein_coding'])} of these are protein-coding genes.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0399d0e3-33cc-4c23-a5d0-a0be75621544",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['transcribed_unprocessed_pseudogene', 'unprocessed_pseudogene',\n",
       "       'miRNA', 'lncRNA', 'protein_coding', 'processed_pseudogene',\n",
       "       'snRNA', 'transcribed_processed_pseudogene', 'misc_RNA', 'TEC',\n",
       "       'transcribed_unitary_pseudogene', 'snoRNA', 'scaRNA',\n",
       "       'rRNA_pseudogene', 'unitary_pseudogene', 'polymorphic_pseudogene',\n",
       "       'pseudogene', 'rRNA', 'IG_V_pseudogene', 'scRNA', 'IG_V_gene',\n",
       "       'IG_C_gene', 'IG_J_gene', 'sRNA', 'ribozyme',\n",
       "       'translated_processed_pseudogene', 'vaultRNA', 'TR_C_gene',\n",
       "       'TR_J_gene', 'TR_V_gene', 'TR_V_pseudogene',\n",
       "       'translated_unprocessed_pseudogene', 'TR_D_gene',\n",
       "       'IG_C_pseudogene', 'TR_J_pseudogene', 'IG_J_pseudogene',\n",
       "       'IG_D_gene', 'IG_pseudogene', 'Mt_tRNA', 'Mt_rRNA'], dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_list.gene_type.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d325820f-5eb9-4aae-a748-44a3b4967354",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list[gene_list[\"gene_type\"]==\"protein_coding\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ccd1130-9638-4716-84a6-3f131365876a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list[\"length\"] = [gene_list.iloc[i].end - gene_list.iloc[i].start if gene_list.iloc[i].strand == \"+\" else gene_list.iloc[i].start - gene_list.end for i in range(len(gene_list))]\n",
    "gene_list.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "398d0e37-fd0b-4179-8223-3b3674a97fe2",
   "metadata": {},
   "source": [
    "For now I will simply keep 20,000 protein coding genes, keeping the first entry for each gene for simplicity. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d6637c1f-2e86-40d4-bf0e-675dd75d216a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chr1</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000186092.6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>chr1</td>\n",
       "      <td>450703</td>\n",
       "      <td>451697</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000284733.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>chr1</td>\n",
       "      <td>685679</td>\n",
       "      <td>686673</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000284662.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>870</th>\n",
       "      <td>chr1</td>\n",
       "      <td>923928</td>\n",
       "      <td>944581</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187634.12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1227</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944203</td>\n",
       "      <td>959309</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188976.11</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NOC2L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2899728</th>\n",
       "      <td>chrM</td>\n",
       "      <td>10470</td>\n",
       "      <td>10766</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000212907.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND4L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2899735</th>\n",
       "      <td>chrM</td>\n",
       "      <td>10760</td>\n",
       "      <td>12137</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000198886.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2899749</th>\n",
       "      <td>chrM</td>\n",
       "      <td>12337</td>\n",
       "      <td>14148</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000198786.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2899755</th>\n",
       "      <td>chrM</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2899763</th>\n",
       "      <td>chrM</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-CYB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19965 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname   start     end strand             gene_id       gene_type  \\\n",
       "57         chr1   65419   71585      +   ENSG00000186092.6  protein_coding   \n",
       "326        chr1  450703  451697      -   ENSG00000284733.1  protein_coding   \n",
       "454        chr1  685679  686673      -   ENSG00000284662.1  protein_coding   \n",
       "870        chr1  923928  944581      +  ENSG00000187634.12  protein_coding   \n",
       "1227       chr1  944203  959309      -  ENSG00000188976.11  protein_coding   \n",
       "...         ...     ...     ...    ...                 ...             ...   \n",
       "2899728    chrM   10470   10766      +   ENSG00000212907.2  protein_coding   \n",
       "2899735    chrM   10760   12137      +   ENSG00000198886.2  protein_coding   \n",
       "2899749    chrM   12337   14148      +   ENSG00000198786.2  protein_coding   \n",
       "2899755    chrM   14149   14673      -   ENSG00000198695.2  protein_coding   \n",
       "2899763    chrM   14747   15887      +   ENSG00000198727.2  protein_coding   \n",
       "\n",
       "        gene_name  \n",
       "57          OR4F5  \n",
       "326        OR4F29  \n",
       "454        OR4F16  \n",
       "870        SAMD11  \n",
       "1227        NOC2L  \n",
       "...           ...  \n",
       "2899728   MT-ND4L  \n",
       "2899735    MT-ND4  \n",
       "2899749    MT-ND5  \n",
       "2899755    MT-ND6  \n",
       "2899763    MT-CYB  \n",
       "\n",
       "[19965 rows x 7 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = gene_list[gene_list[\"gene_type\"] == \"protein_coding\"]\n",
    "genes = genes.drop_duplicates(subset=\"gene_id\",keep=\"first\", inplace=False)\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d1b13e6e-7c90-456b-8bbf-e00c7b9ac0a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1751728</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280193</td>\n",
       "      <td>54287088</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname     start       end strand             gene_id  \\\n",
       "1751728   chr12  54280193  54287088      +  ENSG00000135486.17   \n",
       "\n",
       "              gene_type gene_name  \n",
       "1751728  protein_coding   HNRNPA1  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes[genes.gene_name == \"HNRNPA1\"]#.gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "db388a04-7e00-4af6-a750-d2574b15fd90",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_bed_file = \"%s/gencode_v32_hg38.bed\" % data_dir\n",
    "if not os.path.isfile(gene_bed_file):\n",
    "    gene_bed_file_out = open(gene_bed_file, \"w\")\n",
    "    for index, row in genes.iterrows():\n",
    "        line = \"%s\\t%d\\t%d\\t%s\\t%s\" % (row.seqname, row.start, row.end, row.strand, row.gene_id)\n",
    "        print(line, file = gene_bed_file_out)\n",
    "    gene_bed_file_out.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "faa759a5-d7fb-44d4-a6a3-b46a7c9462dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 38171 human sequences.\n"
     ]
    }
   ],
   "source": [
    "human_regions = pd.read_csv(f\"{data_dir}gcs_basenj/human/sequences.bed\", sep=\"\\t\", header=None)\n",
    "human_regions.columns = ['chrom', 'start', 'end', 'subset']\n",
    "print(f\"There are {len(human_regions)} human sequences.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1a1608e3-fbea-46cc-96d0-424bcfbd5e10",
   "metadata": {},
   "outputs": [],
   "source": [
    "human_regions_file = \"%sgcs_basenj/human/sequences.bed\" % data_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ec4859f2-0c41-4ce1-ad5e-293e1c4f43d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybedtools\n",
    "genes_bed = pybedtools.BedTool(gene_bed_file)\n",
    "human_bed = pybedtools.BedTool(human_regions_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a710cb6a-ff5a-4b9e-be2d-e9c17d9e4178",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_bed = \"%soverlap.bed\" % data_dir\n",
    "out_bed_open = open(out_bed, \"w\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "60622ddf-ae15-4bf5-bfc2-0cfa1827c17b",
   "metadata": {},
   "outputs": [],
   "source": [
    "terminal=False\n",
    "if terminal:\n",
    "    out_bed = \"%soverlap.bed\" % data_dir\n",
    "    out_bed_open = open(out_bed, \"w\")\n",
    "    for overlap in genes_bed.intersect(human_bed, wo=True):\n",
    "        chrom1 = overlap[0]\n",
    "        start1 = int(overlap[1])\n",
    "        end1 = int(overlap[2])\n",
    "        strand = overlap[3]\n",
    "        gene_id = overlap[4]\n",
    "        chrom = overlap[5]\n",
    "        start = int(overlap[6])\n",
    "        end = int(overlap[7])\n",
    "        subset = overlap[8]\n",
    "        lap = int(overlap[9])\n",
    "        #print(type(overlap))\n",
    "        line = \"%s\\t%d\\t%d\\t%s\\t%s\\t%s\\t%d\\t%d\\t%s\\t%d\" % (chrom1, start1, end1, strand, gene_id, chrom, start, end, subset, lap)#(overlap[0], int(overlap[1]), int(overlap[2]), overlap[3], overlap[4], overlap[5], int(overlap[6]), int(overlap[7]), overlap[8], int(overlap[9]))\n",
    "        print(line, file = out_bed_open)\n",
    "    out_bed_open.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f40c94fb-8e2f-497b-9a60-33ae1d8b2410",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_gene</th>\n",
       "      <th>start_gene</th>\n",
       "      <th>end_gene</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chr_input</th>\n",
       "      <th>start_input</th>\n",
       "      <th>end_input</th>\n",
       "      <th>subset</th>\n",
       "      <th>overlap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>923928</td>\n",
       "      <td>944581</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187634.12</td>\n",
       "      <td>chr1</td>\n",
       "      <td>848384</td>\n",
       "      <td>979456</td>\n",
       "      <td>train</td>\n",
       "      <td>20653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>923928</td>\n",
       "      <td>944581</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187634.12</td>\n",
       "      <td>chr1</td>\n",
       "      <td>913983</td>\n",
       "      <td>1045055</td>\n",
       "      <td>train</td>\n",
       "      <td>20653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944203</td>\n",
       "      <td>959309</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188976.11</td>\n",
       "      <td>chr1</td>\n",
       "      <td>848384</td>\n",
       "      <td>979456</td>\n",
       "      <td>train</td>\n",
       "      <td>15106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944203</td>\n",
       "      <td>959309</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188976.11</td>\n",
       "      <td>chr1</td>\n",
       "      <td>913983</td>\n",
       "      <td>1045055</td>\n",
       "      <td>train</td>\n",
       "      <td>15106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>960584</td>\n",
       "      <td>965719</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187961.14</td>\n",
       "      <td>chr1</td>\n",
       "      <td>848384</td>\n",
       "      <td>979456</td>\n",
       "      <td>train</td>\n",
       "      <td>5135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53465</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155258235</td>\n",
       "      <td>155264491</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000155961.5</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155141683</td>\n",
       "      <td>155272755</td>\n",
       "      <td>valid</td>\n",
       "      <td>6256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53466</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155489011</td>\n",
       "      <td>155669944</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185973.11</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155600439</td>\n",
       "      <td>155731511</td>\n",
       "      <td>valid</td>\n",
       "      <td>69505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53467</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155489011</td>\n",
       "      <td>155669944</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185973.11</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155485750</td>\n",
       "      <td>155616822</td>\n",
       "      <td>valid</td>\n",
       "      <td>127811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53468</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155767812</td>\n",
       "      <td>155782459</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000168939.11</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155715128</td>\n",
       "      <td>155846200</td>\n",
       "      <td>valid</td>\n",
       "      <td>14647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53469</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155881345</td>\n",
       "      <td>155943769</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000124333.16</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155829817</td>\n",
       "      <td>155960889</td>\n",
       "      <td>valid</td>\n",
       "      <td>62424</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>53470 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr_gene  start_gene   end_gene strand             gene_id chr_input  \\\n",
       "0         chr1      923928     944581      +  ENSG00000187634.12      chr1   \n",
       "1         chr1      923928     944581      +  ENSG00000187634.12      chr1   \n",
       "2         chr1      944203     959309      -  ENSG00000188976.11      chr1   \n",
       "3         chr1      944203     959309      -  ENSG00000188976.11      chr1   \n",
       "4         chr1      960584     965719      +  ENSG00000187961.14      chr1   \n",
       "...        ...         ...        ...    ...                 ...       ...   \n",
       "53465     chrX   155258235  155264491      -   ENSG00000155961.5      chrX   \n",
       "53466     chrX   155489011  155669944      -  ENSG00000185973.11      chrX   \n",
       "53467     chrX   155489011  155669944      -  ENSG00000185973.11      chrX   \n",
       "53468     chrX   155767812  155782459      +  ENSG00000168939.11      chrX   \n",
       "53469     chrX   155881345  155943769      +  ENSG00000124333.16      chrX   \n",
       "\n",
       "       start_input  end_input subset  overlap  \n",
       "0           848384     979456  train    20653  \n",
       "1           913983    1045055  train    20653  \n",
       "2           848384     979456  train    15106  \n",
       "3           913983    1045055  train    15106  \n",
       "4           848384     979456  train     5135  \n",
       "...            ...        ...    ...      ...  \n",
       "53465    155141683  155272755  valid     6256  \n",
       "53466    155600439  155731511  valid    69505  \n",
       "53467    155485750  155616822  valid   127811  \n",
       "53468    155715128  155846200  valid    14647  \n",
       "53469    155829817  155960889  valid    62424  \n",
       "\n",
       "[53470 rows x 10 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_bed = pd.read_csv(os.path.join(data_dir, \"overlap.bed\"), sep=\"\\t\", header=None)\n",
    "overlap_bed.columns=[\"chr_gene\", \"start_gene\", \"end_gene\", \"strand\", \"gene_id\", \"chr_input\", \"start_input\", \"end_input\", \"subset\", \"overlap\"]\n",
    "overlap_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "88d712b7-93e0-4db1-b9e1-d9738a015167",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_gene</th>\n",
       "      <th>start_gene</th>\n",
       "      <th>end_gene</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chr_input</th>\n",
       "      <th>start_input</th>\n",
       "      <th>end_input</th>\n",
       "      <th>subset</th>\n",
       "      <th>overlap</th>\n",
       "      <th>strand_start_gene</th>\n",
       "      <th>seq_len</th>\n",
       "      <th>center</th>\n",
       "      <th>overlap_tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>923928</td>\n",
       "      <td>944581</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187634.12</td>\n",
       "      <td>chr1</td>\n",
       "      <td>848384</td>\n",
       "      <td>979456</td>\n",
       "      <td>train</td>\n",
       "      <td>20653</td>\n",
       "      <td>923928</td>\n",
       "      <td>131072</td>\n",
       "      <td>913920.0</td>\n",
       "      <td>10008.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>923928</td>\n",
       "      <td>944581</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187634.12</td>\n",
       "      <td>chr1</td>\n",
       "      <td>913983</td>\n",
       "      <td>1045055</td>\n",
       "      <td>train</td>\n",
       "      <td>20653</td>\n",
       "      <td>923928</td>\n",
       "      <td>131072</td>\n",
       "      <td>979519.0</td>\n",
       "      <td>55591.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944203</td>\n",
       "      <td>959309</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188976.11</td>\n",
       "      <td>chr1</td>\n",
       "      <td>848384</td>\n",
       "      <td>979456</td>\n",
       "      <td>train</td>\n",
       "      <td>15106</td>\n",
       "      <td>959309</td>\n",
       "      <td>131072</td>\n",
       "      <td>913920.0</td>\n",
       "      <td>45389.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>944203</td>\n",
       "      <td>959309</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000188976.11</td>\n",
       "      <td>chr1</td>\n",
       "      <td>913983</td>\n",
       "      <td>1045055</td>\n",
       "      <td>train</td>\n",
       "      <td>15106</td>\n",
       "      <td>959309</td>\n",
       "      <td>131072</td>\n",
       "      <td>979519.0</td>\n",
       "      <td>20210.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>960584</td>\n",
       "      <td>965719</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000187961.14</td>\n",
       "      <td>chr1</td>\n",
       "      <td>848384</td>\n",
       "      <td>979456</td>\n",
       "      <td>train</td>\n",
       "      <td>5135</td>\n",
       "      <td>960584</td>\n",
       "      <td>131072</td>\n",
       "      <td>913920.0</td>\n",
       "      <td>46664.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53465</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155258235</td>\n",
       "      <td>155264491</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000155961.5</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155141683</td>\n",
       "      <td>155272755</td>\n",
       "      <td>valid</td>\n",
       "      <td>6256</td>\n",
       "      <td>155264491</td>\n",
       "      <td>131072</td>\n",
       "      <td>155207219.0</td>\n",
       "      <td>57272.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53466</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155489011</td>\n",
       "      <td>155669944</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185973.11</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155600439</td>\n",
       "      <td>155731511</td>\n",
       "      <td>valid</td>\n",
       "      <td>69505</td>\n",
       "      <td>155669944</td>\n",
       "      <td>131072</td>\n",
       "      <td>155665975.0</td>\n",
       "      <td>3969.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53467</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155489011</td>\n",
       "      <td>155669944</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000185973.11</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155485750</td>\n",
       "      <td>155616822</td>\n",
       "      <td>valid</td>\n",
       "      <td>127811</td>\n",
       "      <td>155669944</td>\n",
       "      <td>131072</td>\n",
       "      <td>155551286.0</td>\n",
       "      <td>118658.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53468</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155767812</td>\n",
       "      <td>155782459</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000168939.11</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155715128</td>\n",
       "      <td>155846200</td>\n",
       "      <td>valid</td>\n",
       "      <td>14647</td>\n",
       "      <td>155767812</td>\n",
       "      <td>131072</td>\n",
       "      <td>155780664.0</td>\n",
       "      <td>12852.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53469</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155881345</td>\n",
       "      <td>155943769</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000124333.16</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155829817</td>\n",
       "      <td>155960889</td>\n",
       "      <td>valid</td>\n",
       "      <td>62424</td>\n",
       "      <td>155881345</td>\n",
       "      <td>131072</td>\n",
       "      <td>155895353.0</td>\n",
       "      <td>14008.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>53470 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr_gene  start_gene   end_gene strand             gene_id chr_input  \\\n",
       "0         chr1      923928     944581      +  ENSG00000187634.12      chr1   \n",
       "1         chr1      923928     944581      +  ENSG00000187634.12      chr1   \n",
       "2         chr1      944203     959309      -  ENSG00000188976.11      chr1   \n",
       "3         chr1      944203     959309      -  ENSG00000188976.11      chr1   \n",
       "4         chr1      960584     965719      +  ENSG00000187961.14      chr1   \n",
       "...        ...         ...        ...    ...                 ...       ...   \n",
       "53465     chrX   155258235  155264491      -   ENSG00000155961.5      chrX   \n",
       "53466     chrX   155489011  155669944      -  ENSG00000185973.11      chrX   \n",
       "53467     chrX   155489011  155669944      -  ENSG00000185973.11      chrX   \n",
       "53468     chrX   155767812  155782459      +  ENSG00000168939.11      chrX   \n",
       "53469     chrX   155881345  155943769      +  ENSG00000124333.16      chrX   \n",
       "\n",
       "       start_input  end_input subset  overlap  strand_start_gene  seq_len  \\\n",
       "0           848384     979456  train    20653             923928   131072   \n",
       "1           913983    1045055  train    20653             923928   131072   \n",
       "2           848384     979456  train    15106             959309   131072   \n",
       "3           913983    1045055  train    15106             959309   131072   \n",
       "4           848384     979456  train     5135             960584   131072   \n",
       "...            ...        ...    ...      ...                ...      ...   \n",
       "53465    155141683  155272755  valid     6256          155264491   131072   \n",
       "53466    155600439  155731511  valid    69505          155669944   131072   \n",
       "53467    155485750  155616822  valid   127811          155669944   131072   \n",
       "53468    155715128  155846200  valid    14647          155767812   131072   \n",
       "53469    155829817  155960889  valid    62424          155881345   131072   \n",
       "\n",
       "            center  overlap_tss  \n",
       "0         913920.0      10008.0  \n",
       "1         979519.0      55591.0  \n",
       "2         913920.0      45389.0  \n",
       "3         979519.0      20210.0  \n",
       "4         913920.0      46664.0  \n",
       "...            ...          ...  \n",
       "53465  155207219.0      57272.0  \n",
       "53466  155665975.0       3969.0  \n",
       "53467  155551286.0     118658.0  \n",
       "53468  155780664.0      12852.0  \n",
       "53469  155895353.0      14008.0  \n",
       "\n",
       "[53470 rows x 14 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_bed[\"strand_start_gene\"] = np.where(overlap_bed.strand == \"+\", overlap_bed.start_gene, overlap_bed.end_gene)\n",
    "overlap_bed[\"seq_len\"] = overlap_bed.end_input - overlap_bed.start_input\n",
    "overlap_bed[\"center\"] = overlap_bed.start_input + np.abs(overlap_bed.seq_len / 2)\n",
    "overlap_bed[\"overlap_tss\"] = np.abs(overlap_bed.strand_start_gene - overlap_bed.center)\n",
    "overlap_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "8bd11bdf-b497-4811-b4b4-6a050ffa43cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import kipoiseq\n",
    "target_interval = kipoiseq.Interval('chr11', 35_082_742, 35_197_430)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6e9c4bb1-e189-4503-a13d-eeb91aee9d08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "35254774.0"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "35_197_430 + (35_197_430 - 35_082_742) / 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e2781ed9-6cb1-4232-b688-bd3281fc6887",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_gene</th>\n",
       "      <th>start_gene</th>\n",
       "      <th>end_gene</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chr_input</th>\n",
       "      <th>start_input</th>\n",
       "      <th>end_input</th>\n",
       "      <th>subset</th>\n",
       "      <th>overlap</th>\n",
       "      <th>strand_start_gene</th>\n",
       "      <th>seq_len</th>\n",
       "      <th>center</th>\n",
       "      <th>overlap_tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [chr_gene, start_gene, end_gene, strand, gene_id, chr_input, start_input, end_input, subset, overlap, strand_start_gene, seq_len, center, overlap_tss]\n",
       "Index: []"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_bed[(overlap_bed.overlap_tss < 500) & overlap_bed.chr_input == \"chr11\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "749d07aa-4964-48ff-93bc-50e05e376f34",
   "metadata": {},
   "source": [
    "#### HNRNPA1 gene TSS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4cedd039-cc48-4424-a73d-0d4ee5d42a0b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "54280933.0"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hnrnpa1_start = 54_223_589\n",
    "hnrnpa1_end = 54_338_277\n",
    "length = hnrnpa1_end - hnrnpa1_start\n",
    "center = hnrnpa1_start + length/2\n",
    "center"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a9d75cc1-aaaa-4861-9b79-31e880334296",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_gene</th>\n",
       "      <th>start_gene</th>\n",
       "      <th>end_gene</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chr_input</th>\n",
       "      <th>start_input</th>\n",
       "      <th>end_input</th>\n",
       "      <th>subset</th>\n",
       "      <th>overlap</th>\n",
       "      <th>strand_start_gene</th>\n",
       "      <th>seq_len</th>\n",
       "      <th>center</th>\n",
       "      <th>overlap_tss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>35379</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280193</td>\n",
       "      <td>54287088</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>chr12</td>\n",
       "      <td>54206621</td>\n",
       "      <td>54337693</td>\n",
       "      <td>train</td>\n",
       "      <td>6895</td>\n",
       "      <td>54280193</td>\n",
       "      <td>131072</td>\n",
       "      <td>54272157.0</td>\n",
       "      <td>8036.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35378</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280193</td>\n",
       "      <td>54287088</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>chr12</td>\n",
       "      <td>54272220</td>\n",
       "      <td>54403292</td>\n",
       "      <td>train</td>\n",
       "      <td>6895</td>\n",
       "      <td>54280193</td>\n",
       "      <td>131072</td>\n",
       "      <td>54337756.0</td>\n",
       "      <td>57563.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr_gene  start_gene  end_gene strand             gene_id chr_input  \\\n",
       "35379    chr12    54280193  54287088      +  ENSG00000135486.17     chr12   \n",
       "35378    chr12    54280193  54287088      +  ENSG00000135486.17     chr12   \n",
       "\n",
       "       start_input  end_input subset  overlap  strand_start_gene  seq_len  \\\n",
       "35379     54206621   54337693  train     6895           54280193   131072   \n",
       "35378     54272220   54403292  train     6895           54280193   131072   \n",
       "\n",
       "           center  overlap_tss  \n",
       "35379  54272157.0       8036.0  \n",
       "35378  54337756.0      57563.0  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_bed[overlap_bed.gene_id == \"ENSG00000135486.17\"].sort_values(\"start_input\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c03684a0-ab8d-4291-b7ea-00ac81b433b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-65473"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "54272220 - 54337693"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "0aaa7cbd-3065-4372-b75c-f66c11daf1d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [seqname, start, end, strand, gene_id, gene_type, gene_name]\n",
       "Index: []"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_list[(gene_list.gene_name == \"HNRNPA1\") & (gene_list.start == 54280933)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0599c066-0596-4b9b-99db-e63e1060006a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1751751</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280193</td>\n",
       "      <td>54280807</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751861</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280775</td>\n",
       "      <td>54280807</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751768</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280748</td>\n",
       "      <td>54280807</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751884</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280780</td>\n",
       "      <td>54280807</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751792</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280748</td>\n",
       "      <td>54280807</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751729</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280193</td>\n",
       "      <td>54287087</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751791</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54286953</td>\n",
       "      <td>54287088</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751769</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280748</td>\n",
       "      <td>54287088</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751795</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54286953</td>\n",
       "      <td>54287088</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751728</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280193</td>\n",
       "      <td>54287088</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>210 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname     start       end strand             gene_id  \\\n",
       "1751751   chr12  54280193  54280807      +  ENSG00000135486.17   \n",
       "1751861   chr12  54280775  54280807      +  ENSG00000135486.17   \n",
       "1751768   chr12  54280748  54280807      +  ENSG00000135486.17   \n",
       "1751884   chr12  54280780  54280807      +  ENSG00000135486.17   \n",
       "1751792   chr12  54280748  54280807      +  ENSG00000135486.17   \n",
       "...         ...       ...       ...    ...                 ...   \n",
       "1751729   chr12  54280193  54287087      +  ENSG00000135486.17   \n",
       "1751791   chr12  54286953  54287088      +  ENSG00000135486.17   \n",
       "1751769   chr12  54280748  54287088      +  ENSG00000135486.17   \n",
       "1751795   chr12  54286953  54287088      +  ENSG00000135486.17   \n",
       "1751728   chr12  54280193  54287088      +  ENSG00000135486.17   \n",
       "\n",
       "              gene_type gene_name  \n",
       "1751751  protein_coding   HNRNPA1  \n",
       "1751861  protein_coding   HNRNPA1  \n",
       "1751768  protein_coding   HNRNPA1  \n",
       "1751884  protein_coding   HNRNPA1  \n",
       "1751792  protein_coding   HNRNPA1  \n",
       "...                 ...       ...  \n",
       "1751729  protein_coding   HNRNPA1  \n",
       "1751791  protein_coding   HNRNPA1  \n",
       "1751769  protein_coding   HNRNPA1  \n",
       "1751795  protein_coding   HNRNPA1  \n",
       "1751728  protein_coding   HNRNPA1  \n",
       "\n",
       "[210 rows x 7 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_list[(gene_list.gene_name == \"HNRNPA1\")].sort_values(\"end\")# & (gene_list.end == 35254774)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "3b5544d4-4bac-427b-80fb-60e9a21c3bb9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1751731</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280808</td>\n",
       "      <td>54280822</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751732</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280808</td>\n",
       "      <td>54280810</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751756</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280808</td>\n",
       "      <td>54280822</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751757</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280808</td>\n",
       "      <td>54280810</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751771</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280808</td>\n",
       "      <td>54280822</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751772</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280808</td>\n",
       "      <td>54280810</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751803</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280808</td>\n",
       "      <td>54280822</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751804</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280808</td>\n",
       "      <td>54280810</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751830</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280808</td>\n",
       "      <td>54280822</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751831</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280808</td>\n",
       "      <td>54280810</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751841</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280808</td>\n",
       "      <td>54280822</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751842</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280808</td>\n",
       "      <td>54280810</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751866</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280808</td>\n",
       "      <td>54280822</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751867</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54280808</td>\n",
       "      <td>54280810</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751887</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54281006</td>\n",
       "      <td>54282626</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751888</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54281006</td>\n",
       "      <td>54281094</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751889</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54281023</td>\n",
       "      <td>54281094</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751890</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54281023</td>\n",
       "      <td>54281025</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751901</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54281006</td>\n",
       "      <td>54281022</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751902</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54281186</td>\n",
       "      <td>54282284</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751903</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54281186</td>\n",
       "      <td>54281295</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1751910</th>\n",
       "      <td>chr12</td>\n",
       "      <td>54281186</td>\n",
       "      <td>54281295</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000135486.17</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>HNRNPA1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname     start       end strand             gene_id  \\\n",
       "1751731   chr12  54280808  54280822      +  ENSG00000135486.17   \n",
       "1751732   chr12  54280808  54280810      +  ENSG00000135486.17   \n",
       "1751756   chr12  54280808  54280822      +  ENSG00000135486.17   \n",
       "1751757   chr12  54280808  54280810      +  ENSG00000135486.17   \n",
       "1751771   chr12  54280808  54280822      +  ENSG00000135486.17   \n",
       "1751772   chr12  54280808  54280810      +  ENSG00000135486.17   \n",
       "1751803   chr12  54280808  54280822      +  ENSG00000135486.17   \n",
       "1751804   chr12  54280808  54280810      +  ENSG00000135486.17   \n",
       "1751830   chr12  54280808  54280822      +  ENSG00000135486.17   \n",
       "1751831   chr12  54280808  54280810      +  ENSG00000135486.17   \n",
       "1751841   chr12  54280808  54280822      +  ENSG00000135486.17   \n",
       "1751842   chr12  54280808  54280810      +  ENSG00000135486.17   \n",
       "1751866   chr12  54280808  54280822      +  ENSG00000135486.17   \n",
       "1751867   chr12  54280808  54280810      +  ENSG00000135486.17   \n",
       "1751887   chr12  54281006  54282626      +  ENSG00000135486.17   \n",
       "1751888   chr12  54281006  54281094      +  ENSG00000135486.17   \n",
       "1751889   chr12  54281023  54281094      +  ENSG00000135486.17   \n",
       "1751890   chr12  54281023  54281025      +  ENSG00000135486.17   \n",
       "1751901   chr12  54281006  54281022      +  ENSG00000135486.17   \n",
       "1751902   chr12  54281186  54282284      +  ENSG00000135486.17   \n",
       "1751903   chr12  54281186  54281295      +  ENSG00000135486.17   \n",
       "1751910   chr12  54281186  54281295      +  ENSG00000135486.17   \n",
       "\n",
       "              gene_type gene_name  \n",
       "1751731  protein_coding   HNRNPA1  \n",
       "1751732  protein_coding   HNRNPA1  \n",
       "1751756  protein_coding   HNRNPA1  \n",
       "1751757  protein_coding   HNRNPA1  \n",
       "1751771  protein_coding   HNRNPA1  \n",
       "1751772  protein_coding   HNRNPA1  \n",
       "1751803  protein_coding   HNRNPA1  \n",
       "1751804  protein_coding   HNRNPA1  \n",
       "1751830  protein_coding   HNRNPA1  \n",
       "1751831  protein_coding   HNRNPA1  \n",
       "1751841  protein_coding   HNRNPA1  \n",
       "1751842  protein_coding   HNRNPA1  \n",
       "1751866  protein_coding   HNRNPA1  \n",
       "1751867  protein_coding   HNRNPA1  \n",
       "1751887  protein_coding   HNRNPA1  \n",
       "1751888  protein_coding   HNRNPA1  \n",
       "1751889  protein_coding   HNRNPA1  \n",
       "1751890  protein_coding   HNRNPA1  \n",
       "1751901  protein_coding   HNRNPA1  \n",
       "1751902  protein_coding   HNRNPA1  \n",
       "1751903  protein_coding   HNRNPA1  \n",
       "1751910  protein_coding   HNRNPA1  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_list[(gene_list.gene_name == \"HNRNPA1\") & ((gene_list.start <= 54281200) & (gene_list.start >= 54280800))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e3bc2045-14f6-4580-8853-ce37cb51ac2b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<matplotlib.lines.Line2D at 0x7f66e20bceb0>]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAiMAAAGsCAYAAAAPJKchAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAA9hAAAPYQGoP6dpAABQx0lEQVR4nO3deXwU9f0/8NdyBYQkChhIBDGgIoIgBhWsIBQLBbXar23tt1at2v7KV0AtUlq0rbYeaIsWrQhFOUQUFAOIgkiQHNyQkzMhQMh9kJBkc26u+f2BWbPJ7mZndmY+M7Ov5+ORxyPZzM68d3bmM+/5XGOTJEkCERERkSBdRAdAREREgY3JCBEREQnFZISIiIiEYjJCREREQjEZISIiIqGYjBAREZFQTEaIiIhIKCYjREREJBSTESIiIhKKyQgREREJZapkJCEhAffddx8iIiJgs9mwZcsW2euQJAmLFy/G9ddfj6CgIAwePBivvfaa+sESERGRT7qJDkCOmpoajBkzBo8//jgefPBBRet45plnsHPnTixevBg33XQTKisrUVpaqnKkRERE5CubWR+UZ7PZsHnzZjzwwAPO1xoaGvCXv/wFH3/8MSoqKjBq1Ci88cYbmDx5MgDg1KlTGD16NI4fP47hw4eLCZyIiIhcmKqZpjOPP/449u3bhw0bNuDo0aP4+c9/jh//+MfIzMwEAHz55ZcYOnQovvrqK0RGRuKaa67Bb3/7W1y8eFFw5ERERIHLMsnI2bNnsX79emzcuBETJ07EsGHDMH/+fNx5551YvXo1AODcuXPIzs7Gxo0bsXbtWqxZswZJSUn42c9+Jjh6IiKiwGWqPiPeJCcnQ5IkXH/99S6vOxwO9OvXDwDQ0tICh8OBtWvXOpdbuXIloqKikJGRwaYbIiIiASyTjLS0tKBr165ISkpC165dXf7Xp08fAEB4eDi6devmkrCMGDECAJCTk8NkhIiISADLJCNjx45Fc3MzSkpKMHHiRLfL/OAHP0BTUxPOnj2LYcOGAQBOnz4NABgyZIhusRIREdH3TDWaprq6GmfOnAFwKfl46623MGXKFPTt2xdXX301fv3rX2Pfvn148803MXbsWJSWlmL37t246aabMHPmTLS0tODWW29Fnz59sGTJErS0tGD27NkICQnBzp07BX86IiKiwGSqZCQuLg5Tpkzp8Ppjjz2GNWvWoLGxEa+88grWrl2L/Px89OvXDxMmTMDf//533HTTTQCAgoICzJ07Fzt37kTv3r0xY8YMvPnmm+jbt6/eH4eIiIhgsmSEiIiIrMcyQ3uJiIjInJiMEBERkVCmGE3T0tKCgoICBAcHw2aziQ6HiIiIfCBJEqqqqhAREYEuXTzXf5giGSkoKMDgwYNFh0FEREQK5ObmYtCgQR7/b4pkJDg4GMClDxMSEiI4GiIiIvKF3W7H4MGDnddxT0yRjLQ2zYSEhDAZISIiMpnOuliwAysREREJxWSEiIiIhGIyQkREREIxGSEiIiKhmIwQERGRUExGiIiISCgmI0RERCQUkxEiIiISiskIERERCcVkhIiIiIRiMkJERERCyUpGFi1ahFtvvRXBwcEICwvDAw88gIyMDK/viYuLg81m6/CTnp7uV+BERERkDbKSkfj4eMyePRsHDx5ETEwMmpqaMG3aNNTU1HT63oyMDBQWFjp/rrvuOsVBExGR+Zwpqcb7CedQ39gsOhQyGFlP7d2xY4fL36tXr0ZYWBiSkpIwadIkr+8NCwvD5ZdfLjtAIiKyhrvfigcAVNQ14I/TbxAcDRmJX31GKisrAQB9+/btdNmxY8ciPDwcU6dORWxsrNdlHQ4H7Ha7yw8REVlDSk6F6BDIYBQnI5IkYd68ebjzzjsxatQoj8uFh4djxYoViI6OxqZNmzB8+HBMnToVCQkJHt+zaNEihIaGOn8GDx6sNEwiIiIyOJskSZKSN86ePRvbtm3D3r17MWjQIFnvve+++2Cz2bB161a3/3c4HHA4HM6/7XY7Bg8ejMrKSoSEhCgJl4iIBLvmz9sAAHcM64dPfjdecDSkB7vdjtDQ0E6v34pqRubOnYutW7ciNjZWdiICAOPHj0dmZqbH/wcFBSEkJMTlh4iIiKxJVgdWSZIwd+5cbN68GXFxcYiMjFS00ZSUFISHhyt6LxEREVmLrGRk9uzZ+OSTT/DFF18gODgYRUVFAIDQ0FD06tULALBw4ULk5+dj7dq1AIAlS5bgmmuuwciRI9HQ0IB169YhOjoa0dHRKn8UIiIyA5tNdARkNLKSkWXLlgEAJk+e7PL66tWr8Zvf/AYAUFhYiJycHOf/GhoaMH/+fOTn56NXr14YOXIktm3bhpkzZ/oXOREREVmC4g6sevK1AwwRERlXawfWH1zbDx//lh1YA4GmHViJiIiI1MJkhIiIiIRiMkJERLqygT1YyRWTESIiIhKKyQgREREJxWSEiIiIhGIyQkREREIxGSEiIl1xBlZqj8kIERERCcVkhIiIiIRiMkJERERCMRkhIiIioZiMEBERkVBMRohIN8fyKnGhyiE6DCIymG6iAyCiwHA0rwI/eXcfAOD86/cIjoaIjIQ1I0SkiwNny0SHQEQGxWSEiIiIhGIyQkQd5JTV4uA5dWsyOOsmEXnCZISIOpj0r1j8csVBnCioVG2dNjAbIdJbXnktZr69B9FJeaJD8YrJCBF5dCLfLjoEIvLDS1tP4mShHc9tTBMdildMRohIF2ymIdJfbUOT6BB8wmSEiIiIhGIyQkREurKxmozaYTJCRLpoewE6mlchLhAiMhwmI0TkmYo3sG1XlV9ep96Kicj0mIwQERGRUExGiEgX7CZARJ4wGSEij9TMH5iLUCseC9QekxEiIiKLMkuNJJMRIvJIzSGYHM5JRJ4wGSEiXTAXISJPmIwQERGRUExGiEgXrBghIk+YjBCRR6omEGynoe/wUKD2mIwQERGRUExGiMgjNe9gA+VmuKSqHp8cyjHNo9uJjKCb6ACIKDAEStX8L5YfwPmyWqTlVuCNn40WHQ4FOJtJbgNYM0JEpKLzZbUAgG/TiwVHQmQeTEaIyCN1m2nMcYdG2uORQO0xGSEiXQRKM00rSRIdAZF5MBkhIiIioZiMEJEuAqxihIhkYDJCRB6p2c8j0JppiMh3TEaIyCMmEMqxy4hnfIIztcdkhIh0wdE01Epi717dmCXvYzJCRB7lldfh2Q0pOJ5f6f/KTFIoEpH+OAMrEXn0r28yAABbUgtw/vV7BEdjLrz7J/Ida0aISBdtK0bMUnVMRPpgMkJEumCnRWrFY6GjT4/k4JcrDqCytlF0KEIwGSEi0gAbaUiOP0Ufw8FzF/FubKboUIRgMkJEuuC9MFHnqh1NokMQgskIERERCcVkhIh0wW4CROQJkxEi0kWgJSMc2UvkOyYjREREJBSTESLSBaeDJyJPZCUjixYtwq233org4GCEhYXhgQceQEZGRqfvi4+PR1RUFHr27ImhQ4di+fLligMmInMKvGYattMQ+UpWMhIfH4/Zs2fj4MGDiImJQVNTE6ZNm4aamhqP78nKysLMmTMxceJEpKSk4Pnnn8fTTz+N6Ohov4MnsqK3Yk5j5tt7AnaIHxEFHlnPptmxY4fL36tXr0ZYWBiSkpIwadIkt+9Zvnw5rr76aixZsgQAMGLECCQmJmLx4sV48MEHlUVNZGHvfHtp0qNPDmXj/00aJjgaIvUFWCUZ+cCvPiOVlZee5Nm3b1+Pyxw4cADTpk1zeW369OlITExEY6P7aW8dDgfsdrvLD1GgaWy2VjV/oE0Bbq1vj0hbipMRSZIwb9483HnnnRg1apTH5YqKijBgwACX1wYMGICmpiaUlpa6fc+iRYsQGhrq/Bk8eLDSMImIiMjgFCcjc+bMwdGjR7F+/fpOl21/R9TascvTndLChQtRWVnp/MnNzVUaJhERGQxrjTxTu9+zWWokZfUZaTV37lxs3boVCQkJGDRokNdlBw4ciKKiIpfXSkpK0K1bN/Tr18/te4KCghAUFKQkNCIiIjIZWTUjkiRhzpw52LRpE3bv3o3IyMhO3zNhwgTExMS4vLZz506MGzcO3bt3lxctUQDh0FCT49fnkTnu1UlPspKR2bNnY926dfjkk08QHByMoqIiFBUVoa6uzrnMwoUL8eijjzr/njVrFrKzszFv3jycOnUKq1atwsqVKzF//nz1PgURERGZlqxkZNmyZaisrMTkyZMRHh7u/Pn000+dyxQWFiInJ8f5d2RkJLZv3464uDjcfPPNePnll/HOO+9wWC9RJ1gxQkSBQlafEV+qjdesWdPhtbvuugvJyclyNkVEZGrMJYl8x2fTBLiYk8X44ZtxOJ5fKToUsjj2EyAiT5iMBLjfrU3EuQs1+O2HiaJDoXZ4Z01WZZLRpqQjJiMEAKht4HNQiNTE0VBkBGbJ+5iMEBmUta9lZikiSQvWPrZJCSYjREREJBSTESIiIhKKyQgBMM/zCwKJxC6spsZvzzMWN9QekxEiIiKDCNT+NExGiAzKaoUS74aJOheoNaJMRoiINGC1ZJJIS0xGiIiILMosNZJMRogMijfW8n17qhjL488aYsIxs1wExODOIVeyHpRHRGRkT373WIMxgy7HhGH9BEdDnolPFslYWDNCZFQGuLs3q2J7vegQ+PWRIew/WyY6BJ8wGSEiIrKohqYW0SH4hMkIAWD7NhERicNkhMigrFbLbwuwTouBOl+EbwLrWKDOMRkhXRRU1GF3erEhRjkQEZGxMBkhANp3trvj9d14Yk0ivjlRrO2GLIR5GxEFCiYjBEC/PiMHz5mjZzeRv5hMEvmOyQgRWc7y+LNobmE2QOYTqEkskxEig2IHSOXSi6rw6ZFcoTE4TDKkksgImIwQkSWdLq4SHQIR+YjJCJFBWa26lnPZEJEnTEaIiIhIKCYjRAZ1pqRadAhERLpgMkK64qRnvtt50rpzsrDJJrDx+6f2mIwQAOv1TyAiIvNgMkJERERCMRkhXbEChvTCJkH/ZBZXYfW+LDQ2c74U0l430QEQEZHx/OjfCQCA5hYJv504VHA0ZHWsGSEiXbTts8hKC/NIy6sUHUJACdRTg8kIERERCcVkhIgsKVDvMMncArXWkMkI6SpQTzQiIvKMyQgBACrrGkWHQEREAYrJCBERecQh0qQHJiNERKQrzgZP7TEZISJLOniuTHQIROQjJiOkK4ljHAyvtNqhyXr1fjja6WI+9dioWApQe0xGiMhFjaNJdAhkIDY+Ypd0wGSEiIiIhGIyQrpix3wic9FiNA3rWqg9JiNEREQGkZxTLjoEIZiMkK4uVGnTOZKIyAqySmtEhyAEkxHS1c6TxaJDICIig2EyQkREREIxGSEinbDbIhG5x2TET3xuAxERkX+YjPhh35lS3Prqt9jFfhBEsnAeLSJqi8mIHx7+4BBKqx347dpE0aEQqcbG5hQi0hmTEaI2WlokzP4kGW/FnBYdChFRwGAyQtTG4fMXse1oId75NlN0KESWxWY6ao/JCFEbjqYW0SEQEQUcJiNEREQkFJMRspwdx4tw26u7cOhcmehQqA1WzZsTJy8gPchORhISEnDfffchIiICNpsNW7Zs8bp8XFwcbDZbh5/09HSlMRN5NWtdEkqqHHhoxUHRoRARkQ+6yX1DTU0NxowZg8cffxwPPvigz+/LyMhASEiI8+8rr7xS7qYN4+yFajy9PkV0GLKsO5gNAPj1+CG6bvebE0X46EC2rtskIvWwQov0IDsZmTFjBmbMmCF7Q2FhYbj88stlv8+I5n2aihMFdtFh+Kza0YS/bDkOAPjJzREI6dldt23//qMk3balBs6oS0SkP936jIwdOxbh4eGYOnUqYmNjvS7rcDhgt9tdfoykqr5JdAiyNLQZIdLA0SLUCfbtICK9aZ6MhIeHY8WKFYiOjsamTZswfPhwTJ06FQkJCR7fs2jRIoSGhjp/Bg8erHWYREREJIjsZhq5hg8fjuHDhzv/njBhAnJzc7F48WJMmjTJ7XsWLlyIefPmOf+22+1MSAwqr7wWQd264srgINGhkImwNcw8+FWRHoQM7R0/fjwyMz3PcBkUFISQkBCXHzKeyrpG3PlGLG59dZfoUMgE2PpDrfj8I2pPSDKSkpKC8PBwEZsmFeWU1YoOgYiILEB2M011dTXOnDnj/DsrKwupqano27cvrr76aixcuBD5+flYu3YtAGDJkiW45pprMHLkSDQ0NGDdunWIjo5GdHS0ep+CfMbqce+4e4iI9Cc7GUlMTMSUKVOcf7f27XjsscewZs0aFBYWIicnx/n/hoYGzJ8/H/n5+ejVqxdGjhyJbdu2YebMmSqEL4bZLlisECVJklBW04D+fdi3hy6Nqlt74DwmXnclhg8M1n37kulKUdKa7GRk8uTJXudiWLNmjcvfCxYswIIFC2QHRupp1qE65HDWRdwW2Vfz7ZAyf/3iONYdzMG7vxqLe0dHiA6HBFuzPwuvbU8HcArnX7/H67K8mSE98Nk0ASDh9AXNt3E0r0LzbZBy6w5eqq1c/E2G4EjICNLyKoVunx1YqT0mIwHAwYnOiIjIwJiMELXFpmydcEcT0fc0n/SMjKW149jHh7Kx4XCu4GgokNg4zzwRecBkJAC4uwS8sPm43+stq3H4vQ4iMjbWYZEe2EyjAJ/seklFbaOm6z9TUo3opDzuby9aWiTkXjTj5HOsJRFK8CnFSjJqj8kI6a7a0YT7l+7De3FnvC5391vxeG5jGramFegUmfnM3ZCCif+MxRep+aJDsbTSagfOlFSJDoPIspiMkGJKJy76cP95pOVW4J87fBtmmpYrdhiikW07WggAWBZ3VnAk1jbulV24+60EPgJBJazspPaYjJAq5HROlDvUWM/ZGjkzJHmTFoDz6bBFxTqaW4xbvjEZCTTGPRY94l0UEZH/ThXaRYfgEZMRCzqeX4lZHyXh7IVq0aEQecAMM5CxAyu1x6G9FvSTd/eiRQJOFFZiz4Ifig6HCEBgVvfnV9Thqst7iQ6jAzZHktGwZkQBo5/Grc2CuRfrxAaiEg7t1RfvWtXzg9d348DZMtFhEBkek5EAo9Vlndcvc7DyLKhGTVo/S+RMx2QMRj79mYwEAKMdgHLDsdc3aRKHOwa9nqnGqBdsNZwrrdF8G0Y7l/Rg3SOGjITJCBne5hRO6EWdM/KwRXIViEkdecdkJMCoeWOcWczROkQkn4Ur6AzNZuAGdSYjAUCLA/BiTQPe46yfRKYkJxkw7uWLrITJCCmS0+7hbLzRITm0uDMOxItmS4uE5z5Lw4f7z4sOhcgvTEYUMFsVI+cUICOwQj8Bo1Vz704vQXRyHl7cekJ0KER+YTJCujPyRclsiSbpS0lir+XhXu3Qb6SZmoxcBliZkfc7k5E2iirrkZZbIToMU9LzGD+aV4Edx4t03KLxpRfx8fZEZF6cDr6N8Yu+BQB88+wkDB8YLDgabajVZCNyvoqfvLsPAPD1MxMxIjxEWBxkLBzZS2RerBlxw2q1I0Zr51bLeR0muSL5JEnC1rQCnC7Wt7amrNqh6/aIzMbIzTSsGaGA1FqzY+Xp0UWJO30BT69PAQCcf/0ewdEQkRmwZoRUYaZruiRJeOi/B/HoqsOWnh7dHT2Sr5MFds23IYrRahk5Uo6sgskIKVJsrxcdgmL5FXU4fP4i9mSWorah2eV/LNo7H1H01s4M/Hz5ftQ3Nntf0Ns2FL/TC2PlCd8zalzUqcq6RnyZVuDXsU6+YTKiAO9GgLKaBsXvNdLdpZlqdIzind1ncOR8ObamFch6H/c1mc2Ta45g7voUvGSReVyMVPa2x2TEHeN+X36T2yrR0iIhq7Qm4JozrErN77GxuUXxey18inVkwFOniUOPfJKYXQ6AD+vUA5MR8uqFLccwZXEcVu0773U5ERcXFqe+ETnK5GhehbBtk2cxJ4tFh0AK7DxRhJV7s0SHoQkmI+TV+sO5AIAlMae9LicnMWAzl76iXtklbNu/WX1Et23pUQXNpiZ1+PpdVdY2YmnsGeS2exZWoPp/HyXh5a9O4lhepaL3tz1+jVbbzWSEiDRTXa/fdOWBmOQa7HqiuoWbj+Jf32Tg/qX7Ol12T+YFvLb9lF/Nh2Zxodq/AQSfJebi1ld3KU5qtMBkhBSxaiFotLsFM2pqtu4+VHR4sDZFsf1nywAAF33oMP/IysNYkXAO6w/nqB6H1WrEFnx+FKXVDXhmQ4roUJyYjLjR2XFn5B7JemlfJsvZI9x/1vbxoexOl9EiXeFxZR5a1mLll9epvk6r3KO0P0NaDPTBmIy40dnEUGauDlYrcjMX+wY6/0xlyS7v/YZalVRxWnYr4SzF+qt2NOG5z9IQm1EiOhTdMBkJBCxLPOLdtO+W7Mp0+bvcj7lm9JJTVotqh379VqxIi6ZLnnfe/efbTEQn5+FxDx3Arbj/mIwEGCv1ibDQRzGluIwLokPo1KR/xeK2V9UbTcRKAtJDQaV5Z7hWislIANCi/LRSHpBZXIV/x5xGVX2j6FA0Z+Uqd08frbahGU0WHWHBhFwfVjltjPw5+NReNwz8fSmiSXlloVLwR/9OAACUVNVj8vAwwdFYS4NBkoA/RR/Dm78YI2Tb7qrU88prUeNoxvCBwQIiIjIe1owEAuvkDZpKyzXOmHtSV3RynugQXNz5RiymL0nABXb29ajEXo+KWuvXVrqjV3O6kS4NTEYCjBEqNIxaVXiy0C46hIChxXHYJ8h8Fb3ZZTWiQzCk4/mVuO21bxW915c5SUxPcRlq0MIXTEYUMcIFXRYVjr+q9iMSjJpRkCloMTw+pGd31dfZHo96lXSyIz8+pHziso1JxqoFI98wGSFl2mVkVu4YaSUiR1NZcTgiKWS2GzrSHJMRN3hdJeocryfi6fEd8EbDSox71jIZIVWYdf6S9uWsST8G6YSHB+mhs+PMiukhkxE3eCNAgYwJGbVl1hsNNbGJUXtMRkiRDg/Kk5HB8bS2hkD8HrX8zJIkYU/mBRRWqv+gN8Mx2cFj5ueRmYX5xsKRbJrMwMpzkwzG7DWa8acv4DffPYvk/Ov3CI6GAoGRynHWjChgpC+Q5OP3RyK1bfZoO+nZgXNlIsIxJLMnllqzYqdiJiNuWK19kNdeamWUQowJ4SWVdYE5w6jZ6H5NCMDzg8kIqULENU6bdtwALAX00uYYiU0vERcHERkOkxEi0t2mlHzRIejGIJVRRIbGZIQU4XA//RVW1uHe/+zBZ4m5okPRRbG9HvvOlIoOw4WWiYVRm4eN0rQnEkfTaI/JSIBhDmHeguWVbadwPN+OBZ8fFR0KAO0btG5/7Vs8/MEhxJ++oPGWxDHnkUha66yMsmJ6yGTEDavdCCj9OHIeb27WXWam77q2/cMKA8T+s8aqHVETbw7cM9ppadRaKythMkIeWfmOlDzztebIaMVzIF7Y9WguZZOsdRmpllh2MpKQkID77rsPERERsNls2LJlS6fviY+PR1RUFHr27ImhQ4di+fLlSmIlAzFCO7IWZaSZy90tPnQK5YXFuAxwSukmgD4q+Uh2MlJTU4MxY8bg3Xff9Wn5rKwszJw5ExMnTkRKSgqef/55PP3004iOjpYdrFY2p+Thzjd2iw7DVNpf1ERc4nYbdHhoeU2DkO0++2mqkO2Sd+0vvE0t358tRrozJfNQnrgaNw2UPR38jBkzMGPGDJ+XX758Oa6++mosWbIEADBixAgkJiZi8eLFePDBB+VuXhN/+DRNdAi6MUvhZ6/vfDKouAxlyYiWlQPv7s7E4p2n8epPR+Hh24eoum4j1Eb5w9zRq6ewot75e31ji8BIyFd6n3qBWIGpeZ+RAwcOYNq0aS6vTZ8+HYmJiWhsdH/BcTgcsNvtLj96Mnuh742/B3lcRgneTzjnU0rjqYags927NPYMRr+0U35wCqjdMW3xztMAgBc2H1d1vUZktPLS7KetycN3sTklD3/ZcgzNLUY7SpQJxORAb5o/KK+oqAgDBgxweW3AgAFoampCaWkpwsPDO7xn0aJF+Pvf/651aKRA64O8pt04oJMlgRlv71G0jX99k6HofaSO3kHGen4mRzKoKzmnXNbySm7OWmubJwztj3tGdyzjidrTZTRN+4O5tb+Bp4N84cKFqKysdP7k5gbGJE9msvNkcafLFNnrO13GaKx+A+TLhaVbF1781aTlMaVk3av3nVc7DI8u1irrP2X2Wi6ST/NboIEDB6KoqMjltZKSEnTr1g39+vVz+56goCAEBQVpHRqZnATgiTVHENStC5b9OsrtMvWNzejZvWuH933/u3nSD6OVz+OH9hUdgim1PeaM9p0GMkmScLGmAf36dLz2GC05smJtoeY1IxMmTEBMTIzLazt37sS4cePQvXt3rTevCQ6PNIaK2kbsTi/B18eLUNvgfkKwuoZmr+v49pQxR+S4Y7Sjbkjf3qJDEED+RcDXdxjtghdo5qxPQdQruwz3CAItGelSJjsZqa6uRmpqKlJTUwFcGrqbmpqKnJwcAJeaWB599FHn8rNmzUJ2djbmzZuHU6dOYdWqVVi5ciXmz5+vzifQAMsEbTULHkDQ9vvdHEAPbHNHkiQ0NCn8QniikIVsO1oIAFgef1ZwJIFJdjKSmJiIsWPHYuzYsQCAefPmYezYsfjb3/4GACgsLHQmJgAQGRmJ7du3Iy4uDjfffDNefvllvPPOO4YZ1hto5CTCWtUA7ThR1PlCpIvnPkvD9X/5GrkXa0WHEjDUOq0qaxvx9bFCBduXFwBrgkkPsvuMTJ482evBuWbNmg6v3XXXXUhOTpa7KVN5/et0/OjGAYgacgVyL9aiqUVCZP9ArMbunL2u8zlE9KJmOVtigA67248VYuqIMAR169r5wgA2fVcz9NHBbDw/c4SWoSkeMm+15ou2h5w/n+2x1YeRmlvhbzgdXKhy4JGVh1Rfb1tWnj5BD/7svtPFVeoFoiI+m0Yly+PP4sFl+9HU3IKJ/4zFlMVxqAnQB5uJkCDjOTre8g9/kpMpi+OUv1klT32cjDe/m+tEFLkFZVNzS0DdfXv6rHI7JWqRiADAv3edRnqRthes5hbvTYNW7KAph5anwwubj2m3cj8wGXHDn6zT0ab9/aKgacHbC4RiftY6pTVv6u2dmk46y+rly7QC0SH4zNHUjB+8sRsPrTgoOhSfBMINfX2jesexp921y0Qdx0kfxprdyCACPSsn8ocNvqd4KTkVKLY7UGx3aBmSaqxQgSP3I2jRpKK40zRZFmtGFLBAeSSUnneXnX1XZri4VNU3BlQzhlwfH8rGnW/sRlZpjehQ3PL1eA+EWhdyT26TmxUPFdaMBBg5FzU5d0RWPDmMIKOoCtOXJPg0/b4vrJjUtD4H6K9fqP88oNqGJjz8gbadOclY9mTqP8/IA0v36b5No2HNSADIKdN+2KacS5yIO8CSqno8syEFh7Mu4j+7M/UPQKE1+88DuDT9vs+TZ2kWjRi+fp6mZvUTrU8O5SAlp0L19RJ5Y6aZodXCmhGVqTVsT03vxp4RHYKLWof+HT3/uuU4vjlRjC9SvXfuDMRCQKmcslokZZcjasgVokPRjEODvg1qHGHn/WiSMkix5JVRyk7SD2tG3BBxIrS0SGgSPTWpTsoEjDLKuVin+zbVpsZFzFPTm5JD/t3YM3hw2f4ONW9qbsNo1Cgb2raUtV2dnFX/38fK521ium1+9U3KbuiMnOQxGVFAi+/zf5btx6R/xpq2l7mBj3EAwKlCu9vXWTBf4s9+OHPBdU4KWf2S/NhuS4uEb091/vRo4/H/qCusNEdybeSLnxxG+xhPrEns9LlbZsNkxCBScytQUFlvqNnx5FxUeFEn4NIstC06HQybUvLx5IeJzr/lHK/1jc2Y9VESPk/K0yI0F96mCjDiTKTt47Vip2ej82WXZxjoWqEGJiMKmPnUNHPsSrAg1dd7cfo9ZCwuQ/nEWR/uP48dJ4owf2OaihG5x35IJMp/vu3YWd+o82gxGXHDn6+q7cXPiHc9WjHrNV+rZMWoX71R5+LwRot9WSHw+UgeDzmVP6gkSe5ngfaw/craRrfNxIFUjlnNmzGuj4aQJOMmx0xG3DDSuVdZ14g/bkzDvjP6j32Xw0j7zB9mTarU0qigE7WV91n749pMh/lzG9Nwy8sxiPfhuU0lVfUY84+dmPpWnPaBkWEY6dxlMqIyl6G9Kqxv8TcZ2JiUx4mXNGKgc1GoytpGLP4mA9e98DUOnC3Tbbv+3HVb4bvTshlxU/KlJzIv3d350P49py/d7OQaZNSZmZI+UgfnGTG4vHJ1JywzUiZM2lBygf80Mdf5+4Loo2qG42SF2jM1Th9V1sHz2NIC8etlMkIBjYW6G37sE10TjnZxin58wb4zpQgLDnKzLd9G01ggV+vAqJ0lrcBqe5bJiBt1fjxC2/gXN20CNP7nNj+jFj555cao2hcps7jK56bUtufK5Zd11ygiL9vvpAzYZcq5W7TFTrzaY58RN9YfyvX6f18vvDx+jU+r70jkV59fUYeFm45iya7TnS+sghe3nsCm5EvzdWhxJ+xpnQ6Fs1DKkV3qWzOpnDkf2iYDkf17y44JUP+4bVukVQocaWRURp8i4ExJNR5Yug+x6cqHu4vGZMQNPQo5f9Q2NGFvZqmikQ+BprMipH0ZY+QiR05s6w/nYsku/R4IuFTh84/8uajuOqW84PVlu5tT8lz60gDqJ5ltEy3Nbl46jAgSf5cUHMRKeW/k5j5z16cgNbcCj685ok1AOmAyooCW47Q7K5C2phXgxr99g1+vPIS3dbzYaKWlRcKSXadx/9J9lpve2KjSi9xPje+k8FolSZLhawPrG5t9vstdkZClcTT6OXSuDHM+SUaJvV50KACAH40c4PX/bBaRp6LW9+d9GSEZdYfJiALFdoewbT+9PsX5+yeHc5y/f5Gaj1/89wBKqoxR2PjKZgOW7MpEWm4FNiZ5bx7TgpFrQrTy4yV7VF9neW0jJv4zFiv3+n4Bb9Fr3vjvnCmpxg1/3YEFn2szWkhPcu+cH1pxEF8dLcTzm49rE5BMXZhs+M1qu5DJiEU8syEVh7MuYtH2dNXW+UcZhbYabareHhJY70enYpLHWxnn6Xu+WNMguyPrPf/Zi2YdE5IVCZemqt+ow/NovBHZ/SCvvNawM3Aamd41NVZLNHzBZMQdf44El8eD+39EyS02quqt2fls3Cu7dNlOAJYBsvhzIW1/PpwqtOPshWo/I/peZwlxF5dhtPp90+2Lk7Zhtv2fVjH5stZAeyyCFSg9XoyajDIZcUPk+aN1IWnwTuEeVTuaNFmvdoWweUtho8XuazjFVd6bT9X4WEbbN2oorKxTfHmKzSjBo6sOo7DSHMO7y2saMO+zVNmzDOs9msas5bQ/mIxoSEm5ZcGyzlQCsAzQ3LMbUvDpkRy3/1Mz+e6syUduImGVU7Htx3Z3kWtqVn7UP776CBJOX8Dzm44pXoeeXt52EpuS8/G/7x8UHYoQRr6+MBlRmVGrwIjUoOTo3pJagD9Fi79YdTFIQaxGGSHnopJd5jpXihZ33aXVvo/mECn3orLHaxixRsxmuzQNxbxPU1FYaa6BC+4wGXHDgMedW0YK06gpmJWqO5V830o6/prl+JdL7pwe9W7mG1JSXd/+LRW13/frutCmaUmr/e7ThcpC50kgWX8oB5tS8kWHoQomIypz6Zym8bbKatzdjYi/krRm62ZgxDseNd3ycozs9ygZTWMGcmtGzl2o0SSOtrtQaS2Jib8GUpH7a4A5MRlR2bqD2aJDEKK5RXJeqDYczrVMtm52tZxIzknLxNNbcuBts0+sSZS9rRUJZxVP2e4u+WlQYSZnT5/R01PHmUx1Rp8dZKSbCyYjKnszxr/ngWh9o97oR2c1b17ZdgoPrbjUKaysWtykcEbR/muUJMnvxwwk55T79X41qH30+HK8q3VKGH00ja9rfs2PuYTcXXumvhmvaF1NbZKYxmYJGxNzOyQf9jptRsHpzUr1p5yBlYR4bfspl7/dtYOr5XDWRc3WrRe17hTaX7N+s/oIRvx1h1+JWnmtueeQcXcd17NYdJ1nRByrdHKftiTB+fupQjv++PlRTP5XnE/vFdU6qnxuDnGe2ZCCzSnyJ+rLKtWmmVErTEYspv1JviLhnO4xWKOoVVf86QtokYCvjxeJDqVTevaj0fNYMeb9oDEoSZDc9alp0nmK/0DwRWoB/vBpmuz37fBQ1jS1GPMBq0xG3FCt0PJxRd7uxuXeqJu5wH39a/+nss+vkDf5UrVFZ6yljroYZWyvQJ6KEwN1HQBg3RFdvtLy+0jOqdBu5X5gMuKGlUdYGPmT+XtXtSfzAn7w+m6X1zq74zPqidmZfWdKNVu399E0mm1Wc0Y+9gFrXoCt0iRlwa/GcJiMuNH+wGtpM1JE+23zsFfqowPGGcnU2lG42tHkcuyodcH5XNDD3tS+uOh5tLvcZCj8IpS8q/2mRD+x1swJJX3PaskrkxE32p6rTc0tmPnOHjy+5ojs9dQZcFhli6R9XP4UdnsyL6gXCIAv0wpVXZ8ch86VYdSL3xjmse2+0rOQ82lbKgVklFaant27ig6BLMAGm6USy26iAzC6EwV2pBdVIb2oSvZ7s0prMKRfbw2iUu6JNUcUz1HgK3/unh9ZeVjFSICXvzqJJ++MVHWdvvr3rkvDvNcf/v65LGav+TJz4WecO0lxO9FTDa9WEZn9eG9lhMNei9p5I3yuVqwZcUO9eQ18W5OeBbw/iYi/82S409nDzazGOBdEY9DzYqVl84ico9jT+W7UvmpaRGXmpFYPgbh7mIy4oVaZoCST9XfbWpZnWowIu/HFb9RfKflJm4NI9KW27fbTcitEhWHIC02BzFFovlJaHhmtRsVY0fjGoLmtR0xG3HB9doS+OszcqfP21SAnB2toMuaY91blNQ249z97sGpvluhQdGO2QsxnbT5Y/Gl1+yaZhQT3zaj/2X3G63sUb89gBZhVRveopbCyHrvTi0WHAYDJiCEcy6/0+L8EAxWagXYiJ+eU4724Mzieb8c/vjrpfF2SJHxzokj2nCaAse6w3vk2U3QIxtohfvBeC2qcD3nuQg1iM8SXKSIS3g2Hc3DkvPhHKqhFrX2o5PlIWmAy4oZaX7KvbcCPrlK306ZVtejcv+R/3tuPusbv+8lUfTdB2pdHC/H7j5I6zGnSnrvrk2rHlgrreEvBc5RE3Oka51KuDtG1BUaojfw8KQ/1jfqONvzzpmO6bs8f7pJbNY4bI59LTEbcaNteqcf8ImqObvGnrXXbUe/DYH3dFVrtsW3HvMfn8FDIyvkOvS1aWFkPADhwtszn9clV4/DtwWKBVUdldq7flpGelKpEc4uE/8af9Xs978V6bhrSQykf6GkoTEYsxp8779mfJCO9yO7x/74WoVqVtReqHF6brTz1A+isBsMbtT9LZ8likb1e3Q0qYOS7p1ani+UPtSflXtp6Aq9uu9RUuTExF4t8fHSDt/IoQ/B3OO6VXUK37y8j1HCpicmIhkSMpvHXj5fsERuAFwUVdYqatAoq1bnAm/yGVhUiZmB9+9tMlLRL0qb9O8HD0sbj63Ej+tz35EJVPdbsP4/392ShtqFJVhLh7bPzfPLM3bxWde2atfafLbNUPz4mI+Qz0XOCFFRqM/xQVwa94LRVVe9bU5EafO1XtSD6qMaR+EaNhMGXs+jshWpkKJhoUQtNzd9HHGDTAglT6OYGamS7aRCUHItG/vo4A6s7bb5kf748Ec+gqPaxz4ESM9/2rdbEStm6u0/i69fq9r2dvMcIuUr7O7C2/Lqb9ePDuXtcvVmVVTd0uszUN+MBAMf/Ph19gr4vpv/x5UlPb9FM++NdzjFg1Noe2Uz4OcwWMmtG3DDbl9jWnkztnuaqZCirlVgpyVLqQhU7/flr9ifJPi97sV3ismqf2PlulsWp1+mUZxO1xWTEDZ4kgUvrhMNbs8SpQnuHtmJP/Y5qHE3YlKz/k3ubWE+vGaPNOurO0lj/R9EQucNmmk7o38nK+AWSKCI6vLmdK0SD7dQ3NmOGj81gALDzZDF2ntR/5sTsMus0l+jF7B015SZJxT6OCJO7XyRJMuzze0RQc09UO5pcmgNFYM2IGzzc/aRR4SumUNdn9Iin+UWMVvY++aH2szVW1Dbg/qX7VF+vp1356raT+CI137d1GOz70EP72sLORgne/tq3msRxNM/zTNXkn/EafWdysGbERMwyWZJWURqhz8a/Y07j40M5vi2swgysJvnKVfVe3NkOD7LTMgl4f8+lfhj333xVp8sq+T78OW6NkPyImELd3edmE2FHajXtaTnwwVesGXHD9UT4/gSw1zeipErbSakcTZ5HMvzq/UMdXvu6k1lJST3H8+14283zXNKL7Hh+szZTTVs9GXFXlNY16DtNuBHEppdgzX5zPIxRvUPS85rUnJWazIE1IzKMfmmn5tu45529iJs/Gdf0793hfwfOdZyGfNW+LMy4KVzzuOTQqgYnu6xWk/V60/ajzN+Y5nYZI08UZyTu7uIu1nQ+zNXsfLl7fXzNER0iMY/kHHe1MRbPzFVmhFo1OVgzYkBr9p8XHYIh2XW4W9K6JsJdAVHX0IxmDxs2QtOUWp77LLXDa69uP+X83V5vzbthszfTaMXqtX56sFL5wJoRAyoTfLeYmluBuIwSoTG4Y4XTrv1dcmVdI8b8fSeuDA4SFJF+OpuWf/bHyfjoydt1isa74/nqd5asb2zGuoPZqq+3rdJqB/64MQ2/vO1qzbYRCEnEioSzmj4QkzpSVDPy3nvvITIyEj179kRUVBT27PFcTR0XFwebzdbhJz3dtwctieD61F79t/9lWgHKfUxItIjvgaX7sGRXx74RgUj1B+W1u9NtLfA8TSYWCAV/q9YJ+4xQG+Cp2UTRFNzffYfL4s7ilW2nvC/sp0Xb0xGbcQG//yhJ0+2owciH9mvf7UclWlok/GL5ATz1sbbfgRHOEzXJTkY+/fRTPPvss3jhhReQkpKCiRMnYsaMGcjJ8T7CICMjA4WFhc6f6667TnHQWlPrS/ZnPScKPD8918gkSdLsAiriwtzYbK0nY5JvKmrVr5103w9CXReqrTFDrrtHaWhx/vs6J4ocGcVVOHz+IrYfK1J93Vo6e6Fa6PZlJyNvvfUWnnzySfz2t7/FiBEjsGTJEgwePBjLli3z+r6wsDAMHDjQ+dO1a1fFQdP3zpfVoMpgbe1GvuPpTPsZUL9IKxAUSeAyY22QVjEbcZIvPfop6PWpM4t9uwBX1Tdh1kdJPt2cGPn49bZf92r4KBFfyEpGGhoakJSUhGnTprm8Pm3aNOzfv9/re8eOHYvw8HBMnToVsbGxXpd1OByw2+0uP6IY+LgCAJRWN2DCot2iw9CFHoXgioRzLn+LflKx0Y8/6lyFHx2vjZeKKL/Yth9lZ4R5k369suN0CZ7sOFGEL1J5c6IVWclIaWkpmpubMWDAAJfXBwwYgKIi91VS4eHhWLFiBaKjo7Fp0yYMHz4cU6dORUJCgsftLFq0CKGhoc6fwYMHywnTby0qnSR6PWvCCBPW6MEAZVcH/heo3t9vxM+sNQNWBvglNv1SZ3BP36WaF2UjXODb8jWcYns9NhzOQf13T4x2dwwY4ZN5milZhM6uL+7+v+6gjxM2CqBoNE37qkNvzwwYPnw4hg8f7vx7woQJyM3NxeLFizFp0iS371m4cCHmzZvn/Ntut+uakOReFP90WrMWyJKk3QXUCIVRe6cKqzpfqA0jVrubgRp7Tc6u93QMK7nBaGqR8M0JZf0HrHS4tN+nbf/8ybt7UWx34HRxNf523426xmVEt0X2xeGsi6qvV/RTn72RVTPSv39/dO3atUMtSElJSYfaEm/Gjx+PzEzPozWCgoIQEhLi8qMnoxQAp4vlXehIf97akN01K8k9tKw0j0Ag+/1HSR6/S4NVZnRKrXAr6xqxO70YTc0tKLZf6nhrxCkF2vKl5kmN60f3rp2vRO3rlOjrnqxkpEePHoiKikJMTIzL6zExMbjjjjt8Xk9KSgrCw401a6gnogoKG4D9Z8R2KCI9GCTzNRAj7xGPCQWTRre8XeBScirwxJpE/LddPy3A/WgaI6t2NKnaROZo9K2jrNkSWW9kN9PMmzcPjzzyCMaNG4cJEyZgxYoVyMnJwaxZswBcamLJz8/H2rVrAQBLlizBNddcg5EjR6KhoQHr1q1DdHQ0oqOj1f0kVmOuc9GFVgWzp7k4SL5AmIbdaDz2GfHyHjnNQkYcht42ek+f80sfR6wZ9cJ7pqQKd7+VgKk3hGHlb25VJc4GFb7LtLwK/wPRkexk5KGHHkJZWRn+8Y9/oLCwEKNGjcL27dsxZMgQAEBhYaHLnCMNDQ2YP38+8vPz0atXL4wcORLbtm3DzJkz1fsUFuTrkDOjkWDcQkMLcj9qbYO8DnC+Tn4nV6lF5qOwAm931HIqCLak5KsQTef0OL9LdL7xaPLj4t/aKfTbdPWamHzZx96OjWJ7Pb46Ku8hqqLvfxV1YH3qqafw1FNPuf3fmjVrXP5esGABFixYoGQzhiCqd/qLW0/gRXbkMjV30/pvOJKLh25tO1W39+NL6SyQZvXDN+Nw7kKN6DBM2ehSq9vTjn3fO3L3Y0NzCyRJ0rXm7lheJX763j6flvXl8xihhUnRQ0UFB85n07hhhIOJzMHboeLuoip42hLDE5mIfHYkFwuijyK0V3eP88voNVxfCaMN623Pl/jyyuvwf+uSdYjmey9uPY4mH09MvXZxIPZB4lN73VCrwGFSQ2QeC6KPArg0ykMLni5kJws9T+poxCJE/ec1uX7KHQqHQSuVnFOh2bqVJogGzys1wWSkEyKPCTMekEa/OxOt48VFzOWmsJMn6JL6PN3t/uRdz00EueW+z3lk9DPPU3x1MvtRiWS0fazmDNGiE18mIxr66xfHRYdABtNF9Bn/ncdWHRYdgmJ1uvWNEC8tt0LW8ka7F0gvqkLuRe/9F84r6d9gYGrUiPv6Pe48Wezh/QY7EHzAZMQNtdrrjNARj4yl4wys5is0RGi736ocxnowpJbkzhqrR18DOde5w1kXMfGf3p9F5vt2jXmuaNEc7+snbZ0+Xw2iuxUEfDKSlN3xsd5G7qRmBkYtNMgiFB5eqpzXfqxCyWkhe8ZeA596Ro7NV76UbW0X0fozW2Gftgr4ZOQX/z3Q4bWci7V4ZkMKAGt92XqQwHt9b5jm+k/p8SV6hMIhBc8akfMso2pHE/afLZO9DblE70ej0eIa4esNnZrfheib8IBPRjx1ANLrUdH5FeIfykfKHTinfeFProVzIN0gyKk6335M3iRXSind/4GSxOjZ3OEpgTDjng74ZES0H7y+W3QIqguki8UuDx3IPBHdLmsFQi9qnqZ01ygkOYfLmRJzztrsq/a7uLlFQmx6iWazFBudzWazVHnCZKQTgZLNkz4amoz3/BAzaNtcofTCr2hWStFkXG18nbjLKtbsP4/H1xzB/Us7Do3WKjGTe+wp/UbSi/R/YrvoxIbJiIEFVtFiTnKHJablVbr8HUi1SGpRusu26dSMoSajDAVvS3GfHZWO9Xe+zUR0Uh5e/uokgEt9/Np7en2KOhtrJ7OkY5Ig8iKuZvkh+lDjdPCkqqIAm0zL3wfOeZongDxTOlpLlVopnUts0Z0KjUSSgOP5lXgr5nSny2o1/PuzxDz882djvC6j53dWZO9Y3u4/W4o8GZPlGQWTEQMz4xDZX688hInX9Rcdhmm4u9OijrJKv5+zp77ReE1dJjxVTami1rckI5CTuF+9f0h0CIqwmcbA3M2BYnTZZbUsmEkTJwouNXEFUr8b0e347pjh/BbadIK2I79MsLO+88nhHKHbZzKisRMFlYoLz6p68zyzgUhrredDi4kKeH8Zsc9ISq64myRfk4wuRsziVKb2Jzzarj+b3piMaOyed/bid2sTFb1375lSlaMhIjMxYnOD0sdc+JtDyhnZKHKvGfE7MwMmI51R4SYs/vQF/1diIoFz30qBRvfLDK9rLnzeHdxvpsMOrAKZqT2RtHG+1IRzXwgWSKcNr6nfq3E0o87Dg+HaPzCu7X5raZGwOSVfw8g80/JQ/cd3Q5utgsmIQLtOlYgOgQRzNKn31M1AkV5kFx2CbqzU98HfCSS9NXfPWpfk8vfZNk1JW1Lz8dzGNL+2TdpjM40Xp4ur8KsPtBsmVexmjLgVBNJoB9LfmQvGm/ZcqzvgvWdKXYY1k3txGZ6bwpNz9O1w2zZ/HP/at9iTGVjN9EoxGfHijxpn01atbf48KU90CKbBzm4KWPXEcWNzSj6mLI4THYap6d2s13Z7ZTUNeGTlYX0DMCkmI140Nqt3FLvtHxJIjd/kHnMRn7XuKqOcNVmlNSipsmbtptqK7fXCijutJ8njDYU6mIzo5Kfv7QdwqY/AvM9S8WVageCIyAhYjPlOwqXOiE0q3iQoVVJVjymL43Dbq9+KDsUUvkjNF5ZERiezptYMmIzoJDW3AgDw8cEcbErOx9z1KYa5wyNxLNQ/UXMtkoR7/7MXq/ZliQ4Fp4uM12/FyGywcfQgecVkRGcX/HywGlkLy2ffXaxpwMlCY4yk8XdkiFxNzebuFG7lpFvvY8GqmIzorO3FhxciIt8ZoW1e1FOpP03MFbJdtdisnI2QKpiMeKHFHBBmfYgSEQF/3nQMgP43EjkXzT05ng3G6XisltaJ1oyQJFsBkxEvzip8BoM3H+wR395NxmGlSa20llFcJToEp7YXVkdTs/aT15n8Sm7EB/754+fL9+OGv+7weQK+p9en4ItUMbPAmgVnYNVZc0ubmhGBcZAxMBfx3TvfZooOAQBQWdfoUqs5/C87BEZjDjabzTLN0knZF3Hk/KWJ1JbGnsWVfYI6fc/WtAJsTSvA/Tdf1eF/7aeyD1SsGSEikkPijYRcVkq6F3x+1Ov/5XZo/dX7B/0JxzKYjAhklTsFUs7Tg7/IuKocjbpnI2YvKkTlIlr0y1O7+T45p0LV9ZkVkxGBzF7AkH9EzkpJyt35Rqxz3iA9mbnPgc1mE1LgiTi/2KFVGSYjRIKY+eIS6D4+lK37Nt/fc073baqli82GL4/qO+u0vb5R8/xHkqQOTVCcd0QZdmAViEN7icyJp648Jwsrse5gjq7bzCmrxYjwEM23o/RY+PZUMfadKVM3GBNjMkIkSGw6Hy1uVg1N+s6IavYbF70TkVZbUvStfVzweRqm3TjQp2Wf/DBR42jMhcmIjsw+pTOp68A53hWZVZWjSXQI1InG5hY8tzFN121+lpiHzxI9P5ivsrYRoZd11zEi82CfER3tO+t68TH5zQ4R6YjlhTxt53Qyij9Fex8WHMhYM6Kj5hbXmhF2dCIi0oYeuchXRwtlLb/vTKlGkajjwNkyTBjWT8i2WTMiUFU9q3qJiAKF0Zv3RD4Vm8mIjuobXWtGlsaeERQJEZnJ+3uyUFHbKDoMUzF7p18RRO4zNtPo6KmPk13+NmCTJumguUXCmZJq0WGQyeRX1IkOwVQu1jSIDoFkYDJCpLM/bkzDJp2HHBIFmuc3HxMdglvT/h2Pp6deJzoMt0RWJrGZhkhnTESItFdu0Gat08XVmPNJiugw3GoRmI0wGSEiIiJsTPI8R4rWmIwQERGR0L5sTEaIiIhIqIBORvLKa0WHQEREFPACOhl5cg0fVERERCRaQCcjGcVVokMgIiIKeAGdjBAREZF4TEaIiIhIKCYjREREJBSTESIiIhKKyQgREREJxWSEiIiIhFKUjLz33nuIjIxEz549ERUVhT179nhdPj4+HlFRUejZsyeGDh2K5cuXKwqWiIiIrEd2MvLpp5/i2WefxQsvvICUlBRMnDgRM2bMQE5Ojtvls7KyMHPmTEycOBEpKSl4/vnn8fTTTyM6Otrv4ImIiMj8bJIk75nBt99+O2655RYsW7bM+dqIESPwwAMPYNGiRR2W/9Of/oStW7fi1KlTztdmzZqFtLQ0HDhwwKdt2u12hIaGorKyEiEhIXLC9eqaP29TbV1ERERmd/71e1Rdn6/Xb1k1Iw0NDUhKSsK0adNcXp82bRr279/v9j0HDhzosPz06dORmJiIxsZGt+9xOByw2+0uP0RERGRNspKR0tJSNDc3Y8CAAS6vDxgwAEVFRW7fU1RU5Hb5pqYmlJaWun3PokWLEBoa6vwZPHiwnDCJiIhIpl3z7hK27W5K3mSz2Vz+liSpw2udLe/u9VYLFy7EvHnznH/b7XZNEpKsRTOxeGcGThTYce2VffDhgfNobP6+1eqyHl1R29AMAIgI7YmCynoAwF3XX4krg4PweVIeAODtX96MzSn5aG6RMPWGMKTkVuDF+0Zi+7FC/GXLcQDA/952Na4L64OvjxfilquvwOC+lzn/BwCPjB+C7ccK8eyPrsehc2UYdmUfXHV5L/z9yxO4MSIEE4b2w+C+l+GKy3qgrrEZq/Zl4YGbr8LdNw7A1tQCvLEjHQDw8v0jcaLAjg1Hct1+5l7du2LS9f3xzYlidOtiw8/HDcYPru2HWkczQi/rjrGDL4ejqQV/ij6K/WfLAAA3hofgZKEdQ6/sjTuG9cOQvr1RUdeApbFnMXvKMFwXFox1B7PRxWbD5Zd1xx3D+uGd3Wfw+0lDcdfwK7E3sxR55XXoE9QNa/afx6AreiG7rBZrn7wNf/viBF6+fySSsstRWFmPPZkXcKHKgWpHE0ZdFYpfjBuMrl1s+DwpD9272jBnynUoq3Egp6wWHx3Mxi1XX4Haxmb0690Dm1PynZ/zwyduw5C+l2Hy4jjna7+bGImREaHIKq1Bj25dsDT2DG4efDkmXX8lrr2yD5bHn0Vidjl+FjUINgB9e/fAdQOCcXtkXxw4W4Z1h7JxNK8SADBnyrV4N/aMc92Trr8Sj44fgtiMEnx86Pv+UyMjQjAiPAT7zpTi/puvwvL4sy7fxxWXdcewK/sgMbscABAWHIT3Hr4F467pi9yLtVgaewa/uHUwThdVoaahGTYAjqYWrNybhdJqB664rDuW/zoKoZd1R7HdgVERIfjd2kQU2x1Y8OPhmPdZGu4Y1g8Rob3waaLrMTEiPAQbfjceXx0rwMXqBrwZcxqTv/u+mlokhIf2xOhBoah2NKGwoh4fPnEbbDbgr1uOIzbjAv44fTiuDeuDA2fLsGb/eUwfOQAjI0Lxn92ZmDCsP0J6dsP0kQNRVFmPV7dfaqb9x/0jUVBRjzMl1SivbUBSdjlWPBKF9KIqrNybhW5dbBg9KBR33zgA7yecw4CQnhgZEYpV+7JcYg8P7YkX7hmBuIwLyCqtwVWX90JqbgVyLtZi3ZO3Y+Xec8gsqcYNA4ORXlSF/n2CcPeIMGxMykNDUwvuGNYfv79rKL45XoTy2kaUVjuQXmTH6eJqAMDUG8JwKOsivn5mInaeLMbLX50EAMwYNRAxJ4sxbeQA5JfXoaq+CedKa/DqT0fhhc2XzufpIwegqVnC3KnX4dcfHEK1own33xyBL1ILXD7DvaPD8dXRQgDAz6MGIb2oCkOv7I0uNhskScKW1AK89YsxCO3VHfvPliG7rAaD+16GEQNDUFnXiO5dbXjpy5N4dMIQrD2QDQC4pt9lCL2sB6rqGhHcqzvuuv5KhPTshle2nXKWM9eG9UFNQxM2J+fj2buvx5HzF3Hk/EWcKPi+BvrJOyOxcm8W/vngaLz05QnUNjTjisu645HxQ1Ba04D9Z0rxoxsH4P09WbjnpnBsO3bpc/zf5GFIL7y0H7t3teGHNwzAoawyjB4UitPF1UjKLsfmp+7A8fxKpORUYNN352zUkCvQO6gbrgvrg6bmFnTt0gUbk3Lx0n0j8eLWE/jhDWHYmlaAr+beCXtdI576JBlX9glCYWU9lj58CzKLq/DmztO4d3Q4hoX1wdLdZ1DlaHJ+nr/deyNui+yLBZ8fRVmNAy0SMO3GARh0xWV4Y0c6pt04ABOG9cPAkJ5YFn8W9Y3NOF1cjeCe3SBJQHWbdf3oxgH46dirkJxdjg/2ZiGyf29kldYAAI6+NA1PrjmCI+fLncs/8YNIrNqXhR7duiCyX2+XZ6J17WJDc0vHnhK/nzQU/004BwD4w93Xo7CyDkfOX8ToQZcjtFd3dLHZUFBRhx0nihDcsxuq6pvw07FXIeH0BTw15VpsTsnDmEGXQ/ou3te3pyO3vBYjI0Jw5Hw5bou8VL4UVtbjV7dfjU8O5eC9h2/BtWF9OsSiF1l9RhoaGnDZZZdh48aN+OlPf+p8/ZlnnkFqairi4+M7vGfSpEkYO3Ys3n77bedrmzdvxi9+8QvU1taie/funW5Xqz4jREREpB1N+oz06NEDUVFRiImJcXk9JiYGd9xxh9v3TJgwocPyO3fuxLhx43xKRIiIiMjaZA/tnTdvHj744AOsWrUKp06dwh/+8Afk5ORg1qxZAC41sTz66KPO5WfNmoXs7GzMmzcPp06dwqpVq7By5UrMnz9fvU9BREREpiW7z8hDDz2EsrIy/OMf/0BhYSFGjRqF7du3Y8iQIQCAwsJClzlHIiMjsX37dvzhD3/A0qVLERERgXfeeQcPPvigep+CiIiITEv2PCMisM8IERGR+WjSZ4SIiIhIbUxGiIiISCgmI0RERCQUkxEiIiISiskIERERCcVkhIiIiIRiMkJERERCMRkhIiIioZiMEBERkVCyp4MXoXWSWLvd3smSREREZBSt1+3OJns3RTJSVVUFABg8eLDgSIiIiEiuqqoqhIaGevy/KZ5N09LSgoKCAgQHB8Nms6m2XrvdjsGDByM3N5fPvFEJ96m6uD/Vxf2pPu5TdVltf0qShKqqKkRERKBLF889Q0xRM9KlSxcMGjRIs/WHhIRY4ks3Eu5TdXF/qov7U33cp+qy0v70ViPSih1YiYiISCgmI0RERCRUQCcjQUFBePHFFxEUFCQ6FMvgPlUX96e6uD/Vx32qrkDdn6bowEpERETWFdA1I0RERCQekxEiIiISiskIERERCcVkhIiIiIQK6GTkvffeQ2RkJHr27ImoqCjs2bNHdEi6S0hIwH333YeIiAjYbDZs2bLF5f+SJOGll15CREQEevXqhcmTJ+PEiRMuyzgcDsydOxf9+/dH79698ZOf/AR5eXkuy5SXl+ORRx5BaGgoQkND8cgjj6CiosJlmZycHNx3333o3bs3+vfvj6effhoNDQ1afGzNLFq0CLfeeiuCg4MRFhaGBx54ABkZGS7LcJ/6btmyZRg9erRzAqgJEybg66+/dv6f+9I/ixYtgs1mw7PPPut8jftUnpdeegk2m83lZ+DAgc7/c3/6SApQGzZskLp37y69//770smTJ6VnnnlG6t27t5SdnS06NF1t375deuGFF6To6GgJgLR582aX/7/++utScHCwFB0dLR07dkx66KGHpPDwcMlutzuXmTVrlnTVVVdJMTExUnJysjRlyhRpzJgxUlNTk3OZH//4x9KoUaOk/fv3S/v375dGjRol3Xvvvc7/NzU1SaNGjZKmTJkiJScnSzExMVJERIQ0Z84czfeBmqZPny6tXr1aOn78uJSamirdc8890tVXXy1VV1c7l+E+9d3WrVulbdu2SRkZGVJGRob0/PPPS927d5eOHz8uSRL3pT8OHz4sXXPNNdLo0aOlZ555xvk696k8L774ojRy5EipsLDQ+VNSUuL8P/enbwI2GbntttukWbNmubx2ww03SH/+858FRSRe+2SkpaVFGjhwoPT66687X6uvr5dCQ0Ol5cuXS5IkSRUVFVL37t2lDRs2OJfJz8+XunTpIu3YsUOSJEk6efKkBEA6ePCgc5kDBw5IAKT09HRJki4lRV26dJHy8/Ody6xfv14KCgqSKisrNfm8eigpKZEASPHx8ZIkcZ+q4YorrpA++OAD7ks/VFVVSdddd50UExMj3XXXXc5khPtUvhdffFEaM2aM2/9xf/ouIJtpGhoakJSUhGnTprm8Pm3aNOzfv19QVMaTlZWFoqIil/0UFBSEu+66y7mfkpKS0NjY6LJMREQERo0a5VzmwIEDCA0Nxe233+5cZvz48QgNDXVZZtSoUYiIiHAuM336dDgcDiQlJWn6ObVUWVkJAOjbty8A7lN/NDc3Y8OGDaipqcGECRO4L/0we/Zs3HPPPbj77rtdXuc+VSYzMxMRERGIjIzEL3/5S5w7dw4A96ccpnhQntpKS0vR3NyMAQMGuLw+YMAAFBUVCYrKeFr3hbv9lJ2d7VymR48euOKKKzos0/r+oqIihIWFdVh/WFiYyzLtt3PFFVegR48epv1OJEnCvHnzcOedd2LUqFEAuE+VOHbsGCZMmID6+nr06dMHmzdvxo033ugshLkv5dmwYQOSk5Nx5MiRDv/j8Snf7bffjrVr1+L6669HcXExXnnlFdxxxx04ceIE96cMAZmMtLLZbC5/S5LU4TVStp/aL+NueSXLmMmcOXNw9OhR7N27t8P/uE99N3z4cKSmpqKiogLR0dF47LHHEB8f7/w/96XvcnNz8cwzz2Dnzp3o2bOnx+W4T303Y8YM5+833XQTJkyYgGHDhuHDDz/E+PHjAXB/+iIgm2n69++Prl27dsgWS0pKOmSWgay1R7i3/TRw4EA0NDSgvLzc6zLFxcUd1n/hwgWXZdpvp7y8HI2Njab8TubOnYutW7ciNjYWgwYNcr7OfSpfjx49cO2112LcuHFYtGgRxowZg7fffpv7UoGkpCSUlJQgKioK3bp1Q7du3RAfH4933nkH3bp1c34W7lPlevfujZtuugmZmZk8RmUIyGSkR48eiIqKQkxMjMvrMTExuOOOOwRFZTyRkZEYOHCgy35qaGhAfHy8cz9FRUWhe/fuLssUFhbi+PHjzmUmTJiAyspKHD582LnMoUOHUFlZ6bLM8ePHUVhY6Fxm586dCAoKQlRUlKafU02SJGHOnDnYtGkTdu/ejcjISJf/c5/6T5IkOBwO7ksFpk6dimPHjiE1NdX5M27cODz88MNITU3F0KFDuU/95HA4cOrUKYSHh/MYlUO/vrLG0jq0d+XKldLJkyelZ599Vurdu7d0/vx50aHpqqqqSkpJSZFSUlIkANJbb70lpaSkOIc4v/7661JoaKi0adMm6dixY9L//u//uh2WNmjQIGnXrl1ScnKy9MMf/tDtsLTRo0dLBw4ckA4cOCDddNNNboelTZ06VUpOTpZ27dolDRo0yDTD0lr93//9nxQaGirFxcW5DPWrra11LsN96ruFCxdKCQkJUlZWlnT06FHp+eefl7p06SLt3LlTkiTuSzW0HU0jSdyncj333HNSXFycdO7cOengwYPSvffeKwUHBzuvJdyfvgnYZESSJGnp0qXSkCFDpB49eki33HKLc/hlIImNjZUAdPh57LHHJEm6NDTtxRdflAYOHCgFBQVJkyZNko4dO+ayjrq6OmnOnDlS3759pV69ekn33nuvlJOT47JMWVmZ9PDDD0vBwcFScHCw9PDDD0vl5eUuy2RnZ0v33HOP1KtXL6lv377SnDlzpPr6ei0/vurc7UsA0urVq53LcJ/67oknnnCeo1deeaU0depUZyIiSdyXamifjHCfytM6b0j37t2liIgI6X/+53+kEydOOP/P/ekbmyRJkpg6GSIiIqIA7TNCRERExsFkhIiIiIRiMkJERERCMRkhIiIioZiMEBERkVBMRoiIiEgoJiNEREQkFJMRIiIiEorJCBEREQnFZISIiIiEYjJCREREQjEZISIiIqH+Pz3t3xmYimLVAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.plot(overlap_bed.overlap_tss)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "3e6ba539-e961-40bd-a8d3-0c87e963a523",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3384"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "154490690 - 154487306"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb607c52-5ad3-4188-9667-ff2055cb2701",
   "metadata": {},
   "source": [
    "### Ensembl .gtf file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "a8da7046-65ea-45ee-be5c-87f8bed2959c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_version', 'gene_name', 'gene_source', 'gene_biotype', 'transcript_id', 'transcript_version', 'transcript_name', 'transcript_source', 'transcript_biotype', 'tag', 'ccds_id', 'exon_number', 'exon_id', 'exon_version', 'protein_id', 'protein_version', 'transcript_support_level']\n"
     ]
    }
   ],
   "source": [
    "from gtfparse import read_gtf\n",
    "\n",
    "# returns GTF with essential columns such as \"feature\", \"seqname\", \"start\", \"end\"\n",
    "# alongside the names of any optional keys which appeared in the attribute column\n",
    "df = read_gtf(os.path.join(data_dir, \"genes/Homo_sapiens.GRCh38.109.gtf\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "d4aa5ab3-af28-4c7f-b70e-56debe6f4e0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_source</th>\n",
       "      <th>transcript_biotype</th>\n",
       "      <th>tag</th>\n",
       "      <th>ccds_id</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_version</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_version</th>\n",
       "      <th>transcript_support_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>20</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>20</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,basic,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>CCDS30</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>20</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,basic,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>CCDS30</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003889014</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>20</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,basic,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>CCDS30</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000500094</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>20</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS,basic,Ensembl_canonical,MANE_Select</td>\n",
       "      <td>CCDS30</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3420356</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>32373</td>\n",
       "      <td>32528</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000277475</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3420357</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>31698</td>\n",
       "      <td>31841</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000277475</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3420358</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>gene</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000275405</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3420359</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>transcript</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000275405</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3420360</th>\n",
       "      <td>KI270713.1</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>exon</td>\n",
       "      <td>21861</td>\n",
       "      <td>22024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000275405</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>ensembl</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>basic,Ensembl_canonical</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003747303</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3420361 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            seqname          source          feature    start      end  score  \\\n",
       "0                 1  ensembl_havana             gene  1471765  1497848    NaN   \n",
       "1                 1  ensembl_havana       transcript  1471765  1497848    NaN   \n",
       "2                 1  ensembl_havana             exon  1471765  1472089    NaN   \n",
       "3                 1  ensembl_havana              CDS  1471885  1472089    NaN   \n",
       "4                 1  ensembl_havana      start_codon  1471885  1471887    NaN   \n",
       "...             ...             ...              ...      ...      ...    ...   \n",
       "3420356  KI270713.1         ensembl   five_prime_utr    32373    32528    NaN   \n",
       "3420357  KI270713.1         ensembl  three_prime_utr    31698    31841    NaN   \n",
       "3420358  KI270713.1         ensembl             gene    21861    22024    NaN   \n",
       "3420359  KI270713.1         ensembl       transcript    21861    22024    NaN   \n",
       "3420360  KI270713.1         ensembl             exon    21861    22024    NaN   \n",
       "\n",
       "        strand frame          gene_id gene_version  ... transcript_source  \\\n",
       "0            +     0  ENSG00000160072           20  ...                     \n",
       "1            +     0  ENSG00000160072           20  ...    ensembl_havana   \n",
       "2            +     0  ENSG00000160072           20  ...    ensembl_havana   \n",
       "3            +     0  ENSG00000160072           20  ...    ensembl_havana   \n",
       "4            +     0  ENSG00000160072           20  ...    ensembl_havana   \n",
       "...        ...   ...              ...          ...  ...               ...   \n",
       "3420356      -     0  ENSG00000277475            1  ...           ensembl   \n",
       "3420357      -     0  ENSG00000277475            1  ...           ensembl   \n",
       "3420358      -     0  ENSG00000275405            1  ...                     \n",
       "3420359      -     0  ENSG00000275405            1  ...           ensembl   \n",
       "3420360      -     0  ENSG00000275405            1  ...           ensembl   \n",
       "\n",
       "        transcript_biotype                                       tag ccds_id  \\\n",
       "0                                                                              \n",
       "1           protein_coding  CCDS,basic,Ensembl_canonical,MANE_Select  CCDS30   \n",
       "2           protein_coding  CCDS,basic,Ensembl_canonical,MANE_Select  CCDS30   \n",
       "3           protein_coding  CCDS,basic,Ensembl_canonical,MANE_Select  CCDS30   \n",
       "4           protein_coding  CCDS,basic,Ensembl_canonical,MANE_Select  CCDS30   \n",
       "...                    ...                                       ...     ...   \n",
       "3420356     protein_coding                   basic,Ensembl_canonical           \n",
       "3420357     protein_coding                   basic,Ensembl_canonical           \n",
       "3420358                                                                        \n",
       "3420359              snRNA                   basic,Ensembl_canonical           \n",
       "3420360              snRNA                   basic,Ensembl_canonical           \n",
       "\n",
       "        exon_number          exon_id exon_version       protein_id  \\\n",
       "0                                                                    \n",
       "1                                                                    \n",
       "2                 1  ENSE00003889014            1                    \n",
       "3                 1                                ENSP00000500094   \n",
       "4                 1                                                  \n",
       "...             ...              ...          ...              ...   \n",
       "3420356                                                              \n",
       "3420357                                                              \n",
       "3420358                                                              \n",
       "3420359                                                              \n",
       "3420360           1  ENSE00003747303            1                    \n",
       "\n",
       "        protein_version transcript_support_level  \n",
       "0                                                 \n",
       "1                                                 \n",
       "2                                                 \n",
       "3                     1                           \n",
       "4                                                 \n",
       "...                 ...                      ...  \n",
       "3420356                                       NA  \n",
       "3420357                                       NA  \n",
       "3420358                                           \n",
       "3420359                                       NA  \n",
       "3420360                                       NA  \n",
       "\n",
       "[3420361 rows x 26 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "c7272947-bebe-4c14-97f9-b07e162bc199",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_name', 'level', 'tag', 'transcript_id', 'transcript_type', 'transcript_name', 'transcript_support_level', 'havana_transcript', 'exon_number', 'exon_id', 'hgnc_id', 'havana_gene', 'ont', 'protein_id', 'ccdsid', 'artif_dupl']\n"
     ]
    }
   ],
   "source": [
    "df = read_gtf(os.path.join(data_dir, \"genes/gencode.v43.basic.annotation.gtf\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "e975dc7f-0501-43f6-bbb2-2b5d27d1bdcd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['gene', 'transcript', 'exon', 'CDS', 'start_codon', 'stop_codon',\n",
       "       'UTR', 'Selenocysteine'], dtype=object)"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.feature.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "0b0e325f-a5ff-4431-8564-e17ffe466227",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand',\n",
       "       'frame', 'gene_id', 'gene_type', 'gene_name', 'level', 'tag',\n",
       "       'transcript_id', 'transcript_type', 'transcript_name',\n",
       "       'transcript_support_level', 'havana_transcript', 'exon_number',\n",
       "       'exon_id', 'hgnc_id', 'havana_gene', 'ont', 'protein_id', 'ccdsid',\n",
       "       'artif_dupl'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "5aa4f899-45c0-49f4-a9c9-33e7548a6528",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0            \n",
       "1           1\n",
       "2           1\n",
       "3           1\n",
       "4           1\n",
       "           ..\n",
       "1998575    NA\n",
       "1998576    NA\n",
       "1998577      \n",
       "1998578    NA\n",
       "1998579    NA\n",
       "Name: transcript_support_level, Length: 1998580, dtype: object"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.transcript_support_level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "2945371d-4278-4ff1-884f-e6d59a45ae4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_enformer = df[(df[\"gene_type\"] == \"protein_coding\") & (df[\"feature\"] == \"transcript\")].sort_values([\"seqname\", \"start\"]).drop_duplicates(subset = \"gene_id\", keep = \"first\", inplace = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "cb9348d8-3bdd-4526-ad12-f1863845a0ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>artif_dupl</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1234649</th>\n",
       "      <td>chr12</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>54280193</td>\n",
       "      <td>54287087</td>\n",
       "      <td>NaN</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000135486.19</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>OTTHUMT00000405474.2</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>HGNC:5031</td>\n",
       "      <td>OTTHUMG00000169702.4</td>\n",
       "      <td></td>\n",
       "      <td>ENSP00000448617.1</td>\n",
       "      <td>CCDS41793.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname  source     feature     start       end  score strand frame  \\\n",
       "1234649   chr12  HAVANA  transcript  54280193  54287087    NaN      +     0   \n",
       "\n",
       "                    gene_id       gene_type  ... transcript_support_level  \\\n",
       "1234649  ENSG00000135486.19  protein_coding  ...                        1   \n",
       "\n",
       "            havana_transcript exon_number exon_id    hgnc_id  \\\n",
       "1234649  OTTHUMT00000405474.2                      HGNC:5031   \n",
       "\n",
       "                  havana_gene ont         protein_id       ccdsid artif_dupl  \n",
       "1234649  OTTHUMG00000169702.4      ENSP00000448617.1  CCDS41793.1             \n",
       "\n",
       "[1 rows x 26 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_enformer[df_enformer[\"gene_name\"] == \"HNRNPA1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74fe5bd0-cf9e-42d6-97bc-e6e4bf178450",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyfaidx\n",
    "import kipoiseq\n",
    "import functools\n",
    "from kipoiseq import Interval\n",
    "import random\n",
    "from data import str_to_one_hot, seq_indices_to_one_hot\n",
    "from einops import rearrange, reduce\n",
    "from einops.layers.torch import Rearrange"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd311057-8ad5-4bf9-a800-2fe19d2fa00e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def one_hot_encode(sequence):\n",
    "    return kipoiseq.transforms.functional.one_hot_dna(sequence).astype(np.float32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bd109e4-fc75-4049-94f6-ef9b8e0d423c",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_dir = '/omics/groups/OE0540/internal/users/mikulik/master_thesis/data/gcs_basenj/'\n",
    "human_fasta_f = 'hg38.ml.fa.gz'\n",
    "mouse_fasta_f = 'mm10.ml.fa.gz'\n",
    "human_fasta_gz_path = f\"{fasta_dir}/{human_fasta_f}\"\n",
    "mouse_fasta_gz_path = f\"{fasta_dir}/{mouse_fasta_f}\"\n",
    "human_fasta_path = human_fasta_gz_path.rstrip(\".gz\")\n",
    "mouse_fasta_path = mouse_fasta_gz_path.rstrip(\".gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09479aae-f8ed-4145-8617-3959592b69e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "class FastaStringExtractor:\n",
    "    \n",
    "    def __init__(self, fasta_file):\n",
    "        self.fasta = pyfaidx.Fasta(fasta_file)\n",
    "        self._chromosome_sizes = {k: len(v) for k, v in self.fasta.items()}\n",
    "\n",
    "    def extract(self, interval: Interval, **kwargs) -> str:\n",
    "        # Truncate interval if it extends beyond the chromosome lengths.\n",
    "        chromosome_length = self._chromosome_sizes[interval.chrom]\n",
    "        trimmed_interval = Interval(interval.chrom,\n",
    "                                    max(interval.start, 0),\n",
    "                                    min(interval.end, chromosome_length),\n",
    "                                    )\n",
    "        # pyfaidx wants a 1-based interval\n",
    "        sequence = str(self.fasta.get_seq(trimmed_interval.chrom,\n",
    "                                          trimmed_interval.start + 1,\n",
    "                                          trimmed_interval.stop).seq).upper()\n",
    "        # Fill truncated values with N's.\n",
    "        pad_upstream = 'N' * max(-interval.start, 0)\n",
    "        pad_downstream = 'N' * max(interval.end - chromosome_length, 0)\n",
    "        return pad_upstream + sequence + pad_downstream\n",
    "\n",
    "    def close(self):\n",
    "        return self.fasta.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4e9e327-fd43-4058-a81b-d998cbee3b6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_reader = FastaStringExtractor(mouse_fasta_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc6719a0-f785-41e6-81f5-bc91a7b5dfc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = Interval(\"chr1\", 0, 10)\n",
    "len(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5237bcec-f1d6-4227-8db5-903d9881336c",
   "metadata": {},
   "outputs": [],
   "source": [
    "SEQUENCE_LENGTH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0efd2789-ae4f-4087-b551-ac7986cece45",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq = one_hot_encode(fasta_reader.extract("
   ]
  },
  {
   "cell_type": "markdown",
   "id": "851c02e3-3e4b-4d53-817d-f90c0afc05ed",
   "metadata": {},
   "source": [
    "# Basenji1 data preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "b587d5e6-582e-4572-bc90-ddc56d0d41f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>subset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr4</td>\n",
       "      <td>717798</td>\n",
       "      <td>1422902</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>1478646</td>\n",
       "      <td>2112514</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr4</td>\n",
       "      <td>2115100</td>\n",
       "      <td>4955618</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr4</td>\n",
       "      <td>4960135</td>\n",
       "      <td>5441427</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr4</td>\n",
       "      <td>5486841</td>\n",
       "      <td>5763763</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4433</th>\n",
       "      <td>chr17</td>\n",
       "      <td>45513571</td>\n",
       "      <td>46019356</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4434</th>\n",
       "      <td>chr17</td>\n",
       "      <td>47536713</td>\n",
       "      <td>48042498</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4435</th>\n",
       "      <td>chr17</td>\n",
       "      <td>60906210</td>\n",
       "      <td>61407715</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4436</th>\n",
       "      <td>chr17</td>\n",
       "      <td>71805586</td>\n",
       "      <td>72224641</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4437</th>\n",
       "      <td>chr17</td>\n",
       "      <td>78124600</td>\n",
       "      <td>78652738</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4438 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr     start       end subset\n",
       "0      chr4    717798   1422902  train\n",
       "1      chr4   1478646   2112514  train\n",
       "2      chr4   2115100   4955618  train\n",
       "3      chr4   4960135   5441427  train\n",
       "4      chr4   5486841   5763763  train\n",
       "...     ...       ...       ...    ...\n",
       "4433  chr17  45513571  46019356   test\n",
       "4434  chr17  47536713  48042498   test\n",
       "4435  chr17  60906210  61407715   test\n",
       "4436  chr17  71805586  72224641   test\n",
       "4437  chr17  78124600  78652738   test\n",
       "\n",
       "[4438 rows x 4 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "basenji_dir = \"/omics/groups/OE0540/internal/users/mikulik/master_thesis/basenji/tutorials/data/heart_l131k\"\n",
    "contigs = pd.read_csv(\"%s/contigs.bed\" % basenji_dir, sep=\"\\t\", header=None)\n",
    "contigs.columns = [\"chr\", \"start\", \"end\", \"subset\"]\n",
    "contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "21b05e7d-333e-4d79-b2c5-a05df805a955",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>subset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>140425791</td>\n",
       "      <td>140556863</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr16</td>\n",
       "      <td>27143973</td>\n",
       "      <td>27275045</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr14</td>\n",
       "      <td>72972403</td>\n",
       "      <td>73103475</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>67943788</td>\n",
       "      <td>68074860</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr16</td>\n",
       "      <td>58609927</td>\n",
       "      <td>58740999</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1853</th>\n",
       "      <td>chr5</td>\n",
       "      <td>163844630</td>\n",
       "      <td>163975702</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1854</th>\n",
       "      <td>chr9</td>\n",
       "      <td>21934071</td>\n",
       "      <td>22065143</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1855</th>\n",
       "      <td>chr5</td>\n",
       "      <td>18279816</td>\n",
       "      <td>18410888</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1856</th>\n",
       "      <td>chr2</td>\n",
       "      <td>101839419</td>\n",
       "      <td>101970491</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1857</th>\n",
       "      <td>chr7</td>\n",
       "      <td>27084572</td>\n",
       "      <td>27215644</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1858 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr      start        end subset\n",
       "0      chr2  140425791  140556863  train\n",
       "1     chr16   27143973   27275045  train\n",
       "2     chr14   72972403   73103475  train\n",
       "3      chr1   67943788   68074860  train\n",
       "4     chr16   58609927   58740999  train\n",
       "...     ...        ...        ...    ...\n",
       "1853   chr5  163844630  163975702   test\n",
       "1854   chr9   21934071   22065143   test\n",
       "1855   chr5   18279816   18410888   test\n",
       "1856   chr2  101839419  101970491   test\n",
       "1857   chr7   27084572   27215644   test\n",
       "\n",
       "[1858 rows x 4 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences = pd.read_csv(\"%s/sequences.bed\" % basenji_dir, sep=\"\\t\", header=None)\n",
    "sequences.columns = [\"chr\", \"start\", \"end\", \"subset\"]\n",
    "sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "fd8d9cbf-3cf5-4aad-80e8-4540f5d0e582",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>subset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4756458</td>\n",
       "      <td>4887530</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>863</th>\n",
       "      <td>chr1</td>\n",
       "      <td>4887530</td>\n",
       "      <td>5018602</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1806</th>\n",
       "      <td>chr1</td>\n",
       "      <td>5867723</td>\n",
       "      <td>5998795</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331</th>\n",
       "      <td>chr1</td>\n",
       "      <td>6604225</td>\n",
       "      <td>6735297</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8393983</td>\n",
       "      <td>8525055</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1352</th>\n",
       "      <td>chrX</td>\n",
       "      <td>145940701</td>\n",
       "      <td>146071773</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1196</th>\n",
       "      <td>chrX</td>\n",
       "      <td>147210564</td>\n",
       "      <td>147341636</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>chrX</td>\n",
       "      <td>147472708</td>\n",
       "      <td>147603780</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>494</th>\n",
       "      <td>chrX</td>\n",
       "      <td>151409506</td>\n",
       "      <td>151540578</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278</th>\n",
       "      <td>chrX</td>\n",
       "      <td>152327099</td>\n",
       "      <td>152458171</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1858 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr      start        end subset\n",
       "182   chr1    4756458    4887530  train\n",
       "863   chr1    4887530    5018602  train\n",
       "1806  chr1    5867723    5998795   test\n",
       "331   chr1    6604225    6735297  train\n",
       "299   chr1    8393983    8525055  train\n",
       "...    ...        ...        ...    ...\n",
       "1352  chrX  145940701  146071773  train\n",
       "1196  chrX  147210564  147341636  train\n",
       "361   chrX  147472708  147603780  train\n",
       "494   chrX  151409506  151540578  train\n",
       "278   chrX  152327099  152458171  train\n",
       "\n",
       "[1858 rows x 4 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequences.sort_values(by=[\"chr\", \"start\"], ascending=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "4595f9ea-0488-48f3-badc-c92a42a0cb53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>subset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr15</td>\n",
       "      <td>94421238</td>\n",
       "      <td>94617846</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr5</td>\n",
       "      <td>83350786</td>\n",
       "      <td>83547394</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr5</td>\n",
       "      <td>144532793</td>\n",
       "      <td>144729401</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr15</td>\n",
       "      <td>29905834</td>\n",
       "      <td>30102442</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9</td>\n",
       "      <td>115566410</td>\n",
       "      <td>115763018</td>\n",
       "      <td>train</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11586</th>\n",
       "      <td>chr5</td>\n",
       "      <td>102585056</td>\n",
       "      <td>102781664</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11587</th>\n",
       "      <td>chr6</td>\n",
       "      <td>22661358</td>\n",
       "      <td>22857966</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11588</th>\n",
       "      <td>chr13</td>\n",
       "      <td>89524907</td>\n",
       "      <td>89721515</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11589</th>\n",
       "      <td>chr14</td>\n",
       "      <td>68016258</td>\n",
       "      <td>68212866</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11590</th>\n",
       "      <td>chr6</td>\n",
       "      <td>47467302</td>\n",
       "      <td>47663910</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11591 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr      start        end subset\n",
       "0      chr15   94421238   94617846  train\n",
       "1       chr5   83350786   83547394  train\n",
       "2       chr5  144532793  144729401  train\n",
       "3      chr15   29905834   30102442  train\n",
       "4       chr9  115566410  115763018  train\n",
       "...      ...        ...        ...    ...\n",
       "11586   chr5  102585056  102781664   test\n",
       "11587   chr6   22661358   22857966   test\n",
       "11588  chr13   89524907   89721515   test\n",
       "11589  chr14   68016258   68212866   test\n",
       "11590   chr6   47467302   47663910   test\n",
       "\n",
       "[11591 rows x 4 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "sequences = pd.read_csv(\"/omics/groups/OE0540/internal/users/mikulik/master_thesis/basenji/tutorials/output/sequences.bed\", sep=\"\\t\", header=None)\n",
    "sequences.columns = [\"chr\", \"start\", \"end\", \"subset\"]\n",
    "sequences"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "basenji",
   "language": "python",
   "name": "basenji"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
