{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import pybedtools\n",
    "import kipoiseq\n",
    "import os\n",
    "import pandas as  pd\n",
    "import numpy as np\n",
    "from data_utils import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Collecting CAGE counts at TSS:\n",
    "\n",
    "**For each gene collect CAGE counts at all alternative TSS**\n",
    "\n",
    "1. get bin information for all alternative TSS\n",
    "2. get input region that overlaps with each TSS\n",
    "3. extract the count value at the TSS bin from targets and predictions for the overlapping input region\n",
    "4. sum all raw counts at all alternative TSS bin (+ 2 neighbouring bins)\n",
    "\n",
    "We are left with one single gene expression value per gene. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Find overlap between unique TSS and test set regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "odcf = False\n",
    "if odcf:\n",
    "    data_dir =\"/omics/groups/OE0540/internal/users/mikulik/master_thesis/data/gcs_basenj\"\n",
    "else:\n",
    "    data_dir = \"/data/mikulik/mnt/gcs_basenj/\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(213986, 16)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_124460/3491214571.py:2: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  tss = pd.read_csv(os.path.join(data_dir, \"unique_tss.csv\"), header=0, index_col=0)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_stable_id</th>\n",
       "      <th>version</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>tss</th>\n",
       "      <th>ensembl_canonical</th>\n",
       "      <th>RefSeq_match_transcript_MANE_select</th>\n",
       "      <th>TSL</th>\n",
       "      <th>gene_stable_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>strand_aware_start</th>\n",
       "      <th>strand_aware_end</th>\n",
       "      <th>tss_coord</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10474</th>\n",
       "      <td>ENST00000431340</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>Y</td>\n",
       "      <td>1</td>\n",
       "      <td>8280198</td>\n",
       "      <td>8282209</td>\n",
       "      <td>8280198</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>ENSG00000215601</td>\n",
       "      <td>TSPY24P</td>\n",
       "      <td>8280198</td>\n",
       "      <td>8282209</td>\n",
       "      <td>chrY:8280198_8280199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10478</th>\n",
       "      <td>ENST00000415010</td>\n",
       "      <td>1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>Y</td>\n",
       "      <td>-1</td>\n",
       "      <td>7853686</td>\n",
       "      <td>7854876</td>\n",
       "      <td>7854876</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>ENSG00000215603</td>\n",
       "      <td>ZNF92P1Y</td>\n",
       "      <td>7854876</td>\n",
       "      <td>7853686</td>\n",
       "      <td>chrY:7854876_7854877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10480</th>\n",
       "      <td>ENST00000430062</td>\n",
       "      <td>1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>Y</td>\n",
       "      <td>-1</td>\n",
       "      <td>5207215</td>\n",
       "      <td>5208069</td>\n",
       "      <td>5208069</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>ENSG00000231341</td>\n",
       "      <td>VDAC1P6</td>\n",
       "      <td>5208069</td>\n",
       "      <td>5207215</td>\n",
       "      <td>chrY:5208069_5208070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10484</th>\n",
       "      <td>ENST00000449381</td>\n",
       "      <td>1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>Y</td>\n",
       "      <td>-1</td>\n",
       "      <td>9610571</td>\n",
       "      <td>9621276</td>\n",
       "      <td>9621276</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>ENSG00000231436</td>\n",
       "      <td>RBMY3AP</td>\n",
       "      <td>9621276</td>\n",
       "      <td>9610571</td>\n",
       "      <td>chrY:9621276_9621277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10489</th>\n",
       "      <td>ENST00000651261</td>\n",
       "      <td>1</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>Y</td>\n",
       "      <td>-1</td>\n",
       "      <td>7694707</td>\n",
       "      <td>7809267</td>\n",
       "      <td>7809267</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000234795</td>\n",
       "      <td>RFTN1P1</td>\n",
       "      <td>7809267</td>\n",
       "      <td>7694707</td>\n",
       "      <td>chrY:7809267_7809268</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript_stable_id  version         transcript_type chrom  strand   \n",
       "10474      ENST00000431340        1  unprocessed_pseudogene     Y       1  \\\n",
       "10478      ENST00000415010        1    processed_pseudogene     Y      -1   \n",
       "10480      ENST00000430062        1    processed_pseudogene     Y      -1   \n",
       "10484      ENST00000449381        1  unprocessed_pseudogene     Y      -1   \n",
       "10489      ENST00000651261        1    processed_transcript     Y      -1   \n",
       "\n",
       "         start      end      tss  ensembl_canonical   \n",
       "10474  8280198  8282209  8280198                1.0  \\\n",
       "10478  7853686  7854876  7854876                1.0   \n",
       "10480  5207215  5208069  5208069                1.0   \n",
       "10484  9610571  9621276  9621276                1.0   \n",
       "10489  7694707  7809267  7809267                NaN   \n",
       "\n",
       "      RefSeq_match_transcript_MANE_select    TSL   gene_stable_id gene_name   \n",
       "10474                                 NaN  tslNA  ENSG00000215601   TSPY24P  \\\n",
       "10478                                 NaN  tslNA  ENSG00000215603  ZNF92P1Y   \n",
       "10480                                 NaN  tslNA  ENSG00000231341   VDAC1P6   \n",
       "10484                                 NaN  tslNA  ENSG00000231436   RBMY3AP   \n",
       "10489                                 NaN    NaN  ENSG00000234795   RFTN1P1   \n",
       "\n",
       "       strand_aware_start  strand_aware_end             tss_coord  \n",
       "10474             8280198           8282209  chrY:8280198_8280199  \n",
       "10478             7854876           7853686  chrY:7854876_7854877  \n",
       "10480             5208069           5207215  chrY:5208069_5208070  \n",
       "10484             9621276           9610571  chrY:9621276_9621277  \n",
       "10489             7809267           7694707  chrY:7809267_7809268  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in all unique tss \n",
    "tss = pd.read_csv(os.path.join(data_dir, \"unique_tss.csv\"), header=0, index_col=0)\n",
    "print(tss.shape)\n",
    "tss.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reduce the length of the test set regions to 114,688 bp and write them to a bed file. The same for the unique tss from Karollus et al. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_input_regions_to_bed(data_dir, species, subset):\n",
    "    '''\n",
    "    data_dir (str): path to input region bed file for both species\n",
    "    seq_bed (str): name of the bed file that stores the input regions for the species specified\n",
    "    species (str): one of human or mouse\n",
    "    subset (str): on of train/test/valid\n",
    "    '''\n",
    "    # read in the input regions\n",
    "    region_df = pd.read_csv(os.path.join(data_dir, species, \"sequences.bed\"), sep=\"\\t\", header=None)\n",
    "    region_df.columns = [\"Chromosome\", \"Start\", \"End\", \"subset\"]\n",
    "    region_df.reset_index(inplace=True)\n",
    "    region_df.columns = [\"index\", \"Chromosome\", \"Start\", \"End\", \"subset\"]\n",
    "    region_df = region_df[[\"Chromosome\", \"Start\", \"End\", \"subset\", \"index\"]]\n",
    "    region_df_test = region_df[region_df[\"subset\"] == subset].copy()\n",
    "    print(region_df_test.shape)\n",
    "\n",
    "    # crop the sequence length to 114688, to contain only the center 896 bins for which we make predictions\n",
    "    seq_len_crop = 131072 - (128 * 64 * 2)\n",
    "    print(f\"Crop to {seq_len_crop}\")\n",
    "    for row, region in region_df_test.iterrows():\n",
    "        interval = kipoiseq.Interval(region.Chromosome, region.Start, region.End).resize(seq_len_crop) \n",
    "        region_df_test.at[row, \"Start\"] = interval.start\n",
    "        region_df_test.at[row, \"End\"] = interval.end\n",
    "        assert region_df_test.at[row, \"End\"] - region_df_test.at[row, \"Start\"] == seq_len_crop\n",
    "    region_bed = pybedtools.BedTool.from_dataframe(region_df_test)\n",
    "    print(f'Save at {os.path.join(data_dir, species, f\"{subset}_{species}_center_114688.bed\")}')\n",
    "    region_bed.saveas(os.path.join(data_dir, species, f\"{subset}_{species}_center_114688.bed\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(34021, 5)\n",
      "Crop to 114688\n",
      "Save at /data/mikulik/mnt/gcs_basenj/human/train_human_center_114688.bed\n",
      "(1937, 5)\n",
      "Crop to 114688\n",
      "Save at /data/mikulik/mnt/gcs_basenj/human/test_human_center_114688.bed\n",
      "(2213, 5)\n",
      "Crop to 114688\n",
      "Save at /data/mikulik/mnt/gcs_basenj/human/valid_human_center_114688.bed\n"
     ]
    }
   ],
   "source": [
    "# write all region subsets to separate bed files \n",
    "write_input_regions_to_bed(data_dir, \"human\", \"train\")\n",
    "write_input_regions_to_bed(data_dir, \"human\", \"test\")\n",
    "write_input_regions_to_bed(data_dir, \"human\", \"valid\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr18</td>\n",
       "      <td>936578</td>\n",
       "      <td>1051266</td>\n",
       "      <td>train</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>113639139</td>\n",
       "      <td>113753827</td>\n",
       "      <td>train</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr11</td>\n",
       "      <td>18435912</td>\n",
       "      <td>18550600</td>\n",
       "      <td>train</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16</td>\n",
       "      <td>85813873</td>\n",
       "      <td>85928561</td>\n",
       "      <td>train</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3</td>\n",
       "      <td>158394380</td>\n",
       "      <td>158509068</td>\n",
       "      <td>train</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34016</th>\n",
       "      <td>chr7</td>\n",
       "      <td>50523314</td>\n",
       "      <td>50638002</td>\n",
       "      <td>train</td>\n",
       "      <td>34016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34017</th>\n",
       "      <td>chr7</td>\n",
       "      <td>135610961</td>\n",
       "      <td>135725649</td>\n",
       "      <td>train</td>\n",
       "      <td>34017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34018</th>\n",
       "      <td>chr4</td>\n",
       "      <td>189012390</td>\n",
       "      <td>189127078</td>\n",
       "      <td>train</td>\n",
       "      <td>34018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34019</th>\n",
       "      <td>chr4</td>\n",
       "      <td>10446291</td>\n",
       "      <td>10560979</td>\n",
       "      <td>train</td>\n",
       "      <td>34019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34020</th>\n",
       "      <td>chr1</td>\n",
       "      <td>185738464</td>\n",
       "      <td>185853152</td>\n",
       "      <td>train</td>\n",
       "      <td>34020</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34021 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0          1          2      3      4\n",
       "0      chr18     936578    1051266  train      0\n",
       "1       chr4  113639139  113753827  train      1\n",
       "2      chr11   18435912   18550600  train      2\n",
       "3      chr16   85813873   85928561  train      3\n",
       "4       chr3  158394380  158509068  train      4\n",
       "...      ...        ...        ...    ...    ...\n",
       "34016   chr7   50523314   50638002  train  34016\n",
       "34017   chr7  135610961  135725649  train  34017\n",
       "34018   chr4  189012390  189127078  train  34018\n",
       "34019   chr4   10446291   10560979  train  34019\n",
       "34020   chr1  185738464  185853152  train  34020\n",
       "\n",
       "[34021 rows x 5 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_regions = pd.read_csv(os.path.join(data_dir, \"human\", \"train_human_center_114688.bed\"),sep=\"\\t\",header=None)\n",
    "train_regions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reduce the length of the test set regions to 114,688 bp and write them to a bed file. The same for the unique tss from Karollus et al. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pybedtools.bedtool.BedTool'>\n",
      "213986\n"
     ]
    }
   ],
   "source": [
    "# write the tss coordinates from Karollus to a bed file\n",
    "tss_sub = tss[[\"tss_coord\", \"gene_stable_id\"]].copy()\n",
    "tss_sub[[\"Chromosome\", \"tmp\"]] = tss_sub.tss_coord.str.split(\":\", n=1, expand=True)\n",
    "tss_sub[[\"Start\", \"End\"]] = tss_sub.tmp.str.split(\"_\", expand=True)\n",
    "tss_bed = pybedtools.BedTool.from_dataframe(tss_sub[[\"Chromosome\", \"Start\", \"End\", \"gene_stable_id\"]])\n",
    "print(type(tss_bed))\n",
    "print(len(tss_bed))\n",
    "tss_bed\n",
    "# save\n",
    "save = False\n",
    "if save:\n",
    "    tss_bed.saveas(os.path.join(data_dir, \"tss_all.bed\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pybedtools.bedtool.BedTool'>\n",
      "74633\n"
     ]
    }
   ],
   "source": [
    "# write only the tss coordinates of protein-coding genes from Karollus to a bed file\n",
    "tss = tss[tss.transcript_type == \"protein_coding\" ]\n",
    "tss_sub = tss[[\"tss_coord\", \"gene_stable_id\"]].copy()\n",
    "tss_sub[[\"Chromosome\", \"tmp\"]] = tss_sub.tss_coord.str.split(\":\", n=1, expand=True)\n",
    "tss_sub[[\"Start\", \"End\"]] = tss_sub.tmp.str.split(\"_\", expand=True)\n",
    "tss_bed = pybedtools.BedTool.from_dataframe(tss_sub[[\"Chromosome\", \"Start\", \"End\", \"gene_stable_id\"]])\n",
    "print(type(tss_bed))\n",
    "print(len(tss_bed))\n",
    "tss_bed\n",
    "# save\n",
    "save = False\n",
    "if save:\n",
    "    tss_bed.saveas(os.path.join(data_dir, \"tss_protein_coding.bed\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Find TSSs in the input sequences\n",
    "\n",
    "Now  we want to know which of the regions in the test set overlap with at least one of the TSSs.\n",
    "\n",
    "The bedtools overlap can be runf with the script `/omics/groups/OE0540/internal/users/mikulik/master_thesis/data/gcs_basenj/bedtools_overlap_karollus_tss.py`. Make sure to use the tss coordinates first and the input region bed file second! Otherwise the `get_overlap` function will not work anymore. Some of the regions in the test set do not contain any of the TSSs!\n",
    "\n",
    "\n",
    "**For all tss**\n",
    "\n",
    "```\n",
    "python bedtools_overlap_karollus_tss.py tss_all.bed /human/train_human_center_114688.bed /human/overlap_train.bed /omics/groups/OE0540/internal/users/mikulik/master_thesis/data/gcs_basenj/\n",
    "```\n",
    "\n",
    "```\n",
    "python bedtools_overlap_karollus_tss.py tss_all.bed /human/valid_human_center_114688.bed /human/overlap_valid.bed /omics/groups/OE0540/internal/users/mikulik/master_thesis/data/gcs_basenj/\n",
    "```\n",
    "\n",
    "\n",
    "**For only protein-coding gene tss**\n",
    "```\n",
    "python bedtools_overlap_karollus_tss.py tss_protein_coding.bed /human/train_human_center_114688.bed /human/overlap_train_protein_coding.bed /omics/groups/OE0540/internal/users/mikulik/master_thesis/data/gcs_basenj/\n",
    "```\n",
    "```\n",
    "python bedtools_overlap_karollus_tss.py tss_protein_coding.bed /human/valid_human_center_114688.bed /human/overlap_valid_protein_coding.bed /omics/groups/OE0540/internal/users/mikulik/master_thesis/data/gcs_basenj/\n",
    "```\n",
    "```\n",
    "python bedtools_overlap_karollus_tss.py tss_protein_coding.bed /human/test_human_center_114688.bed /human/overlap_test_protein_coding.bed /omics/groups/OE0540/internal/users/mikulik/master_thesis/data/gcs_basenj/\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_overlap(data_dir, file_name):\n",
    "    overlap_tss = pd.read_csv(os.path.join(data_dir, f\"{file_name}.bed\"), sep=\"\\t\", header=None)\n",
    "    overlap_tss.columns = [\"tss_chrom\", \"tss_start\", \"tss_end\", \"gene_id\",  \"test_region_chr\", \"test_region_start\", \"test_region_end\", \"subset\", \"region_index\",  \"overlap_size\"]\n",
    "    overlap_tss[\"region_coord\"] = overlap_tss.test_region_chr + \":\" + overlap_tss.test_region_start.astype(\"str\") + \"_\" + overlap_tss.test_region_end.astype(\"str\")\n",
    "    overlap_tss[\"tss_coord\"] = overlap_tss.tss_chrom + \":\" + overlap_tss.tss_start.astype(\"str\") + \"_\" + overlap_tss.tss_end.astype(\"str\")\n",
    "    overlap_tss[\"seq_len\"] = overlap_tss.test_region_end -  overlap_tss.test_region_start\n",
    "    # calculate the position of the TSS within the test region\n",
    "    overlap_tss[\"pos_tss\"] = np.abs(overlap_tss.test_region_start - overlap_tss.tss_start)\n",
    "    assert 0 < overlap_tss.pos_tss.max() < 114688 \n",
    "    # convert the position of the TSS to the 128 bp bin number\n",
    "    overlap_tss[\"bin\"] = np.round(overlap_tss.pos_tss // 128)\n",
    "    print((overlap_tss.bin < 0).sum())\n",
    "    assert all(overlap_tss.bin >= 0)\n",
    "\n",
    "    overlap_tss  = overlap_tss.sort_values(by=\"region_index\", ascending=True)\n",
    "    #print(f\"The number of input sequences in the test set is {len(region_df_test)}, while the number of regions which have at least one TSS overlapping is {len(overlap_tss.region_index.unique())}\")\n",
    "    assert overlap_tss.bin.max() < 896\n",
    "    assert overlap_tss.bin.min() >= 0\n",
    "    return overlap_tss\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>subset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>36234</th>\n",
       "      <td>chr10</td>\n",
       "      <td>37555537</td>\n",
       "      <td>37686609</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36235</th>\n",
       "      <td>chr14</td>\n",
       "      <td>87048845</td>\n",
       "      <td>87179917</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36236</th>\n",
       "      <td>chrX</td>\n",
       "      <td>136527085</td>\n",
       "      <td>136658157</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36237</th>\n",
       "      <td>chr11</td>\n",
       "      <td>34042349</td>\n",
       "      <td>34173421</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36238</th>\n",
       "      <td>chr19</td>\n",
       "      <td>10352757</td>\n",
       "      <td>10483829</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38166</th>\n",
       "      <td>chr19</td>\n",
       "      <td>33204702</td>\n",
       "      <td>33335774</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38167</th>\n",
       "      <td>chr14</td>\n",
       "      <td>41861379</td>\n",
       "      <td>41992451</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38168</th>\n",
       "      <td>chr19</td>\n",
       "      <td>30681544</td>\n",
       "      <td>30812616</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38169</th>\n",
       "      <td>chr14</td>\n",
       "      <td>61473198</td>\n",
       "      <td>61604270</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38170</th>\n",
       "      <td>chr2</td>\n",
       "      <td>129664471</td>\n",
       "      <td>129795543</td>\n",
       "      <td>test</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1937 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr      start        end subset\n",
       "36234  chr10   37555537   37686609   test\n",
       "36235  chr14   87048845   87179917   test\n",
       "36236   chrX  136527085  136658157   test\n",
       "36237  chr11   34042349   34173421   test\n",
       "36238  chr19   10352757   10483829   test\n",
       "...      ...        ...        ...    ...\n",
       "38166  chr19   33204702   33335774   test\n",
       "38167  chr14   41861379   41992451   test\n",
       "38168  chr19   30681544   30812616   test\n",
       "38169  chr14   61473198   61604270   test\n",
       "38170   chr2  129664471  129795543   test\n",
       "\n",
       "[1937 rows x 4 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions = pd.read_csv(os.path.join(data_dir, \"human\", \"sequences.bed\"), sep=\"\\t\", header=None)\n",
    "regions.columns = [\"chr\", \"start\", \"end\",\"subset\"]\n",
    "regions[regions.subset ==\"test\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_tss = get_overlap(os.path.join(data_dir, \"human\"), \"overlap_test_protein_coding\")\n",
    "overlap_tss.region_index.isin(regions[regions.subset ==\"test\"].index).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "ds = EnformerDataset(\"human\", \n",
    "                     \"test\",\n",
    "                     131072,\n",
    "                     data_dir, \n",
    "                     human_fasta_path = os.path.join(data_dir, \"hg38.ml.fa\"), \n",
    "                     mouse_fasta_path = os.path.join(data_dir, \"mm10.ml.fa\"), \n",
    "                     random=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlap_tss.region_index.isin(ds.input_sequence.region_df[\"index\"]).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def update_gene_dict(single_region_df:pd.DataFrame, gene_dict:dict, target:np.array, prediction:np.array):\n",
    "    for i, row in single_region_df.iterrows():\n",
    "        gene = row.gene_id\n",
    "        bin = int(row.bin)\n",
    "        counts_target = target[:, bin-1 : bin+2, :].sum(axis=1).squeeze()\n",
    "        counts_prediction = prediction[:, bin-1 : bin+2, :].sum(axis=1).squeeze()\n",
    "        #print(counts_prediction)\n",
    "        gene_dict[\"pred\"][gene] = gene_dict[\"pred\"][gene] + counts_prediction\n",
    "        gene_dict[\"tar\"][gene] = gene_dict[\"tar\"][gene] + counts_target\n",
    "        #print(gene_dict)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "transformer",
   "language": "python",
   "name": "transformer"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
